<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>checkAndPrepInputs_perFOV — checkAndPrepInputs_perFOV • FastReseg</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="checkAndPrepInputs_perFOV — checkAndPrepInputs_perFOV"><meta name="description" content="supporting function for fastReseg_perFOV_full_process, checks and preps inputs for full resegmentation pipeline on the provided transcript_df and calculates distance cutoffs and set values for config_spatNW_transcript and leiden_config if not provided."><meta property="og:description" content="supporting function for fastReseg_perFOV_full_process, checks and preps inputs for full resegmentation pipeline on the provided transcript_df and calculates distance cutoffs and set values for config_spatNW_transcript and leiden_config if not provided."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-#E6F3FC" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FastReseg</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="external-link dropdown-item" href="https://www.nature.com/articles/s41598-025-08733-5">Working principles and Manuscript</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Workflow and core functions</h6></li>
    <li><a class="dropdown-item" href="../articles/tutorial.html">FastReseg to detect and correct segmentation error in spatial transcriptome data</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Nanostring-Biostats/FastReseg/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>checkAndPrepInputs_perFOV</h1>
      <small class="dont-index">Source: <a href="https://github.com/Nanostring-Biostats/FastReseg/blob/HEAD/R/fastReseg_perFOV_full_process.R" class="external-link"><code>R/fastReseg_perFOV_full_process.R</code></a></small>
      <div class="d-none name"><code>checkAndPrepInputs_perFOV.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>supporting function for <code>fastReseg_perFOV_full_process</code>, checks and preps inputs for full resegmentation pipeline on the provided <code>transcript_df</code> and calculates distance cutoffs and set values for <code>config_spatNW_transcript</code> and <code>leiden_config</code> if not provided.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">checkAndPrepInputs_perFOV</span><span class="op">(</span></span>
<span>  <span class="va">all_celltypes</span>,</span>
<span>  <span class="va">all_genes</span>,</span>
<span>  <span class="va">transcript_df</span>,</span>
<span>  transID_coln <span class="op">=</span> <span class="st">"UMI_transID"</span>,</span>
<span>  transGene_coln <span class="op">=</span> <span class="st">"target"</span>,</span>
<span>  cellID_coln <span class="op">=</span> <span class="st">"UMI_cellID"</span>,</span>
<span>  spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span>, <span class="st">"z"</span><span class="op">)</span>,</span>
<span>  extracellular_cellID <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  flagModel_TransNum_cutoff <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  molecular_distance_cutoff <span class="op">=</span> <span class="fl">2.7</span>,</span>
<span>  cellular_distance_cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  score_baseline <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  lowerCutoff_transNum <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  higherCutoff_transNum <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  groupTranscripts_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dbscan"</span>, <span class="st">"delaunay"</span><span class="op">)</span>,</span>
<span>  spatialMergeCheck_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"leidenCut"</span>, <span class="st">"geometryDiff"</span><span class="op">)</span>,</span>
<span>  cutoff_spatialMerge <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  leiden_config <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  config_spatNW_transcript <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-all-celltypes">all_celltypes<a class="anchor" aria-label="anchor" href="#arg-all-celltypes"></a></dt>
<dd><p>vector of all cell types consider in the analysis as listed in columns of <code>score_GeneMatrix</code> in parent function</p></dd>


<dt id="arg-all-genes">all_genes<a class="anchor" aria-label="anchor" href="#arg-all-genes"></a></dt>
<dd><p>vector of all genes consider in the analysis as listed in rows of <code>score_GeneMatrix</code> in parent function</p></dd>


<dt id="arg-transcript-df">transcript_df<a class="anchor" aria-label="anchor" href="#arg-transcript-df"></a></dt>
<dd><p>the data.frame for each transcript with columns for transcript_id, target or gene name, original cell_id, spatial coordinates.</p></dd>


<dt id="arg-transid-coln">transID_coln<a class="anchor" aria-label="anchor" href="#arg-transid-coln"></a></dt>
<dd><p>the column name of transcript_ID in <code>transcript_df</code></p></dd>


<dt id="arg-transgene-coln">transGene_coln<a class="anchor" aria-label="anchor" href="#arg-transgene-coln"></a></dt>
<dd><p>the column name of target or gene name in <code>transcript_df</code></p></dd>


<dt id="arg-cellid-coln">cellID_coln<a class="anchor" aria-label="anchor" href="#arg-cellid-coln"></a></dt>
<dd><p>the column name of cell_ID in <code>transcript_df</code></p></dd>


<dt id="arg-spatlocs-colns">spatLocs_colns<a class="anchor" aria-label="anchor" href="#arg-spatlocs-colns"></a></dt>
<dd><p>column names for 1st, 2nd and optional 3rd dimension of spatial coordinates in <code>transcript_df</code></p></dd>


<dt id="arg-extracellular-cellid">extracellular_cellID<a class="anchor" aria-label="anchor" href="#arg-extracellular-cellid"></a></dt>
<dd><p>a vector of cell_ID for extracellular transcripts which would be removed from the resegmention pipeline (default = NULL)</p></dd>


<dt id="arg-flagmodel-transnum-cutoff">flagModel_TransNum_cutoff<a class="anchor" aria-label="anchor" href="#arg-flagmodel-transnum-cutoff"></a></dt>
<dd><p>the cutoff of transcript number to do spatial modeling for identification of wrongly segmented cells (default = 50)</p></dd>


<dt id="arg-molecular-distance-cutoff">molecular_distance_cutoff<a class="anchor" aria-label="anchor" href="#arg-molecular-distance-cutoff"></a></dt>
<dd><p>maximum molecule-to-molecule distance within connected transcript group, same unit as input spatial coordinate (default = 2.7 micron).
If set to NULL, the pipeline would first randomly choose no more than 2500 cells from up to 10 random picked ROIs with search radius to be 5 times of <code>cellular_distance_cutoff</code>, and then calculate the minimal molecular distance between picked cells. The pipeline would further use the 5 times of 90% quantile of minimal molecular distance as <code>molecular_distance_cutoff</code>. This calculation is slow and is not recommended for large transcript data.frame.</p></dd>


<dt id="arg-cellular-distance-cutoff">cellular_distance_cutoff<a class="anchor" aria-label="anchor" href="#arg-cellular-distance-cutoff"></a></dt>
<dd><p>maximum cell-to-cell distance in x, y between the center of query cells to the center of neighbor cells with direct contact, same unit as input spatial coordinate. Default = NULL to use the 2 times of average 2D cell diameter.</p></dd>


<dt id="arg-score-baseline">score_baseline<a class="anchor" aria-label="anchor" href="#arg-score-baseline"></a></dt>
<dd><p>a named vector of score baseline under each cell type listed in <code>score_GeneMatrix</code> such that per cell transcript score higher than the baseline is required to call a cell type of high enough confidence</p></dd>


<dt id="arg-lowercutoff-transnum">lowerCutoff_transNum<a class="anchor" aria-label="anchor" href="#arg-lowercutoff-transnum"></a></dt>
<dd><p>a named vector of transcript number cutoff under each cell type such that higher than the cutoff is required to keep query cell as it is</p></dd>


<dt id="arg-highercutoff-transnum">higherCutoff_transNum<a class="anchor" aria-label="anchor" href="#arg-highercutoff-transnum"></a></dt>
<dd><p>a named vector of transcript number cutoff under each cell type such that lower than the cutoff is required to keep query cell as it is when there is neighbor cell of consistent cell type.</p></dd>


<dt id="arg-grouptranscripts-method">groupTranscripts_method<a class="anchor" aria-label="anchor" href="#arg-grouptranscripts-method"></a></dt>
<dd><p>use either "dbscan" or "delaunay method" to group transcripts in space (default = "dbscan")</p></dd>


<dt id="arg-spatialmergecheck-method">spatialMergeCheck_method<a class="anchor" aria-label="anchor" href="#arg-spatialmergecheck-method"></a></dt>
<dd><p>use either "leidenCut" (in 2D or 3D) or "geometryDiff" (in 2D only) method to determine whether a cell pair merging event is allowed in space (default = "leidenCut")</p></dd>


<dt id="arg-cutoff-spatialmerge">cutoff_spatialMerge<a class="anchor" aria-label="anchor" href="#arg-cutoff-spatialmerge"></a></dt>
<dd><p>spatial constraint on a valid merging event between two source transcript groups, default = 0.5 for 50% cutoff, set to 0 to skip spatial constraint evaluation for merging.
For <code>spatialMergeCheck_method = "leidenCut"</code>, this is the minimal percentage of transcripts shared membership between query cell and neighbor cells in leiden clustering results for a valid merging event.
For <code>spatialMergeCheck_method = "geometryDiff"</code>, this is the maximum percentage of white space change upon merging of query cell and neighbor cell for a valid merging event.</p></dd>


<dt id="arg-leiden-config">leiden_config<a class="anchor" aria-label="anchor" href="#arg-leiden-config"></a></dt>
<dd><p>(leidenCut) a list of configuration to pass to reticulate and <code><a href="https://r.igraph.org/reference/cluster_leiden.html" class="external-link">igraph::cluster_leiden</a></code> function, including objective_function, resolution_parameter, beta, n_iterations.</p></dd>


<dt id="arg-config-spatnw-transcript">config_spatNW_transcript<a class="anchor" aria-label="anchor" href="#arg-config-spatnw-transcript"></a></dt>
<dd><p>configuration list to create spatial network at transcript level, see manual for <code>createSpatialDelaunayNW_from_spatLocs</code> for more details, set to NULL to use default config (default = NULL)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a list</p><dl><dt>transcript_df</dt>
<dd><p>transcript data.frame ready for downstream full resgmentation pipeline</p></dd>

<dt>cellular_distance_cutoff</dt>
<dd><p>maximum cell-to-cell distance in x, y between the center of query cells to the center of neighbor cells with direct contact, same unit as input spatial coordinate.</p></dd>

<dt>molecular_distance_cutoff</dt>
<dd><p>maximum molecule-to-molecule distance within connected transcript group, same unit as input spatial coordinate</p></dd>

<dt>config_spatNW_transcript</dt>
<dd><p>configuration list to create spatial network at transcript level</p></dd>

<dt>leiden_config</dt>
<dd><p>configuration list to do leiden clustering on spatial network at transcript level for merge event evaluation</p></dd>


</dl></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Patrick Danaher, Lidan Wu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

