<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>fastReseg_perFOV_full_process — fastReseg_perFOV_full_process • FastReseg</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="fastReseg_perFOV_full_process — fastReseg_perFOV_full_process"><meta name="description" content="core wrapper for resegmentation pipeline using transcript score matrix derived from external reference profiles and preset cutoffs"><meta property="og:description" content="core wrapper for resegmentation pipeline using transcript score matrix derived from external reference profiles and preset cutoffs"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-#E6F3FC" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FastReseg</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="external-link dropdown-item" href="https://www.nature.com/articles/s41598-025-08733-5">Working principles and Manuscript</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Workflow and core functions</h6></li>
    <li><a class="dropdown-item" href="../articles/tutorial.html">FastReseg to detect and correct segmentation error in spatial transcriptome data</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Nanostring-Biostats/FastReseg/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>fastReseg_perFOV_full_process</h1>
      <small class="dont-index">Source: <a href="https://github.com/Nanostring-Biostats/FastReseg/blob/HEAD/R/fastReseg_perFOV_full_process.R" class="external-link"><code>R/fastReseg_perFOV_full_process.R</code></a></small>
      <div class="d-none name"><code>fastReseg_perFOV_full_process.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>core wrapper for resegmentation pipeline using transcript score matrix derived from external reference profiles and preset cutoffs</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">fastReseg_perFOV_full_process</span><span class="op">(</span></span>
<span>  <span class="va">score_GeneMatrix</span>,</span>
<span>  <span class="va">transcript_df</span>,</span>
<span>  transID_coln <span class="op">=</span> <span class="st">"UMI_transID"</span>,</span>
<span>  transGene_coln <span class="op">=</span> <span class="st">"target"</span>,</span>
<span>  cellID_coln <span class="op">=</span> <span class="st">"UMI_cellID"</span>,</span>
<span>  spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span>, <span class="st">"z"</span><span class="op">)</span>,</span>
<span>  invert_y <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  extracellular_cellID <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  flagModel_TransNum_cutoff <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  flagCell_lrtest_cutoff <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  svmClass_score_cutoff <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>,</span>
<span>  svm_args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>kernel <span class="op">=</span> <span class="st">"radial"</span>, scale <span class="op">=</span> <span class="cn">FALSE</span>, gamma <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span>,</span>
<span>  molecular_distance_cutoff <span class="op">=</span> <span class="fl">2.7</span>,</span>
<span>  cellular_distance_cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  score_baseline <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  lowerCutoff_transNum <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  higherCutoff_transNum <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  groupTranscripts_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dbscan"</span>, <span class="st">"delaunay"</span><span class="op">)</span>,</span>
<span>  spatialMergeCheck_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"leidenCut"</span>, <span class="st">"geometryDiff"</span><span class="op">)</span>,</span>
<span>  cutoff_spatialMerge <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  leiden_config <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>objective_function <span class="op">=</span> <span class="st">"CPM"</span>, resolution_parameter <span class="op">=</span> <span class="fl">1</span>, beta <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>    n_iterations <span class="op">=</span> <span class="fl">200</span><span class="op">)</span>,</span>
<span>  config_spatNW_transcript <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  return_intermediates <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  return_perCellData <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  includeAllRefGenes <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  seed_process <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-score-genematrix">score_GeneMatrix<a class="anchor" aria-label="anchor" href="#arg-score-genematrix"></a></dt>
<dd><p>the gene x cell-type matrix of log-like score of gene in each cell type</p></dd>


<dt id="arg-transcript-df">transcript_df<a class="anchor" aria-label="anchor" href="#arg-transcript-df"></a></dt>
<dd><p>the data.frame for each transcript with columns for transcript_id, target or gene name, original cell_id, spatial coordinates.</p></dd>


<dt id="arg-transid-coln">transID_coln<a class="anchor" aria-label="anchor" href="#arg-transid-coln"></a></dt>
<dd><p>the column name of transcript_ID in <code>transcript_df</code></p></dd>


<dt id="arg-transgene-coln">transGene_coln<a class="anchor" aria-label="anchor" href="#arg-transgene-coln"></a></dt>
<dd><p>the column name of target or gene name in <code>transcript_df</code></p></dd>


<dt id="arg-cellid-coln">cellID_coln<a class="anchor" aria-label="anchor" href="#arg-cellid-coln"></a></dt>
<dd><p>the column name of cell_ID in <code>transcript_df</code></p></dd>


<dt id="arg-spatlocs-colns">spatLocs_colns<a class="anchor" aria-label="anchor" href="#arg-spatlocs-colns"></a></dt>
<dd><p>column names for 1st, 2nd and optional 3rd dimension of spatial coordinates in <code>transcript_df</code></p></dd>


<dt id="arg-invert-y">invert_y<a class="anchor" aria-label="anchor" href="#arg-invert-y"></a></dt>
<dd><p>flag to invert y axis of local coordinates during stitching (default = TRUE)</p></dd>


<dt id="arg-extracellular-cellid">extracellular_cellID<a class="anchor" aria-label="anchor" href="#arg-extracellular-cellid"></a></dt>
<dd><p>a vector of cell_ID for extracellular transcripts which would be removed from the resegmention pipeline (default = NULL)</p></dd>


<dt id="arg-flagmodel-transnum-cutoff">flagModel_TransNum_cutoff<a class="anchor" aria-label="anchor" href="#arg-flagmodel-transnum-cutoff"></a></dt>
<dd><p>the cutoff of transcript number to do spatial modeling for identification of wrongly segmented cells (default = 50)</p></dd>


<dt id="arg-flagcell-lrtest-cutoff">flagCell_lrtest_cutoff<a class="anchor" aria-label="anchor" href="#arg-flagcell-lrtest-cutoff"></a></dt>
<dd><p>the cutoff of lrtest_nlog10P to identify putative wrongly segemented cells with strong spatial dependency in transcript score profile</p></dd>


<dt id="arg-svmclass-score-cutoff">svmClass_score_cutoff<a class="anchor" aria-label="anchor" href="#arg-svmclass-score-cutoff"></a></dt>
<dd><p>the cutoff of transcript score to separate between high and low score transcripts in SVM (default = -2)</p></dd>


<dt id="arg-svm-args">svm_args<a class="anchor" aria-label="anchor" href="#arg-svm-args"></a></dt>
<dd><p>a list of arguments to pass to svm function for identifying low-score transcript groups in space, typically involve kernel, gamma, scale</p></dd>


<dt id="arg-molecular-distance-cutoff">molecular_distance_cutoff<a class="anchor" aria-label="anchor" href="#arg-molecular-distance-cutoff"></a></dt>
<dd><p>maximum molecule-to-molecule distance within connected transcript group, same unit as input spatial coordinate (default = 2.7 micron).
If set to NULL, the pipeline would first randomly choose no more than 2500 cells from up to 10 random picked ROIs with search radius to be 5 times of <code>cellular_distance_cutoff</code>, and then calculate the minimal molecular distance between picked cells. The pipeline would further use the 5 times of 90% quantile of minimal molecular distance as <code>molecular_distance_cutoff</code>. This calculation is slow and is not recommended for large transcript data.frame.</p></dd>


<dt id="arg-cellular-distance-cutoff">cellular_distance_cutoff<a class="anchor" aria-label="anchor" href="#arg-cellular-distance-cutoff"></a></dt>
<dd><p>maximum cell-to-cell distance in x, y between the center of query cells to the center of neighbor cells with direct contact, same unit as input spatial coordinate. Default = NULL to use the 2 times of average 2D cell diameter.</p></dd>


<dt id="arg-score-baseline">score_baseline<a class="anchor" aria-label="anchor" href="#arg-score-baseline"></a></dt>
<dd><p>a named vector of score baseline under each cell type listed in <code>score_GeneMatrix</code> such that per cell transcript score higher than the baseline is required to call a cell type of high enough confidence</p></dd>


<dt id="arg-lowercutoff-transnum">lowerCutoff_transNum<a class="anchor" aria-label="anchor" href="#arg-lowercutoff-transnum"></a></dt>
<dd><p>a named vector of transcript number cutoff under each cell type such that higher than the cutoff is required to keep query cell as it is</p></dd>


<dt id="arg-highercutoff-transnum">higherCutoff_transNum<a class="anchor" aria-label="anchor" href="#arg-highercutoff-transnum"></a></dt>
<dd><p>a named vector of transcript number cutoff under each cell type such that lower than the cutoff is required to keep query cell as it is when there is neighbor cell of consistent cell type.</p></dd>


<dt id="arg-grouptranscripts-method">groupTranscripts_method<a class="anchor" aria-label="anchor" href="#arg-grouptranscripts-method"></a></dt>
<dd><p>use either "dbscan" or "delaunay method" to group transcripts in space (default = "dbscan")</p></dd>


<dt id="arg-spatialmergecheck-method">spatialMergeCheck_method<a class="anchor" aria-label="anchor" href="#arg-spatialmergecheck-method"></a></dt>
<dd><p>use either "leidenCut" (in 2D or 3D) or "geometryDiff" (in 2D only) method to determine whether a cell pair merging event is allowed in space (default = "leidenCut")</p></dd>


<dt id="arg-cutoff-spatialmerge">cutoff_spatialMerge<a class="anchor" aria-label="anchor" href="#arg-cutoff-spatialmerge"></a></dt>
<dd><p>spatial constraint on a valid merging event between two source transcript groups, default = 0.5 for 50% cutoff, set to 0 to skip spatial constraint evaluation for merging.
For <code>spatialMergeCheck_method = "leidenCut"</code>, this is the minimal percentage of transcripts shared membership between query cell and neighbor cells in leiden clustering results for a valid merging event.
For <code>spatialMergeCheck_method = "geometryDiff"</code>, this is the maximum percentage of white space change upon merging of query cell and neighbor cell for a valid merging event.</p></dd>


<dt id="arg-leiden-config">leiden_config<a class="anchor" aria-label="anchor" href="#arg-leiden-config"></a></dt>
<dd><p>(leidenCut) a list of configuration to pass to reticulate and <code><a href="https://r.igraph.org/reference/cluster_leiden.html" class="external-link">igraph::cluster_leiden</a></code> function, including objective_function, resolution_parameter, beta, n_iterations.</p></dd>


<dt id="arg-config-spatnw-transcript">config_spatNW_transcript<a class="anchor" aria-label="anchor" href="#arg-config-spatnw-transcript"></a></dt>
<dd><p>configuration list to create spatial network at transcript level, see manual for <code>createSpatialDelaunayNW_from_spatLocs</code> for more details, set to NULL to use default config (default = NULL)</p></dd>


<dt id="arg-return-intermediates">return_intermediates<a class="anchor" aria-label="anchor" href="#arg-return-intermediates"></a></dt>
<dd><p>flag to return intermediate outputs, including data.frame for spatial modeling statistics of each cell</p></dd>


<dt id="arg-return-percelldata">return_perCellData<a class="anchor" aria-label="anchor" href="#arg-return-percelldata"></a></dt>
<dd><p>flag to return gene x cell count matrix and per cell DF with updated mean spatial coordinates and new cell type</p></dd>


<dt id="arg-includeallrefgenes">includeAllRefGenes<a class="anchor" aria-label="anchor" href="#arg-includeallrefgenes"></a></dt>
<dd><p>flag to include all genes in <code>score_GeneMatrix</code> in the returned <code>updated_perCellExprs</code> with missing genes of value 0 (default = FALSE)</p></dd>


<dt id="arg-seed-process">seed_process<a class="anchor" aria-label="anchor" href="#arg-seed-process"></a></dt>
<dd><p>seed for per FOV processing, used in transcript error detection and correction steps, default = NULL to skip the seed</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a list</p><dl><dt>modStats_ToFlagCells</dt>
<dd><p>a data.frame for spatial modeling statistics of each cell, output of <code>score_cell_segmentation_error</code> function, return when <code>return_intermediates</code> = TRUE</p></dd>

<dt>groupDF_ToFlagTrans</dt>
<dd><p>data.frame for the group assignment of transcripts within putative wrongly segmented cells, merged output of <code>flag_bad_transcripts</code> and <code>groupTranscripts_Delaunay</code> or <code>groupTranscripts_dbscan</code> functions, return when <code>return_intermediates</code> = TRUE</p></dd>

<dt>neighborhoodDF_ToReseg</dt>
<dd><p>a data.frame for neighborhood environment of low-score transcript groups, output of <code>get_neighborhood_content</code> function, return when <code>return_intermediates</code> = TRUE</p></dd>

<dt>reseg_actions</dt>
<dd><p>a list of 4 elements describing how the resegmenation would be performed on original <code>transcript_df</code> by the group assignment of transcripts listed in <code>groupDF_ToFlagTrans</code>, output of <code>decide_ReSegment_Operations</code> function, return when <code>return_intermediates</code> = TRUE</p></dd>

<dt>updated_transDF</dt>
<dd><p>the updated transcript_df with <code>updated_cellID</code> and <code>updated_celltype</code> column based on reseg_full_converter</p></dd>

<dt>updated_perCellDT</dt>
<dd><p>a per cell data.table with mean spatial coordinates, new cell type and resegmentation action after resegmentation, return when <code>return_perCellData</code> = TRUE</p></dd>

<dt>updated_perCellExprs</dt>
<dd><p>a gene x cell count sparse matrix for updated transcript data.frame after resegmentation, return when <code>return_perCellData</code> = TRUE</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The pipeline would score each transcript based on the provided cell type-specific reference profiles, evaluate the goodness-of-fit of each transcript within original cell segment, identify the low-score transcript groups within cells that has strong spatial dependency in transcript score profile, evaluate the neighborhood environment of low-score transcript groups and perform resegmentation actions including triming to extracellular space, merging to neighbor cell or labeling as new cell.</p>
<p>To account for genes missing in <code>score_GeneMatrix</code> but present in input transcript data.frame, genes in <code>ctrl_genes</code> would be assigned with goodness-of-fit score equal to <code>svmClass_score_cutoff</code> for all cell types to minimize the impact of those genes on the identification of low-score transcript groups via SVM. To avoid significant interference from those <code>ctrl_genes</code>, it's recommended to have total counts of those genes below 1% of total counts of all genes in each cell.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">example_refProfiles</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">mini_transcriptDF</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">example_baselineCT</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co">#' cell_ID for extracellualr transcripts</span></span></span>
<span class="r-in"><span><span class="va">extracellular_cellID</span> <span class="op">&lt;-</span> <span class="va">mini_transcriptDF</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">mini_transcriptDF</span><span class="op">$</span><span class="va">CellId</span> <span class="op">==</span><span class="fl">0</span><span class="op">)</span>, <span class="st">'cell_ID'</span><span class="op">]</span> </span></span>
<span class="r-in"><span><span class="va">score_baseline</span> <span class="op">&lt;-</span> <span class="va">example_baselineCT</span><span class="op">[[</span><span class="st">'span_score'</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="st">"25%"</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">lowerCutoff_transNum</span>  <span class="op">&lt;-</span> <span class="va">example_baselineCT</span><span class="op">[[</span><span class="st">'span_transNum'</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="st">"25%"</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">higherCutoff_transNum</span>  <span class="op">&lt;-</span> <span class="va">example_baselineCT</span><span class="op">[[</span><span class="st">'span_transNum'</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="st">"50%"</span><span class="op">]</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># calculate log-likelihood of each gene under each cell type and center the </span></span></span>
<span class="r-in"><span><span class="co"># score matrix on per gene basis</span></span></span>
<span class="r-in"><span><span class="va">score_GeneMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="scoreGenesInRef.html">scoreGenesInRef</a></span><span class="op">(</span>genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">mini_transcriptDF</span><span class="op">[[</span><span class="st">"target"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>                                                      <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">example_refProfiles</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                                    ref_profiles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmax</a></span><span class="op">(</span><span class="va">example_refProfiles</span>, <span class="fl">1e-5</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>                                    </span></span>
<span class="r-in"><span><span class="co"># case 1: run with default methods: "dbscan" for transcript grouping, </span></span></span>
<span class="r-in"><span><span class="co"># "leidenCut" for merging check </span></span></span>
<span class="r-in"><span><span class="va">final_res1</span> <span class="op">&lt;-</span> <span class="fu">fastReseg_perFOV_full_process</span><span class="op">(</span>score_GeneMatrix<span class="op">=</span> <span class="va">score_GeneMatrix</span>,</span></span>
<span class="r-in"><span>                                           transcript_df <span class="op">=</span> <span class="va">mini_transcriptDF</span>,</span></span>
<span class="r-in"><span>                                           extracellular_cellID <span class="op">=</span> <span class="va">extracellular_cellID</span>,</span></span>
<span class="r-in"><span>                                           molecular_distance_cutoff <span class="op">=</span> <span class="fl">2.7</span>,</span></span>
<span class="r-in"><span>                                           cellular_distance_cutoff <span class="op">=</span> <span class="fl">25</span>,</span></span>
<span class="r-in"><span>                                           score_baseline <span class="op">=</span> <span class="va">score_baseline</span>,</span></span>
<span class="r-in"><span>                                           lowerCutoff_transNum <span class="op">=</span> <span class="va">lowerCutoff_transNum</span>,</span></span>
<span class="r-in"><span>                                           higherCutoff_transNum<span class="op">=</span> <span class="va">higherCutoff_transNum</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3 Dimension of spaital coordinates are provided.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Perform leiden clustering at resolution_parameter = 1.000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Create Delanay network when config$method is NULL.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Name the spatial network based on method as `Delaunay_network` when config$name is NULL.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use cellular_distance_cutoff = 25.0000 for searching of neighbor cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use `molecular_distance_cutoff` = 2.7000 for defining direct neighbor cells based on molecule-to-molecule distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 960 unique genes are found in `transcript_df`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 960 unique genes are shared by the provided `score_GeneMatrix` that contains cluster profiles of 960 genes for 22 clusters.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 756783 transcript records and 1375 cells in input `transcript_df`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use the providied `molecular_distance_cutoff` = 2.7000 for defining direct neighbor cells based on molecule-to-molecule distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use the providied `cellular_distance_cutoff` = 25.0000 for searching of neighbor cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 960 common genes among transcript_df and score_GeneMatrix. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 1375 cells and assigned cell type based on the provided `refProfiles` cluster profiles.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Run linear regreassion in 3 Dimension.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Below model_cutoff = 50, skip 37 cells with fewer transcripts. Move forward with remaining 1338 cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 373 cells, 0.2788 of all evaluated cells, are flagged for resegmentation with lrtest_nlog10P &gt; 5.0.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Run SVM in 3 Dimension.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 373 common cells and 960 common genes among chosen_cells, transcript_df, and score_GeneMatrix. </span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Below model_cutoff = 50, skip 0 cells with fewer transcripts. Move forward with remaining 373 cells.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Skip 0 cells with all transcripts in same class given `score_cutoff = -2`. Move forward with remaining 373 cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Remove 0 cells with raw transcript score all in same class based on cutoff -2.00 when running spatial SVM model.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Do spatial network analysis in 3 Dimension.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 10652 chosen_transcripts are found in common cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> SVM spatial model further identified 17 cells with transcript score all in same class, exclude from transcript group analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 960 common genes among transcript_df and score_GeneMatrix. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Perform leiden clustering at resolution_parameter = 1.000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use neighbor_distance_xy = 25.0000 for searching of neighbor cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use distance_cutoff = 2.7000 for defining direct neighbor cells based on molecule-to-molecule distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use first 2D for searching cell neighborhood, but all 3 Dimension to identify direct neighbors based on molecular distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 2289 common cells and 960 common genes among transcript_df, cell_networkDT, and score_GeneMatrix. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 914 chosen_cells are found in common cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use `leidenCut` method to evaluate putative merging event in space. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Perform leiden clustering at resolution_parameter = 1.000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> A valid merging event must have query cell with 0.500 transcript shared same membership as neighbor cell of consistent cell type. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Run delanuay network in 3 Dimension.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Perform ledien clustering on 59 potential merging events. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> (`c_1_2_1339_g3`, `c_1_2_1403_g5`) cell pair with all 7 transcripts in same z plane, run 2D network analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> (`c_1_2_1403_g5`, `c_1_2_1339_g3`) cell pair with all 7 transcripts in same z plane, run 2D network analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> (`c_1_2_1731_g2`, `c_1_2_1792_g6`) cell pair with all 5 transcripts in same z plane, run 2D network analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> (`c_1_2_1792_g6`, `c_1_2_1731_g2`) cell pair with all 5 transcripts in same z plane, run 2D network analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> (`c_1_2_921_g5`, `c_1_2_941_g1`) cell pair with all 15 transcripts in same z plane, run 2D network analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> (`c_1_2_941_g1`, `c_1_2_921_g5`) cell pair with all 15 transcripts in same z plane, run 2D network analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 915 common cells and 960 common genes among `names(reseg_full_converter)`, `transcript_df`, and `score_GeneMatrix`. </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># case 2: run with alternative methods: "delaunay" for transcript grouping, </span></span></span>
<span class="r-in"><span><span class="co"># "geometryDiff" for merging check </span></span></span>
<span class="r-in"><span><span class="va">final_res2</span> <span class="op">&lt;-</span> <span class="fu">fastReseg_perFOV_full_process</span><span class="op">(</span>score_GeneMatrix<span class="op">=</span> <span class="va">score_GeneMatrix</span>,</span></span>
<span class="r-in"><span>                                           transcript_df <span class="op">=</span> <span class="va">mini_transcriptDF</span>,</span></span>
<span class="r-in"><span>                                           extracellular_cellID <span class="op">=</span> <span class="va">extracellular_cellID</span>,</span></span>
<span class="r-in"><span>                                           molecular_distance_cutoff <span class="op">=</span> <span class="fl">2.7</span>,</span></span>
<span class="r-in"><span>                                           cellular_distance_cutoff <span class="op">=</span> <span class="fl">25</span>,</span></span>
<span class="r-in"><span>                                           score_baseline <span class="op">=</span> <span class="va">score_baseline</span>,</span></span>
<span class="r-in"><span>                                           lowerCutoff_transNum <span class="op">=</span> <span class="va">lowerCutoff_transNum</span>,</span></span>
<span class="r-in"><span>                                           higherCutoff_transNum<span class="op">=</span> <span class="va">higherCutoff_transNum</span>,</span></span>
<span class="r-in"><span>                                           groupTranscripts_method <span class="op">=</span> <span class="st">"delaunay"</span>,</span></span>
<span class="r-in"><span>                                           spatialMergeCheck_method <span class="op">=</span> <span class="st">"geometryDiff"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3 Dimension of spaital coordinates are provided.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Create Delanay network when config$method is NULL.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Name the spatial network based on method as `Delaunay_network` when config$name is NULL.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use cellular_distance_cutoff = 25.0000 for searching of neighbor cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use `molecular_distance_cutoff` = 2.7000 for defining direct neighbor cells based on molecule-to-molecule distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 960 unique genes are found in `transcript_df`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 960 unique genes are shared by the provided `score_GeneMatrix` that contains cluster profiles of 960 genes for 22 clusters.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 756783 transcript records and 1375 cells in input `transcript_df`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use the providied `molecular_distance_cutoff` = 2.7000 for defining direct neighbor cells based on molecule-to-molecule distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use the providied `cellular_distance_cutoff` = 25.0000 for searching of neighbor cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 960 common genes among transcript_df and score_GeneMatrix. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 1375 cells and assigned cell type based on the provided `refProfiles` cluster profiles.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Run linear regreassion in 3 Dimension.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Below model_cutoff = 50, skip 37 cells with fewer transcripts. Move forward with remaining 1338 cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 373 cells, 0.2788 of all evaluated cells, are flagged for resegmentation with lrtest_nlog10P &gt; 5.0.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Run SVM in 3 Dimension.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 373 common cells and 960 common genes among chosen_cells, transcript_df, and score_GeneMatrix. </span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Below model_cutoff = 50, skip 0 cells with fewer transcripts. Move forward with remaining 373 cells.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Skip 0 cells with all transcripts in same class given `score_cutoff = -2`. Move forward with remaining 373 cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Remove 0 cells with raw transcript score all in same class based on cutoff -2.00 when running spatial SVM model.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Do spatial network analysis in 3 Dimension.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 10652 chosen_transcripts are found in common cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_1172 has 1 solo transcripts, flag based on distance cutoff = 2.7000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_1184 with all 4 flagged transcripts of unique transcripts in same z plane, run 2D network analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_1211 with all 7 flagged transcripts of unique transcripts in same z plane, run 2D network analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_1403 has 1 solo transcripts, flag based on distance cutoff = 2.7000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_1412 with all 7 flagged transcripts of unique transcripts in same z plane, run 2D network analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_1436 has 1 solo transcripts, flag based on distance cutoff = 2.7000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_1585 has 1 solo transcripts, flag based on distance cutoff = 2.7000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_1620 has 1 solo transcripts, flag based on distance cutoff = 2.7000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_1625 has 1 solo transcripts, flag based on distance cutoff = 2.7000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_1674 with all 5 flagged transcripts of unique transcripts in same z plane, run 2D network analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_1689 has 1 solo transcripts, flag based on distance cutoff = 2.7000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_1732 has 1 solo transcripts, flag based on distance cutoff = 2.7000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_1736 with all 12 flagged transcripts of unique transcripts in same z plane, run 2D network analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_1736 has 1 solo transcripts, flag based on distance cutoff = 2.7000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_1801 has 1 solo transcripts, flag based on distance cutoff = 2.7000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_1830 has 1 solo transcripts, flag based on distance cutoff = 2.7000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_1887 has 2 solo transcripts, flag based on distance cutoff = 2.7000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_1930 with all 11 flagged transcripts of unique transcripts in same z plane, run 2D network analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_1968 with all 5 flagged transcripts of unique transcripts in same z plane, run 2D network analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_2157 has 1 solo transcripts, flag based on distance cutoff = 2.7000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_2279 has 1 solo transcripts, flag based on distance cutoff = 2.7000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_2466 has 1 solo transcripts, flag based on distance cutoff = 2.7000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_2503 has 1 solo transcripts, flag based on distance cutoff = 2.7000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_2549 has 1 solo transcripts, flag based on distance cutoff = 2.7000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_2566 with all 6 flagged transcripts of unique transcripts in same z plane, run 2D network analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_2590 with all 12 flagged transcripts of unique transcripts in same z plane, run 2D network analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_2786 with all 5 flagged transcripts of unique transcripts in same z plane, run 2D network analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_417 has 1 solo transcripts, flag based on distance cutoff = 2.7000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_4201 has 1 solo transcripts, flag based on distance cutoff = 2.7000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_4364 with all 13 flagged transcripts of unique transcripts in same z plane, run 2D network analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_4421 with all 4 flagged transcripts of unique transcripts in same z plane, run 2D network analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_477 has 2 solo transcripts, flag based on distance cutoff = 2.7000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_562 has 1 solo transcripts, flag based on distance cutoff = 2.7000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_712 has 1 solo transcripts, flag based on distance cutoff = 2.7000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> c_1_2_941 with all 9 flagged transcripts of unique transcripts in same z plane, run 2D network analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> SVM spatial model further identified 17 cells with transcript score all in same class, exclude from transcript group analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 960 common genes among transcript_df and score_GeneMatrix. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use neighbor_distance_xy = 25.0000 for searching of neighbor cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use distance_cutoff = 2.7000 for defining direct neighbor cells based on molecule-to-molecule distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use first 2D for searching cell neighborhood, but all 3 Dimension to identify direct neighbors based on molecular distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 2292 common cells and 960 common genes among transcript_df, cell_networkDT, and score_GeneMatrix. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 917 chosen_cells are found in common cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use `geometryDiff` method to evaluate putative merging event in space. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> A valid merging event to neighbor cell of consistent cell type must have no more than 0.500 area change in white space upon merging with respect to the concave area of either source cells. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Perform geometry analysis in 2D for potential merging events despite the provided 3 Dimension data.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Perform geometry analysis on 127 potential merging events. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 922 common cells and 960 common genes among `names(reseg_full_converter)`, `transcript_df`, and `score_GeneMatrix`. </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># case 3: run with default "dbscan" for transcript grouping, but apply no </span></span></span>
<span class="r-in"><span><span class="co"># spatial constraint on merging</span></span></span>
<span class="r-in"><span><span class="va">final_res3</span> <span class="op">&lt;-</span> <span class="fu">fastReseg_perFOV_full_process</span><span class="op">(</span>score_GeneMatrix<span class="op">=</span> <span class="va">score_GeneMatrix</span>,</span></span>
<span class="r-in"><span>                                            transcript_df <span class="op">=</span> <span class="va">mini_transcriptDF</span>,</span></span>
<span class="r-in"><span>                                            extracellular_cellID <span class="op">=</span> <span class="va">extracellular_cellID</span>,</span></span>
<span class="r-in"><span>                                            molecular_distance_cutoff <span class="op">=</span> <span class="fl">2.7</span>,</span></span>
<span class="r-in"><span>                                            cellular_distance_cutoff <span class="op">=</span> <span class="fl">25</span>,</span></span>
<span class="r-in"><span>                                            score_baseline <span class="op">=</span> <span class="va">score_baseline</span>,</span></span>
<span class="r-in"><span>                                            lowerCutoff_transNum <span class="op">=</span> <span class="va">lowerCutoff_transNum</span>,</span></span>
<span class="r-in"><span>                                            higherCutoff_transNum<span class="op">=</span> <span class="va">higherCutoff_transNum</span>,</span></span>
<span class="r-in"><span>                                            cutoff_spatialMerge <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3 Dimension of spaital coordinates are provided.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use cellular_distance_cutoff = 25.0000 for searching of neighbor cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use `molecular_distance_cutoff` = 2.7000 for defining direct neighbor cells based on molecule-to-molecule distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 960 unique genes are found in `transcript_df`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 960 unique genes are shared by the provided `score_GeneMatrix` that contains cluster profiles of 960 genes for 22 clusters.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 756783 transcript records and 1375 cells in input `transcript_df`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use the providied `molecular_distance_cutoff` = 2.7000 for defining direct neighbor cells based on molecule-to-molecule distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use the providied `cellular_distance_cutoff` = 25.0000 for searching of neighbor cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 960 common genes among transcript_df and score_GeneMatrix. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 1375 cells and assigned cell type based on the provided `refProfiles` cluster profiles.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Run linear regreassion in 3 Dimension.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Below model_cutoff = 50, skip 37 cells with fewer transcripts. Move forward with remaining 1338 cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 373 cells, 0.2788 of all evaluated cells, are flagged for resegmentation with lrtest_nlog10P &gt; 5.0.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Run SVM in 3 Dimension.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 373 common cells and 960 common genes among chosen_cells, transcript_df, and score_GeneMatrix. </span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Below model_cutoff = 50, skip 0 cells with fewer transcripts. Move forward with remaining 373 cells.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Skip 0 cells with all transcripts in same class given `score_cutoff = -2`. Move forward with remaining 373 cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Remove 0 cells with raw transcript score all in same class based on cutoff -2.00 when running spatial SVM model.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Do spatial network analysis in 3 Dimension.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 10652 chosen_transcripts are found in common cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> SVM spatial model further identified 17 cells with transcript score all in same class, exclude from transcript group analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 960 common genes among transcript_df and score_GeneMatrix. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use neighbor_distance_xy = 25.0000 for searching of neighbor cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use distance_cutoff = 2.7000 for defining direct neighbor cells based on molecule-to-molecule distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use first 2D for searching cell neighborhood, but all 3 Dimension to identify direct neighbors based on molecular distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 2289 common cells and 960 common genes among transcript_df, cell_networkDT, and score_GeneMatrix. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 914 chosen_cells are found in common cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The provided `cutoff_spatialMerge = 0`, no spatial evaluation would be done for potential cell merging events. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 928 common cells and 960 common genes among `names(reseg_full_converter)`, `transcript_df`, and `score_GeneMatrix`. </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Patrick Danaher, Lidan Wu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

