<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>fastReseg_full_pipeline — fastReseg_full_pipeline • FastReseg</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="fastReseg_full_pipeline — fastReseg_full_pipeline"><meta name="description" content="wrapper for full resegmentation pipeline using internal reference profiles and cutoffs. This function first estimates proper reference profiles and cutoffs from the provided data and then use fastReseg_perFOV_full_process function to process each transcript data.frame."><meta property="og:description" content="wrapper for full resegmentation pipeline using internal reference profiles and cutoffs. This function first estimates proper reference profiles and cutoffs from the provided data and then use fastReseg_perFOV_full_process function to process each transcript data.frame."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-#E6F3FC" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FastReseg</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="external-link dropdown-item" href="https://www.nature.com/articles/s41598-025-08733-5">Working principles and Manuscript</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Workflow and core functions</h6></li>
    <li><a class="dropdown-item" href="../articles/tutorial.html">FastReseg to detect and correct segmentation error in spatial transcriptome data</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Nanostring-Biostats/FastReseg/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>fastReseg_full_pipeline</h1>
      <small class="dont-index">Source: <a href="https://github.com/Nanostring-Biostats/FastReseg/blob/HEAD/R/fastReseg_full_pipeline.R" class="external-link"><code>R/fastReseg_full_pipeline.R</code></a></small>
      <div class="d-none name"><code>fastReseg_full_pipeline.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>wrapper for full resegmentation pipeline using internal reference profiles and cutoffs. This function first estimates proper reference profiles and cutoffs from the provided data and then use <code>fastReseg_perFOV_full_process</code> function to process each transcript data.frame.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">fastReseg_full_pipeline</span><span class="op">(</span></span>
<span>  <span class="va">counts</span>,</span>
<span>  clust <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  refProfiles <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transDF_fileInfo <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filepath_coln <span class="op">=</span> <span class="st">"file_path"</span>,</span>
<span>  prefix_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"slide"</span>, <span class="st">"fov"</span><span class="op">)</span>,</span>
<span>  fovOffset_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"stage_X"</span>, <span class="st">"stage_Y"</span><span class="op">)</span>,</span>
<span>  pixel_size <span class="op">=</span> <span class="fl">0.18</span>,</span>
<span>  zstep_size <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  transcript_df <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transID_coln <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transGene_coln <span class="op">=</span> <span class="st">"target"</span>,</span>
<span>  cellID_coln <span class="op">=</span> <span class="st">"CellId"</span>,</span>
<span>  spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span>, <span class="st">"z"</span><span class="op">)</span>,</span>
<span>  invert_y <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  extracellular_cellID <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  flagModel_TransNum_cutoff <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  flagCell_lrtest_cutoff <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  svmClass_score_cutoff <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>,</span>
<span>  svm_args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>kernel <span class="op">=</span> <span class="st">"radial"</span>, scale <span class="op">=</span> <span class="cn">FALSE</span>, gamma <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span>,</span>
<span>  molecular_distance_cutoff <span class="op">=</span> <span class="fl">2.7</span>,</span>
<span>  cellular_distance_cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  score_baseline <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  lowerCutoff_transNum <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  higherCutoff_transNum <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  imputeFlag_missingCTs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  groupTranscripts_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dbscan"</span>, <span class="st">"delaunay"</span><span class="op">)</span>,</span>
<span>  spatialMergeCheck_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"leidenCut"</span>, <span class="st">"geometryDiff"</span><span class="op">)</span>,</span>
<span>  cutoff_spatialMerge <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  leiden_config <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>objective_function <span class="op">=</span> <span class="st">"CPM"</span>, resolution_parameter <span class="op">=</span> <span class="fl">1</span>, beta <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>    n_iterations <span class="op">=</span> <span class="fl">200</span><span class="op">)</span>,</span>
<span>  config_spatNW_transcript <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  path_to_output <span class="op">=</span> <span class="st">"reSeg_res"</span>,</span>
<span>  transDF_export_option <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>  save_intermediates <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  return_perCellData <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  combine_extra <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ctrl_genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seed_process <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  percentCores <span class="op">=</span> <span class="fl">0.75</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-counts">counts<a class="anchor" aria-label="anchor" href="#arg-counts"></a></dt>
<dd><p>Counts matrix for entire dataset, cells X genes.</p></dd>


<dt id="arg-clust">clust<a class="anchor" aria-label="anchor" href="#arg-clust"></a></dt>
<dd><p>Vector of cluster assignments for each cell in <code>counts</code>, when NULL to automatically assign the cell cluster for each cell based on maximum transcript score of given the provided <code>refProfiles</code></p></dd>


<dt id="arg-refprofiles">refProfiles<a class="anchor" aria-label="anchor" href="#arg-refprofiles"></a></dt>
<dd><p>A matrix of cluster profiles, genes X clusters, default = NULL to use external cluster assignments</p></dd>


<dt id="arg-transdf-fileinfo">transDF_fileInfo<a class="anchor" aria-label="anchor" href="#arg-transdf-fileinfo"></a></dt>
<dd><p>a data.frame with each row for each individual file of per FOV transcript data.frame within which the coordinates and CellId are unique, columns include the file path of per FOV transcript data.frame file, annotation columns like slide and fov to be used as prefix when creating unique cell_ID across entire dataset; when NULL, use the provided <code>transcript_df</code> directly</p></dd>


<dt id="arg-filepath-coln">filepath_coln<a class="anchor" aria-label="anchor" href="#arg-filepath-coln"></a></dt>
<dd><p>the column name of each individual file of per FOV transcript data.frame in <code>transDF_fileInfo</code></p></dd>


<dt id="arg-prefix-colns">prefix_colns<a class="anchor" aria-label="anchor" href="#arg-prefix-colns"></a></dt>
<dd><p>the column names of annotation in <code>transDF_fileInfo</code>, to be added to the CellId as prefix when creating unique cell_ID for entire dataset; set to NULL if use the original <code>transID_coln</code> or <code>cellID_coln</code></p></dd>


<dt id="arg-fovoffset-colns">fovOffset_colns<a class="anchor" aria-label="anchor" href="#arg-fovoffset-colns"></a></dt>
<dd><p>the column name of coordinate offsets in 1st and 2nd dimension for each per FOV transcript data.frame in <code>transDF_fileInfo</code>, unit in micron
Notice that some assays like SMI has XY axes swapped between stage and each FOV such that <code>fovOffset_colns</code> should be c("stage_Y", "stage_X").</p></dd>


<dt id="arg-pixel-size">pixel_size<a class="anchor" aria-label="anchor" href="#arg-pixel-size"></a></dt>
<dd><p>the micrometer size of image pixel listed in 1st and 2nd dimension of <code>spatLocs_colns</code> of each <code>transcript_df</code></p></dd>


<dt id="arg-zstep-size">zstep_size<a class="anchor" aria-label="anchor" href="#arg-zstep-size"></a></dt>
<dd><p>the micrometer size of z-step for the optional 3rd dimension of <code>spatLocs_colns</code> of each <code>transcript_df</code></p></dd>


<dt id="arg-transcript-df">transcript_df<a class="anchor" aria-label="anchor" href="#arg-transcript-df"></a></dt>
<dd><p>the data.frame of transcript level information with unique CellId, default = NULL to read from the <code>transDF_fileInfo</code></p></dd>


<dt id="arg-transid-coln">transID_coln<a class="anchor" aria-label="anchor" href="#arg-transid-coln"></a></dt>
<dd><p>the column name of transcript_ID in <code>transcript_df</code>, default = NULL to use row index of transcript in each <code>transcript_df</code>; when <code>prefix_colns</code> != NULL, unique transcript_id would be generated from <code>prefix_colns</code> and <code>transID_coln</code> in each <code>transcript_df</code></p></dd>


<dt id="arg-transgene-coln">transGene_coln<a class="anchor" aria-label="anchor" href="#arg-transgene-coln"></a></dt>
<dd><p>the column name of target or gene name in <code>transcript_df</code></p></dd>


<dt id="arg-cellid-coln">cellID_coln<a class="anchor" aria-label="anchor" href="#arg-cellid-coln"></a></dt>
<dd><p>the column name of cell_ID in <code>transcript_df</code>; when <code>prefix_colns</code> != NULL, unique cell_ID would be generated from <code>prefix_colns</code> and <code>cellID_coln</code> in each <code>transcript_df</code></p></dd>


<dt id="arg-spatlocs-colns">spatLocs_colns<a class="anchor" aria-label="anchor" href="#arg-spatlocs-colns"></a></dt>
<dd><p>column names for 1st, 2nd and optional 3rd dimension of spatial coordinates in <code>transcript_df</code></p></dd>


<dt id="arg-invert-y">invert_y<a class="anchor" aria-label="anchor" href="#arg-invert-y"></a></dt>
<dd><p>flag to invert y axis of local coordinates during stitching (default = TRUE)</p></dd>


<dt id="arg-extracellular-cellid">extracellular_cellID<a class="anchor" aria-label="anchor" href="#arg-extracellular-cellid"></a></dt>
<dd><p>a vector of cell_ID for extracellular transcripts which would be removed from the resegmention pipeline (default = NULL)</p></dd>


<dt id="arg-flagmodel-transnum-cutoff">flagModel_TransNum_cutoff<a class="anchor" aria-label="anchor" href="#arg-flagmodel-transnum-cutoff"></a></dt>
<dd><p>the cutoff of transcript number to do spatial modeling for identification of wrongly segmented cells (default = 50)</p></dd>


<dt id="arg-flagcell-lrtest-cutoff">flagCell_lrtest_cutoff<a class="anchor" aria-label="anchor" href="#arg-flagcell-lrtest-cutoff"></a></dt>
<dd><p>the cutoff of lrtest_nlog10P to identify putative wrongly segemented cells with strong spatial dependency in transcript score profile</p></dd>


<dt id="arg-svmclass-score-cutoff">svmClass_score_cutoff<a class="anchor" aria-label="anchor" href="#arg-svmclass-score-cutoff"></a></dt>
<dd><p>the cutoff of transcript score to separate between high and low score transcripts in SVM (default = -2)</p></dd>


<dt id="arg-svm-args">svm_args<a class="anchor" aria-label="anchor" href="#arg-svm-args"></a></dt>
<dd><p>a list of arguments to pass to svm function for identifying low-score transcript groups in space, typically involve kernel, gamma, scale</p></dd>


<dt id="arg-molecular-distance-cutoff">molecular_distance_cutoff<a class="anchor" aria-label="anchor" href="#arg-molecular-distance-cutoff"></a></dt>
<dd><p>maximum molecule-to-molecule distance within connected transcript group, unit in micron (default = 2.7 micron).
If set to NULL, the pipeline would first randomly choose no more than 2500 cells from up to 10 random picked ROIs with search radius to be 5 times of <code>cellular_distance_cutoff</code>, and then calculate the minimal molecular distance between picked cells. The pipeline would further use the 5 times of 90% quantile of minimal molecular distance as <code>molecular_distance_cutoff</code>. This calculation is slow and is not recommended for large transcript data.frame.</p></dd>


<dt id="arg-cellular-distance-cutoff">cellular_distance_cutoff<a class="anchor" aria-label="anchor" href="#arg-cellular-distance-cutoff"></a></dt>
<dd><p>maximum cell-to-cell distance in x, y between the center of query cells to the center of neighbor cells with direct contact, unit in micron. Default = NULL to use the 2 times of average 2D cell diameter.</p></dd>


<dt id="arg-score-baseline">score_baseline<a class="anchor" aria-label="anchor" href="#arg-score-baseline"></a></dt>
<dd><p>a named vector of score baseline under each cell type listed in <code>refProfiles</code> such that  per cell transcript score higher than the baseline is required to call a cell type of high enough confidence; default = NULL to calculate from <code>counts</code> and <code>refProfiles</code></p></dd>


<dt id="arg-lowercutoff-transnum">lowerCutoff_transNum<a class="anchor" aria-label="anchor" href="#arg-lowercutoff-transnum"></a></dt>
<dd><p>a named vector of transcript number cutoff under each cell type such that higher than the cutoff is required to keep query cell as it is; default = NULL to calculate from <code>counts</code> and <code>refProfiles</code></p></dd>


<dt id="arg-highercutoff-transnum">higherCutoff_transNum<a class="anchor" aria-label="anchor" href="#arg-highercutoff-transnum"></a></dt>
<dd><p>a named vector of transcript number cutoff under each cell type such that lower than the cutoff is required to keep query cell as it is when there is neighbor cell of consistent cell type; default = NULL to calculate from <code>counts</code> and <code>refProfiles</code></p></dd>


<dt id="arg-imputeflag-missingcts">imputeFlag_missingCTs<a class="anchor" aria-label="anchor" href="#arg-imputeflag-missingcts"></a></dt>
<dd><p>flag to impute <code>score_baseline</code>, <code>lowerCutoff_transNum</code>,<code>higherCutoff_transNum</code> for cell types present in <code>refProfiles</code> but missing in the provided transcript data files or the provided baseline and cutoffs; when TRUE, the median values of existing cell types would be used as the values for missing cell types.</p></dd>


<dt id="arg-grouptranscripts-method">groupTranscripts_method<a class="anchor" aria-label="anchor" href="#arg-grouptranscripts-method"></a></dt>
<dd><p>use either "dbscan" or "delaunay method" to group transcripts in space (default = "dbscan")</p></dd>


<dt id="arg-spatialmergecheck-method">spatialMergeCheck_method<a class="anchor" aria-label="anchor" href="#arg-spatialmergecheck-method"></a></dt>
<dd><p>use either "leidenCut" (in 2D or 3D) or "geometryDiff" (in 2D only) method to determine whether a cell pair merging event is allowed in space (default = "leidenCut")</p></dd>


<dt id="arg-cutoff-spatialmerge">cutoff_spatialMerge<a class="anchor" aria-label="anchor" href="#arg-cutoff-spatialmerge"></a></dt>
<dd><p>spatial constraint on a valid merging event between two source transcript groups, default = 0.5 for 50% cutoff, set to 0 to skip spatial constraint evaluation for merging.
For <code>spatialMergeCheck_method = "leidenCut"</code>, this is the minimal percentage of transcripts shared membership between query cell and neighbor cells in leiden clustering results for a valid merging event.
For <code>spatialMergeCheck_method = "geometryDiff"</code>, this is the maximum percentage of white space change upon merging of query cell and neighbor cell for a valid merging event.</p></dd>


<dt id="arg-leiden-config">leiden_config<a class="anchor" aria-label="anchor" href="#arg-leiden-config"></a></dt>
<dd><p>(leidenCut) a list of configuration to pass to reticulate and <code><a href="https://r.igraph.org/reference/cluster_leiden.html" class="external-link">igraph::cluster_leiden</a></code> function, including objective_function, resolution_parameter, beta, n_iterations.</p></dd>


<dt id="arg-config-spatnw-transcript">config_spatNW_transcript<a class="anchor" aria-label="anchor" href="#arg-config-spatnw-transcript"></a></dt>
<dd><p>configuration list to create spatial network at transcript level, see manual for <code>createSpatialDelaunayNW_from_spatLocs</code> for more details, set to NULL to use default config (default = NULL)</p></dd>


<dt id="arg-path-to-output">path_to_output<a class="anchor" aria-label="anchor" href="#arg-path-to-output"></a></dt>
<dd><p>the file path to output folder where the resegmentation data is saved; directory would be created by function if not exists; transcript data.frame <code>updated_transDF</code> is saved as individual csv files for each FOV, while cell data of all FOVs, <code>updated_perCellDT</code> and <code>updated_perCellExprs</code>, are combined to save as <code>.rds</code> object.</p></dd>


<dt id="arg-transdf-export-option">transDF_export_option<a class="anchor" aria-label="anchor" href="#arg-transdf-export-option"></a></dt>
<dd><p>option on how to export updated transcript_df, 0 for no export, 1 for write to <code>path_to_output</code> in disk as csv for each FOV, 2 for return to function as list (default = 1)</p></dd>


<dt id="arg-save-intermediates">save_intermediates<a class="anchor" aria-label="anchor" href="#arg-save-intermediates"></a></dt>
<dd><p>flag to save intermediate outputs into output folder, including data.frame for spatial modeling statistics of each cell,</p></dd>


<dt id="arg-return-percelldata">return_perCellData<a class="anchor" aria-label="anchor" href="#arg-return-percelldata"></a></dt>
<dd><p>flag to return for gene x cell count matrix and per cell DF with updated mean spatial coordinates and new cell type, also export to output folder in disk when <code>transDF_export_option = 1</code>.</p></dd>


<dt id="arg-combine-extra">combine_extra<a class="anchor" aria-label="anchor" href="#arg-combine-extra"></a></dt>
<dd><p>flag to combine original extracellular transcripts and trimmed transcripts back to the updated transcript data.frame, slow process if many transcript in each FOV file. (default = FALSE)</p></dd>


<dt id="arg-ctrl-genes">ctrl_genes<a class="anchor" aria-label="anchor" href="#arg-ctrl-genes"></a></dt>
<dd><p>a vector of control genes that are present in input transcript data.frame but not in <code>refProfiles</code> and expect no cell type dependency, e.g. negative control probes; the <code>ctrl_genes</code> would be included in FastReseg analysis. (default = NULL)</p></dd>


<dt id="arg-seed-process">seed_process<a class="anchor" aria-label="anchor" href="#arg-seed-process"></a></dt>
<dd><p>seed for per FOV processing, used in transcript error detection and correction steps, default = NULL to skip the seed</p></dd>


<dt id="arg-percentcores">percentCores<a class="anchor" aria-label="anchor" href="#arg-percentcores"></a></dt>
<dd><p>percent of cores to use for parallel processing (0-1] (default = 0.75)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a list</p><dl><dt>refProfiles</dt>
<dd><p>a genes X clusters matrix of cluster-specific reference profiles used in resegmenation pipeline</p></dd>

<dt>baselineData</dt>
<dd><p>a list of two matrice in cluster X percentile format for the cluster-specific percentile distribution of per cell value; <code>span_score</code> is for the average per molecule transcript tLLR score of each cell, <code>span_transNum</code> is for the transcript number of each cell.</p></dd>

<dt>cutoffs_list</dt>
<dd><p>a list of cutoffs used in resegmentation pipeline, including, <code>score_baseline</code>, <code>lowerCutoff_transNum</code>, <code>higherCutoff_transNum</code>, <code>cellular_distance_cutoff</code>, <code>molecular_distance_cutoff</code></p></dd>

<dt>ctrl_genes</dt>
<dd><p>a vector of control genes whose transcript scores are set to fixed value for all cell types, return when <code>ctrl_genes</code> is not NULL.</p></dd>

<dt>updated_perCellDT</dt>
<dd><p>a per cell data.table with mean spatial coordinates, new cell type and resegmentation action after resegmentation, return when <code>return_perCellData</code> = TRUE</p></dd>

<dt>updated_perCellExprs</dt>
<dd><p>a gene x cell count sparse matrix for updated transcript data.frame after resegmentation, return when <code>return_perCellData</code> = TRUE</p></dd>

<dt>reseg_actions</dt>
<dd><p>a list of 4 elements describing how the resegmenation would be performed on original <code>transcript_df</code> by the group assignment of transcripts listed in <code>groupDF_ToFlagTrans</code>, output of <code>decide_ReSegment_Operations</code> function, return when <code>save_intermediates</code> = TRUE</p></dd>

<dt>updated_transDF_list</dt>
<dd><p>a list of per-FOV transcript data.frame with updated cell segmenation in <code>updated_cellID</code> and <code>updated_celltype</code> columns, return when <code>transDF_export_option = 2</code></p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The pipeline would first estimate mean profile for each cell cluster based on the provided cell x gene count matrix and cluster assignment for entire data set.
And then, the pipeline would use the estimated cluster-specific profile as reference profiles and calculate suitable cutoff for distance search, transcript number and score in first provided per FOV transcript data frame when those cutoffs are not provided.
When transcript data.frame is provided as multiple file paths in <code>transDF_fileInfo</code> data.frame, the pipeline would further perform resegmentation on individual transcript data.frame using the baseline and cutoff defined globally.
For each transcript data.frame, the pipeline would score each transcript based on the provided cell type-specific reference profiles, evaluate the goodness-of-fit of each transcript within original cell segment,
identify the low-score transcript groups within cells that has strong spatial dependency in transcript score profile,
evaluate the neighborhood environment of low-score transcript groups and perform resegmentation actions including triming to extracellular space, merging to neighbor cell or labeling as new cell.</p>
<p>To account for genes missing in <code>refProfiles</code> but present in input transcript data.frame, genes in <code>ctrl_genes</code> would be assigned with goodness-of-fit score equal to <code>svmClass_score_cutoff</code> for all cell types to minimize the impact of those genes on the identification of low-score transcript groups via SVM. To avoid significant interference from those <code>ctrl_genes</code>, it's recommended to have total counts of those genes below 1% of total counts of all genes in each cell.</p>
<p>The pipeline would save the each per FOV output as individual file in <code>path_to_output</code> directory; <code>updated_transDF</code> would be saved as csv file.
When <code>save_intermediates</code> = TRUE, all intermediate files and resegmenation outputs of each FOV would be saved as single <code>.rds</code> object which is a list containing the following elements:</p><dl><dt>modStats_ToFlagCells</dt>
<dd><p>a data.frame for spatial modeling statistics of each cell, output of <code>score_cell_segmentation_error</code> function, save when <code>save_intermediates</code> = TRUE</p></dd>

<dt>groupDF_ToFlagTrans</dt>
<dd><p>data.frame for the group assignment of transcripts within putative wrongly segmented cells, merged output of <code>flag_bad_transcripts</code> and <code>groupTranscripts_Delaunay</code> or <code>groupTranscripts_dbscan</code> functions, save when <code>save_intermediates</code> = TRUE</p></dd>

<dt>neighborhoodDF_ToReseg</dt>
<dd><p>a data.frame for neighborhood environment of low-score transcript groups, output of <code>get_neighborhood_content</code> function, save when <code>save_intermediates</code> = TRUE</p></dd>

<dt>reseg_actions</dt>
<dd><p>a list of 4 elements describing how the resegmenation would be performed on original <code>transcript_df</code> by the group assignment of transcripts listed in <code>groupDF_ToFlagTrans</code>, output of <code>decide_ReSegment_Operations</code> function, save when <code>save_intermediates</code> = TRUE</p></dd>

<dt>updated_transDF</dt>
<dd><p>the updated transcript_df with <code>updated_cellID</code> and <code>updated_celltype</code> column based on reseg_full_converter, write to disk when <code>transDF_export_option =1</code></p></dd>

<dt>updated_perCellDT</dt>
<dd><p>a per cell data.table with mean spatial coordinates, new cell type and resegmentation action after resegmentation, return when <code>return_perCellData</code> = TRUE</p></dd>

<dt>updated_perCellExprs</dt>
<dd><p>a gene x cell count sparse matrix for updated transcript data.frame after resegmentation, return when <code>return_perCellData</code> = TRUE</p></dd>


</dl><p>The pipeline would also combine per cell data for all FOVs and return the combined data when <code>return_perCellData</code> = TRUE; <code>updated_perCellDT</code> and <code>updated_perCellExprs</code> would also be saved in a list as single <code>.rds</code> object in <code>path_to_output</code> directory when  <code>transDF_export_option = 1</code>.</p><dl><dt>updated_perCellDT</dt>
<dd><p>a per cell data.table with mean spatial coordinates, new cell type and resegmentation action after resegmentation, return when <code>return_perCellData</code> = TRUE</p></dd>

<dt>updated_perCellExprs</dt>
<dd><p>a gene x cell count sparse matrix for updated transcript data.frame after resegmentation, return when <code>return_perCellData</code> = TRUE</p></dd>


</dl></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># get example based on example dataset</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"mini_transcriptDF"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ori_RawExprs"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_refProfiles"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_baselineCT"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># cell_ID for extracellualr transcripts</span></span></span>
<span class="r-in"><span><span class="va">extracellular_cellID</span> <span class="op">&lt;-</span> <span class="va">mini_transcriptDF</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">mini_transcriptDF</span><span class="op">$</span><span class="va">CellId</span> <span class="op">==</span><span class="fl">0</span><span class="op">)</span>, <span class="st">'cell_ID'</span><span class="op">]</span> </span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># case #'1: provide `transcript_df` directly,</span></span></span>
<span class="r-in"><span><span class="co"># do auto-calculation of distance cutoff from data while using the provided </span></span></span>
<span class="r-in"><span><span class="co"># cutoffs for score and transcript numbers.</span></span></span>
<span class="r-in"><span><span class="va">res1</span> <span class="op">&lt;-</span> <span class="fu">fastReseg_full_pipeline</span><span class="op">(</span>counts <span class="op">=</span> <span class="va">ori_RawExprs</span>,</span></span>
<span class="r-in"><span>                                clust <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>                                refProfiles <span class="op">=</span> <span class="va">example_refProfiles</span>,</span></span>
<span class="r-in"><span>                                pixel_size <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>                                zstep_size <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>                                transcript_df <span class="op">=</span> <span class="va">mini_transcriptDF</span>,</span></span>
<span class="r-in"><span>                                transID_coln <span class="op">=</span> <span class="st">"UMI_transID"</span>,</span></span>
<span class="r-in"><span>                                transGene_coln <span class="op">=</span> <span class="st">"target"</span>,</span></span>
<span class="r-in"><span>                                cellID_coln <span class="op">=</span> <span class="st">"UMI_cellID"</span>,</span></span>
<span class="r-in"><span>                                spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>,<span class="st">"y"</span>,<span class="st">"z"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                                extracellular_cellID <span class="op">=</span> <span class="va">extracellular_cellID</span>,</span></span>
<span class="r-in"><span>                                molecular_distance_cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>                                cellular_distance_cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>                                score_baseline <span class="op">=</span> <span class="va">example_baselineCT</span><span class="op">[[</span><span class="st">'score_baseline'</span><span class="op">]</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>                                lowerCutoff_transNum <span class="op">=</span> <span class="va">example_baselineCT</span><span class="op">[[</span><span class="st">'lowerCutoff_transNum'</span><span class="op">]</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>                                higherCutoff_transNum<span class="op">=</span> <span class="va">example_baselineCT</span><span class="op">[[</span><span class="st">'higherCutoff_transNum'</span><span class="op">]</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>                                imputeFlag_missingCTs <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                                path_to_output <span class="op">=</span> <span class="st">"res1_directDF"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Per-FOV transcript data.frame with updated cell segmentation would be exported to disk at `path_to_output = 'res1_directDF'`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The intermediate results and per cell data would be exported to output directory at `path_to_output`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 960 common genes among `refProfiles` and `counts`. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No common cell types/clusters found between `clust` and `refProfiles`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Perform cluster assignment based on maximum transcript score given the provided `refProfiles`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extract distance cutoff from first input transcript data.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3 Dimension of spaital coordinates are provided.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> A single `transcript_df` is provided with unique `cellID_coln` = UMI_cellID and `transID_coln` = UMI_transID (use row idx if NULL).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 2 times of average 2D cell diameter as cellular_distance_cutoff = 24.2375 for searching of neighbor cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 3D coordinates with variance. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of minimal molecular distance between 1375 cells: 0, 0.08, 0.14, 0.21, 0.3, 0.4, 0.51, 0.63, 0.78, 0.86, 3.92, at quantile = 0%, 10%, 20%, 30%, 40%, 50%, 60%, 70%, 80%, 90%, 100%.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 5 times of 90% quantile of minimal 3D molecular distance between picked cells as `molecular_distance_cutoff` = 4.2790 for defining direct neighbor cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use `molecular_distance_cutoff` = 4.2790 for defining direct neighbor cells based on molecule-to-molecule distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use cellular_distance_cutoff = 24.2375 for searching of neighbor cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ##############</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Processing file `1`: NA</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ##------ Sat Dec  6 01:23:33 2025 ------##</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3 Dimension of spaital coordinates are provided.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Perform leiden clustering at resolution_parameter = 1.000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Create Delanay network when config$method is NULL.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Name the spatial network based on method as `Delaunay_network` when config$name is NULL.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use cellular_distance_cutoff = 24.2375 for searching of neighbor cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use `molecular_distance_cutoff` = 4.2790 for defining direct neighbor cells based on molecule-to-molecule distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 960 unique genes are found in `transcript_df`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 960 unique genes are shared by the provided `score_GeneMatrix` that contains cluster profiles of 960 genes for 22 clusters.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 756783 transcript records and 1375 cells in input `transcript_df`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use the providied `molecular_distance_cutoff` = 4.2790 for defining direct neighbor cells based on molecule-to-molecule distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use the providied `cellular_distance_cutoff` = 24.2375 for searching of neighbor cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 960 common genes among transcript_df and score_GeneMatrix. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 1375 cells and assigned cell type based on the provided `refProfiles` cluster profiles.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Run linear regreassion in 3 Dimension.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Below model_cutoff = 50, skip 37 cells with fewer transcripts. Move forward with remaining 1338 cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 373 cells, 0.2788 of all evaluated cells, are flagged for resegmentation with lrtest_nlog10P &gt; 5.0.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Run SVM in 3 Dimension.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 373 common cells and 960 common genes among chosen_cells, transcript_df, and score_GeneMatrix. </span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Below model_cutoff = 50, skip 0 cells with fewer transcripts. Move forward with remaining 373 cells.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Skip 0 cells with all transcripts in same class given `score_cutoff = -2`. Move forward with remaining 373 cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Remove 0 cells with raw transcript score all in same class based on cutoff -2.00 when running spatial SVM model.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Do spatial network analysis in 3 Dimension.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 10652 chosen_transcripts are found in common cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> SVM spatial model further identified 17 cells with transcript score all in same class, exclude from transcript group analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 960 common genes among transcript_df and score_GeneMatrix. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Perform leiden clustering at resolution_parameter = 1.000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use neighbor_distance_xy = 24.2375 for searching of neighbor cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use distance_cutoff = 4.2790 for defining direct neighbor cells based on molecule-to-molecule distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use first 2D for searching cell neighborhood, but all 3 Dimension to identify direct neighbors based on molecular distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 1998 common cells and 960 common genes among transcript_df, cell_networkDT, and score_GeneMatrix. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 623 chosen_cells are found in common cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use `leidenCut` method to evaluate putative merging event in space. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Perform leiden clustering at resolution_parameter = 1.000.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> A valid merging event must have query cell with 0.500 transcript shared same membership as neighbor cell of consistent cell type. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Run delanuay network in 3 Dimension.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Perform ledien clustering on 28 potential merging events. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> (`c_1_2_1731_g2`, `c_1_2_1792_g5`) cell pair with all 5 transcripts in same z plane, run 2D network analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> (`c_1_2_1792_g5`, `c_1_2_1731_g2`) cell pair with all 5 transcripts in same z plane, run 2D network analysis.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 623 common cells and 960 common genes among `names(reseg_full_converter)`, `transcript_df`, and `score_GeneMatrix`. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Trim 5626 transcripts during resegmentation, 0.0074 of all intracellular molecules.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>object 'transcript_df' not found</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># case #'2: provide file paths to per FOV transcript data files and specify </span></span></span>
<span class="r-in"><span><span class="co"># the spatial offset for each FOV,</span></span></span>
<span class="r-in"><span><span class="co"># do auto-calculation of score and transcript number cutoffs from gene </span></span></span>
<span class="r-in"><span><span class="co"># expression matrix, `counts`, and cluster assignment of each cell, `clust`,</span></span></span>
<span class="r-in"><span><span class="co"># do auto-calculation of distance cutoff from the 1st per FOV transcript data.</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_CellGeneExpr"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_clust"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># the example individual transcript files are stored under `data` directory of this package</span></span></span>
<span class="r-in"><span><span class="co"># update your path accordingly</span></span></span>
<span class="r-in"><span><span class="co"># Notice that some assays like SMI has XY axes swapped between stage and each FOV;</span></span></span>
<span class="r-in"><span><span class="co"># coordinates for each FOV should have units in micron</span></span></span>
<span class="r-in"><span><span class="va">dataDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"FastReseg"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fileInfo_DF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  file_path <span class="op">=</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="va">dataDir</span>,  </span></span>
<span class="r-in"><span>                       <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Run4104_FOV001__complete_code_cell_target_call_coord.csv"</span>,</span></span>
<span class="r-in"><span>                         <span class="st">"Run4104_FOV002__complete_code_cell_target_call_coord.csv"</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  slide <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  fov <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  stage_X <span class="op">=</span> <span class="fl">1000</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5.13</span>, <span class="op">-</span><span class="fl">2.701</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  stage_Y <span class="op">=</span> <span class="fl">1000</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.452</span>, <span class="fl">0.081</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">res2</span> <span class="op">&lt;-</span> <span class="fu">fastReseg_full_pipeline</span><span class="op">(</span>counts <span class="op">=</span> <span class="va">example_CellGeneExpr</span>,</span></span>
<span class="r-in"><span>                                clust <span class="op">=</span> <span class="va">example_clust</span>,</span></span>
<span class="r-in"><span>                                refProfiles <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>                                transDF_fileInfo <span class="op">=</span><span class="va">fileInfo_DF</span>,</span></span>
<span class="r-in"><span>                                filepath_coln <span class="op">=</span> <span class="st">'file_path'</span>,</span></span>
<span class="r-in"><span>                                prefix_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'slide'</span>,<span class="st">'fov'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                                </span></span>
<span class="r-in"><span>                                <span class="co"># match XY axes between stage and each FOV</span></span></span>
<span class="r-in"><span>                                fovOffset_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'stage_Y'</span>,<span class="st">'stage_X'</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>                                <span class="co"># 0.18 micron per pixel in transcript data</span></span></span>
<span class="r-in"><span>                                pixel_size <span class="op">=</span> <span class="fl">0.18</span>, </span></span>
<span class="r-in"><span>                                <span class="co"># 0.8 micron per z step in transcript data</span></span></span>
<span class="r-in"><span>                                zstep_size <span class="op">=</span> <span class="fl">0.8</span>, </span></span>
<span class="r-in"><span>                                </span></span>
<span class="r-in"><span>                                transcript_df <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>                                </span></span>
<span class="r-in"><span>                                <span class="co"># row index as transcript_id</span></span></span>
<span class="r-in"><span>                                transID_coln <span class="op">=</span> <span class="cn">NULL</span>, </span></span>
<span class="r-in"><span>                                </span></span>
<span class="r-in"><span>                                transGene_coln <span class="op">=</span> <span class="st">"target"</span>,</span></span>
<span class="r-in"><span>                                cellID_coln <span class="op">=</span> <span class="st">"CellId"</span>,</span></span>
<span class="r-in"><span>                                spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>,<span class="st">"y"</span>,<span class="st">"z"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                                </span></span>
<span class="r-in"><span>                                <span class="co"># CellId = 0 means extracelluar transcripts in raw data</span></span></span>
<span class="r-in"><span>                                extracellular_cellID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>                                </span></span>
<span class="r-in"><span>                                molecular_distance_cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>                                cellular_distance_cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>                                score_baseline <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>                                lowerCutoff_transNum <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>                                higherCutoff_transNum<span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>                                imputeFlag_missingCTs <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                                path_to_output <span class="op">=</span> <span class="st">"res2_multiFiles"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Per-FOV transcript data.frame with updated cell segmentation would be exported to disk at `path_to_output = 'res2_multiFiles'`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The intermediate results and per cell data would be exported to output directory at `path_to_output`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 960 common genes among `refProfiles` and `counts`. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extract distance cutoff from first input transcript data.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3 Dimension of spaital coordinates are provided.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2 individual per FOV files are provided in `transDF_fileInfo`, use the 1st file to calculate distance cutoffs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `transID_coln` and `cellID_coln` of each per FOV transcript_df would be re-named based on `prefix_colns` = `slide`,`fov`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 2 times of average 2D cell diameter as cellular_distance_cutoff = 19.1980 for searching of neighbor cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 3D coordinates with variance. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of minimal molecular distance between 385 cells: 0, 0.16, 0.22, 0.29, 0.36, 0.44, 0.52, 0.63, 0.76, 0.84, 8.01, at quantile = 0%, 10%, 20%, 30%, 40%, 50%, 60%, 70%, 80%, 90%, 100%.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 5 times of 90% quantile of minimal 3D molecular distance between picked cells as `molecular_distance_cutoff` = 4.2191 for defining direct neighbor cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use `molecular_distance_cutoff` = 4.2191 for defining direct neighbor cells based on molecule-to-molecule distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use cellular_distance_cutoff = 19.1980 for searching of neighbor cells.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># case #'3: provide file paths to per FOV transcript data files and specify </span></span></span>
<span class="r-in"><span><span class="co"># the spatial offset for each FOV,</span></span></span>
<span class="r-in"><span><span class="co"># do auto-calculation of score and transcript number cutoffs from gene </span></span></span>
<span class="r-in"><span><span class="co"># expression matrix, `counts`, and cluster-specific reference profiles, `refProfiles`,</span></span></span>
<span class="r-in"><span><span class="co"># use the provided distance cutoff for `molecular_distance_cutoff` but </span></span></span>
<span class="r-in"><span><span class="co"># calculate the `cellular_distance_cutoff`</span></span></span>
<span class="r-in"><span><span class="va">res3</span> <span class="op">&lt;-</span> <span class="fu">fastReseg_full_pipeline</span><span class="op">(</span>counts <span class="op">=</span> <span class="va">example_CellGeneExpr</span>,</span></span>
<span class="r-in"><span>                                clust <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>                                refProfiles <span class="op">=</span> <span class="va">example_refProfiles</span>,</span></span>
<span class="r-in"><span>                                transDF_fileInfo <span class="op">=</span><span class="va">fileInfo_DF</span>,</span></span>
<span class="r-in"><span>                                filepath_coln <span class="op">=</span> <span class="st">'file_path'</span>,</span></span>
<span class="r-in"><span>                                prefix_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'slide'</span>,<span class="st">'fov'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                                fovOffset_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'stage_Y'</span>,<span class="st">'stage_X'</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>                                pixel_size <span class="op">=</span> <span class="fl">0.18</span>, </span></span>
<span class="r-in"><span>                                zstep_size <span class="op">=</span> <span class="fl">0.8</span>, </span></span>
<span class="r-in"><span>                                transcript_df <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>                                transID_coln <span class="op">=</span> <span class="cn">NULL</span>, </span></span>
<span class="r-in"><span>                                transGene_coln <span class="op">=</span> <span class="st">"target"</span>,</span></span>
<span class="r-in"><span>                                cellID_coln <span class="op">=</span> <span class="st">"CellId"</span>,</span></span>
<span class="r-in"><span>                                spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>,<span class="st">"y"</span>,<span class="st">"z"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                                extracellular_cellID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>                                molecular_distance_cutoff <span class="op">=</span> <span class="fl">2.7</span>,</span></span>
<span class="r-in"><span>                                cellular_distance_cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>                                score_baseline <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>                                lowerCutoff_transNum <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>                                higherCutoff_transNum<span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>                                imputeFlag_missingCTs <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                                path_to_output <span class="op">=</span> <span class="st">"res3_multiFiles"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Per-FOV transcript data.frame with updated cell segmentation would be exported to disk at `path_to_output = 'res3_multiFiles'`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The intermediate results and per cell data would be exported to output directory at `path_to_output`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 960 common genes among `refProfiles` and `counts`. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No common cell types/clusters found between `clust` and `refProfiles`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Perform cluster assignment based on maximum transcript score given the provided `refProfiles`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extract distance cutoff from first input transcript data.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3 Dimension of spaital coordinates are provided.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2 individual per FOV files are provided in `transDF_fileInfo`, use the 1st file to calculate distance cutoffs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `transID_coln` and `cellID_coln` of each per FOV transcript_df would be re-named based on `prefix_colns` = `slide`,`fov`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 2 times of average 2D cell diameter as cellular_distance_cutoff = 19.1980 for searching of neighbor cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use `cellular_distance_cutoff` = 19.1980 for searching of neighbor cells.</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Patrick Danaher, Lidan Wu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

