<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ FastReseg</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-#E6F3FC" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FastReseg</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="external-link dropdown-item" href="https://www.nature.com/articles/s41598-025-08733-5">Working principles and Manuscript</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Workflow and core functions</h6></li>
    <li><a class="dropdown-item" href="../articles/tutorial.html">FastReseg to detect and correct segmentation error in spatial transcriptome data</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Nanostring-Biostats/FastReseg/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="pipeline-wrapper">Pipeline wrapper<a class="anchor" aria-label="anchor" href="#pipeline-wrapper"></a></h2>

      <div class="section-desc"><p>Wrapper function to process multi-FOV dataset or per-FOV data</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="fastReseg_flag_all_errors.html">fastReseg_flag_all_errors()</a></code>

        </dt>
        <dd>fastReseg_flag_all_errors</dd>
      </dl><dl><dt>

          <code><a href="fastReseg_full_pipeline.html">fastReseg_full_pipeline()</a></code>

        </dt>
        <dd>fastReseg_full_pipeline</dd>
      </dl><dl><dt>

          <code><a href="fastReseg_perFOV_full_process.html">fastReseg_perFOV_full_process()</a></code>

        </dt>
        <dd>fastReseg_perFOV_full_process</dd>
      </dl></div><div class="section level2">
      <h2 id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a></h2>

      <div class="section-desc"><p>Functions to prepare input data and get baseline</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="runPreprocess.html">runPreprocess()</a></code>

        </dt>
        <dd>runPreprocess</dd>
      </dl><dl><dt>

          <code><a href="get_baselineCT.html">get_baselineCT()</a></code>

        </dt>
        <dd>get_baselineCT</dd>
      </dl><dl><dt>

          <code><a href="choose_distance_cutoff.html">choose_distance_cutoff()</a></code>

        </dt>
        <dd>choose_distance_cutoff</dd>
      </dl><dl><dt>

          <code><a href="estimate_MeanProfile.html">estimate_MeanProfile()</a></code>

        </dt>
        <dd>estimate_MeanProfile</dd>
      </dl><dl><dt>

          <code><a href="scoreGenesInRef.html">scoreGenesInRef()</a></code>

        </dt>
        <dd>scoreGenesInRef</dd>
      </dl><dl><dt>

          <code><a href="getCellType_maxScore.html">getCellType_maxScore()</a></code>

        </dt>
        <dd>getCellType_maxScore</dd>
      </dl><dl><dt>

          <code><a href="getScoreCellType_gene.html">getScoreCellType_gene()</a></code>

        </dt>
        <dd>getScoreCellType_gene</dd>
      </dl></div><div class="section level2">
      <h2 id="example-data">Example data<a class="anchor" aria-label="anchor" href="#example-data"></a></h2>

      <div class="section-desc"><p>Example data to use with tutorial</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="example_CellGeneExpr.html">example_CellGeneExpr</a></code>

        </dt>
        <dd>per cell gene expression matrix based on raw transcript data files generated by SMI</dd>
      </dl><dl><dt>

          <code><a href="example_baselineCT.html">example_baselineCT</a></code>

        </dt>
        <dd>Percentile profile for single cell distribution of transcript number and average tLLR score from example dataset for spatial transcriptional profiling of tissue</dd>
      </dl><dl><dt>

          <code><a href="example_clust.html">example_clust</a></code>

        </dt>
        <dd>Cluster assignment for cells based on per cell gene expression profile and reference profiles in example data set</dd>
      </dl><dl><dt>

          <code><a href="example_refProfiles.html">example_refProfiles</a></code>

        </dt>
        <dd>Data for cell type specific reference profiles</dd>
      </dl><dl><dt>

          <code><a href="mini_transcriptDF.html">mini_transcriptDF</a></code>

        </dt>
        <dd>Transcript level data frame from Example dataset for spatial transcriptional profiling of tissue</dd>
      </dl><dl><dt>

          <code><a href="ori_CellStatsDF.html">ori_CellStatsDF</a></code>

        </dt>
        <dd>Cell level data frame from Example dataset for spatial transcriptional profiling of tissue</dd>
      </dl><dl><dt>

          <code><a href="ori_RawExprs.html">ori_RawExprs</a></code>

        </dt>
        <dd>Cell x Gene expression matrix from Example dataset for spatial transcriptional profiling of tissue</dd>
      </dl></div><div class="section level2">
      <h2 id="modular-functions-of-pipeline">Modular functions of pipeline<a class="anchor" aria-label="anchor" href="#modular-functions-of-pipeline"></a></h2>

      <div class="section-desc"><p>Functions to perform individual tasks for each FOV</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="runPreprocess.html">runPreprocess()</a></code>

        </dt>
        <dd>runPreprocess</dd>
      </dl><dl><dt>

          <code><a href="runSegErrorEvaluation.html">runSegErrorEvaluation()</a></code>

        </dt>
        <dd>runSegErrorEvaluation</dd>
      </dl><dl><dt>

          <code><a href="runTranscriptErrorDetection.html">runTranscriptErrorDetection()</a></code>

        </dt>
        <dd>runTranscriptErrorDetection</dd>
      </dl><dl><dt>

          <code><a href="runSegRefinement.html">runSegRefinement()</a></code>

        </dt>
        <dd>runSegRefinement</dd>
      </dl></div><div class="section level2">
      <h2 id="flag-poor-fitted-cells">Flag poor fitted cells<a class="anchor" aria-label="anchor" href="#flag-poor-fitted-cells"></a></h2>

      <div class="section-desc"><p>Functions to detect cell segmentation errors</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="runSegErrorEvaluation.html">runSegErrorEvaluation()</a></code>

        </dt>
        <dd>runSegErrorEvaluation</dd>
      </dl><dl><dt>

          <code><a href="score_cell_segmentation_error.html">score_cell_segmentation_error()</a></code>

        </dt>
        <dd>score_cell_segmentation_error</dd>
      </dl></div><div class="section level2">
      <h2 id="flag-poor-fitted-transcripts">Flag poor fitted transcripts<a class="anchor" aria-label="anchor" href="#flag-poor-fitted-transcripts"></a></h2>

      <div class="section-desc"><p>Functions to identify wrongly segmented transcript groups</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="runTranscriptErrorDetection.html">runTranscriptErrorDetection()</a></code>

        </dt>
        <dd>runTranscriptErrorDetection</dd>
      </dl><dl><dt>

          <code><a href="flag_bad_transcripts.html">flag_bad_transcripts()</a></code>

        </dt>
        <dd>flag_bad_transcripts</dd>
      </dl><dl><dt>

          <code><a href="groupTranscripts_dbscan.html">groupTranscripts_dbscan()</a></code>

        </dt>
        <dd>groupTranscripts_dbscan</dd>
      </dl><dl><dt>

          <code><a href="groupTranscripts_Delaunay.html">groupTranscripts_Delaunay()</a></code>

        </dt>
        <dd>groupTranscripts_Delaunay</dd>
      </dl><dl><dt>

          <code><a href="createSpatialDelaunayNW_from_spatLocs.html">createSpatialDelaunayNW_from_spatLocs()</a></code>

        </dt>
        <dd>createSpatialDelaunayNW_from_spatLocs</dd>
      </dl><dl><dt>

          <code><a href="flagTranscripts_LDA_hyperplane.html">flagTranscripts_LDA_hyperplane()</a></code>

        </dt>
        <dd>flagTranscripts_LDA_hyperplane</dd>
      </dl></div><div class="section level2">
      <h2 id="segmentation-refinement">Segmentation refinement<a class="anchor" aria-label="anchor" href="#segmentation-refinement"></a></h2>

      <div class="section-desc"><p>Functions to evaluate neighborhood of flagged transcript groups and perform refinement</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="runSegRefinement.html">runSegRefinement()</a></code>

        </dt>
        <dd>runSegRefinement</dd>
      </dl><dl><dt>

          <code><a href="prepResegDF.html">prepResegDF()</a></code>

        </dt>
        <dd>prepResegDF</dd>
      </dl><dl><dt>

          <code><a href="get_neighborhood_content.html">get_neighborhood_content()</a></code>

        </dt>
        <dd>get_neighborhood_content</dd>
      </dl><dl><dt>

          <code><a href="getNeighbors_transDF.html">getNeighbors_transDF()</a></code>

        </dt>
        <dd>getNeighbors_transDF</dd>
      </dl><dl><dt>

          <code><a href="decide_ReSegment_Operations.html">decide_ReSegment_Operations()</a></code>

        </dt>
        <dd>decide_ReSegment_Operations</dd>
      </dl><dl><dt>

          <code><a href="update_transDF_ResegActions.html">update_transDF_ResegActions()</a></code>

        </dt>
        <dd>update_transDF_ResegActions</dd>
      </dl><dl><dt>

          <code><a href="transDF_to_perCell_data.html">transDF_to_perCell_data()</a></code>

        </dt>
        <dd>transDF_to_perCell_data</dd>
      </dl></div><div class="section level2">
      <h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a></h2>

      <div class="section-desc"><p>Auxillary function to visualize spatial pattern of transcript-level information of multiple cells</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="plotSpatialScoreMultiCells.html">plotSpatialScoreMultiCells()</a></code>

        </dt>
        <dd>plotSpatialScoreMultiCells</dd>
      </dl></div><div class="section level2">
      <h2 id="ultilties">Ultilties<a class="anchor" aria-label="anchor" href="#ultilties"></a></h2>

      <div class="section-desc"><p>Auxillary function to set parallel processing and digesting input files</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="numCores.html">numCores()</a></code>

        </dt>
        <dd>Get number of cores for parallelized operations</dd>
      </dl><dl><dt>

          <code><a href="checkTransFileInputsAndLoadFirst.html">checkTransFileInputsAndLoadFirst()</a></code>

        </dt>
        <dd>checkTransFileInputsAndLoadFirst</dd>
      </dl><dl><dt>

          <code><a href="prepare_perFOV_transDF.html">prepare_perFOV_transDF()</a></code>

        </dt>
        <dd>prepare_perFOV_transDF</dd>
      </dl><dl><dt>

          <code><a href="myFun_fov_load.html">myFun_fov_load()</a></code>

        </dt>
        <dd>myFun_fov_load</dd>
      </dl></div><div class="section level2">
      <h2 id="miscellaneous">Miscellaneous<a class="anchor" aria-label="anchor" href="#miscellaneous"></a></h2>

      <div class="section-desc"><p>Supporting functions of various purposes</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="check_config_leiden.html">check_config_leiden()</a></code>

        </dt>
        <dd>check_config_leiden</dd>
      </dl><dl><dt>

          <code><a href="check_config_spatialNW.html">check_config_spatialNW()</a></code>

        </dt>
        <dd>check_config_spatialNW</dd>
      </dl><dl><dt>

          <code><a href="checkAndPrepInputs_perFOV.html">checkAndPrepInputs_perFOV()</a></code>

        </dt>
        <dd>checkAndPrepInputs_perFOV</dd>
      </dl><dl><dt>

          <code><a href="checkTypeLengthValue.html">checkTypeLengthValue()</a></code>

        </dt>
        <dd>checkTypeLengthValue</dd>
      </dl><dl><dt>

          <code><a href="combine_matrices_fast.html">combine_matrices_fast()</a></code>

        </dt>
        <dd>combine_matrices_fast</dd>
      </dl><dl><dt>

          <code><a href="compartment_and_add_extra.html">compartment_and_add_extra()</a></code>

        </dt>
        <dd>compartment_and_add_extra</dd>
      </dl><dl><dt>

          <code><a href="initializeAllSegRes.html">initializeAllSegRes()</a></code>

        </dt>
        <dd>initializeAllSegRes</dd>
      </dl><dl><dt>

          <code><a href="makeDummyOuts_perFOV.html">makeDummyOuts_perFOV()</a></code>

        </dt>
        <dd>makeDummyOuts_perFOV</dd>
      </dl><dl><dt>

          <code><a href="myFun_3point_singleCell.html">myFun_3point_singleCell()</a></code>

        </dt>
        <dd>myFun_3point_singleCell</dd>
      </dl><dl><dt>

          <code><a href="run_igraph_leiden.html">run_igraph_leiden()</a></code>

        </dt>
        <dd>run_igraph_leiden</dd>
      </dl><dl><dt>

          <code><a href="igraph_delete_edges.html">igraph_delete_edges()</a></code>

        </dt>
        <dd>igraph_delete_edges</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Patrick Danaher, Lidan Wu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

