<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>makeDummyOuts_perFOV — makeDummyOuts_perFOV • FastReseg</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="makeDummyOuts_perFOV — makeDummyOuts_perFOV"><meta name="description" content="supporting function for fastReseg_perFOV_full_process, makes dummy outputs based on input transcript_df in same format as the outputs of fastReseg_perFOV_full_process() but returns no modStats_ToFlagCells data.frame, used in case of no cells being flagged for cell segmentation error."><meta property="og:description" content="supporting function for fastReseg_perFOV_full_process, makes dummy outputs based on input transcript_df in same format as the outputs of fastReseg_perFOV_full_process() but returns no modStats_ToFlagCells data.frame, used in case of no cells being flagged for cell segmentation error."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-#E6F3FC" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FastReseg</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="external-link dropdown-item" href="https://www.nature.com/articles/s41598-025-08733-5">Working principles and Manuscript</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Workflow and core functions</h6></li>
    <li><a class="dropdown-item" href="../articles/tutorial.html">FastReseg to detect and correct segmentation error in spatial transcriptome data</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Nanostring-Biostats/FastReseg/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>makeDummyOuts_perFOV</h1>
      <small class="dont-index">Source: <a href="https://github.com/Nanostring-Biostats/FastReseg/blob/HEAD/R/fastReseg_perFOV_full_process.R" class="external-link"><code>R/fastReseg_perFOV_full_process.R</code></a></small>
      <div class="d-none name"><code>makeDummyOuts_perFOV.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>supporting function for <code>fastReseg_perFOV_full_process</code>, makes dummy outputs based on input <code>transcript_df</code> in same format as the outputs of <code><a href="fastReseg_perFOV_full_process.html">fastReseg_perFOV_full_process()</a></code> but returns no <code>modStats_ToFlagCells</code> data.frame, used in case of no cells being flagged for cell segmentation error.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">makeDummyOuts_perFOV</span><span class="op">(</span></span>
<span>  <span class="va">all_genes</span>,</span>
<span>  <span class="va">transcript_df</span>,</span>
<span>  transID_coln <span class="op">=</span> <span class="st">"UMI_transID"</span>,</span>
<span>  transGene_coln <span class="op">=</span> <span class="st">"target"</span>,</span>
<span>  cellID_coln <span class="op">=</span> <span class="st">"UMI_cellID"</span>,</span>
<span>  spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span>, <span class="st">"z"</span><span class="op">)</span>,</span>
<span>  return_intermediates <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  return_perCellData <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  includeAllRefGenes <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-all-genes">all_genes<a class="anchor" aria-label="anchor" href="#arg-all-genes"></a></dt>
<dd><p>vector of all genes consider in the analysis as listed in rows of <code>score_GeneMatrix</code> in parent function</p></dd>


<dt id="arg-transcript-df">transcript_df<a class="anchor" aria-label="anchor" href="#arg-transcript-df"></a></dt>
<dd><p>the data.frame for each transcript with columns for transcript_id, target or gene name, original cell_id, spatial coordinates, as well as <code>tLLR_maxCellType</code> and <code>score_tLLR_maxCellType</code> columns from earlier step in parent function</p></dd>


<dt id="arg-transid-coln">transID_coln<a class="anchor" aria-label="anchor" href="#arg-transid-coln"></a></dt>
<dd><p>the column name of transcript_ID in <code>transcript_df</code></p></dd>


<dt id="arg-transgene-coln">transGene_coln<a class="anchor" aria-label="anchor" href="#arg-transgene-coln"></a></dt>
<dd><p>the column name of target or gene name in <code>transcript_df</code></p></dd>


<dt id="arg-cellid-coln">cellID_coln<a class="anchor" aria-label="anchor" href="#arg-cellid-coln"></a></dt>
<dd><p>the column name of cell_ID in <code>transcript_df</code></p></dd>


<dt id="arg-spatlocs-colns">spatLocs_colns<a class="anchor" aria-label="anchor" href="#arg-spatlocs-colns"></a></dt>
<dd><p>column names for 1st, 2nd and optional 3rd dimension of spatial coordinates in <code>transcript_df</code></p></dd>


<dt id="arg-return-intermediates">return_intermediates<a class="anchor" aria-label="anchor" href="#arg-return-intermediates"></a></dt>
<dd><p>flag to return intermediate outputs, including data.frame for spatial modeling statistics of each cell</p></dd>


<dt id="arg-return-percelldata">return_perCellData<a class="anchor" aria-label="anchor" href="#arg-return-percelldata"></a></dt>
<dd><p>flag to return gene x cell count matrix and per cell DF with updated mean spatial coordinates and new cell type</p></dd>


<dt id="arg-includeallrefgenes">includeAllRefGenes<a class="anchor" aria-label="anchor" href="#arg-includeallrefgenes"></a></dt>
<dd><p>flag to include all genes in <code>score_GeneMatrix</code> in the returned <code>updated_perCellExprs</code> with missing genes of value 0 (default = FALSE)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a list</p><dl><dt>groupDF_ToFlagTrans</dt>
<dd><p>data.frame for the group assignment of transcripts within putative wrongly segmented cells, merged output of <code>flag_bad_transcripts</code> and <code>groupTranscripts_Delaunay</code> or <code>groupTranscripts_dbscan</code> functions, return when <code>return_intermediates</code> = TRUE</p></dd>

<dt>neighborhoodDF_ToReseg</dt>
<dd><p>a data.frame for neighborhood environment of low-score transcript groups, output of <code>get_neighborhood_content</code> function, return when <code>return_intermediates</code> = TRUE</p></dd>

<dt>reseg_actions</dt>
<dd><p>a list of 4 elements describing how the resegmenation would be performed on original <code>transcript_df</code> by the group assignment of transcripts listed in <code>groupDF_ToFlagTrans</code>, output of <code>decide_ReSegment_Operations</code> function, return when <code>return_intermediates</code> = TRUE</p></dd>

<dt>updated_transDF</dt>
<dd><p>the updated transcript_df with <code>updated_cellID</code> and <code>updated_celltype</code> column based on reseg_full_converter</p></dd>

<dt>updated_perCellDT</dt>
<dd><p>a per cell data.table with mean spatial coordinates, new cell type and resegmentation action after resegmentation, return when <code>return_perCellData</code> = TRUE</p></dd>

<dt>updated_perCellExprs</dt>
<dd><p>a gene x cell count sparse matrix for updated transcript data.frame after resegmentation, return when <code>return_perCellData</code> = TRUE</p></dd>


</dl></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Patrick Danaher, Lidan Wu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

