<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>runPreprocess — runPreprocess • FastReseg</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="runPreprocess — runPreprocess"><meta name="description" content="modular wrapper to get baseline data and cutoffs from entire dataset"><meta property="og:description" content="modular wrapper to get baseline data and cutoffs from entire dataset"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-#E6F3FC" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FastReseg</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="external-link dropdown-item" href="https://www.nature.com/articles/s41598-025-08733-5">Working principles and Manuscript</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Workflow and core functions</h6></li>
    <li><a class="dropdown-item" href="../articles/tutorial.html">FastReseg to detect and correct segmentation error in spatial transcriptome data</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Nanostring-Biostats/FastReseg/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>runPreprocess</h1>
      <small class="dont-index">Source: <a href="https://github.com/Nanostring-Biostats/FastReseg/blob/HEAD/R/runPreprocess.R" class="external-link"><code>R/runPreprocess.R</code></a></small>
      <div class="d-none name"><code>runPreprocess.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>modular wrapper to get baseline data and cutoffs from entire dataset</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">runPreprocess</span><span class="op">(</span></span>
<span>  <span class="va">counts</span>,</span>
<span>  clust <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  refProfiles <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  score_baseline <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  lowerCutoff_transNum <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  higherCutoff_transNum <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  imputeFlag_missingCTs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ctrl_genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  svmClass_score_cutoff <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>,</span>
<span>  molecular_distance_cutoff <span class="op">=</span> <span class="fl">2.7</span>,</span>
<span>  cellular_distance_cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transcript_df <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transDF_fileInfo <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filepath_coln <span class="op">=</span> <span class="st">"file_path"</span>,</span>
<span>  prefix_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"slide"</span>, <span class="st">"fov"</span><span class="op">)</span>,</span>
<span>  fovOffset_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"stage_X"</span>, <span class="st">"stage_Y"</span><span class="op">)</span>,</span>
<span>  pixel_size <span class="op">=</span> <span class="fl">0.18</span>,</span>
<span>  zstep_size <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  transID_coln <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transGene_coln <span class="op">=</span> <span class="st">"target"</span>,</span>
<span>  cellID_coln <span class="op">=</span> <span class="st">"CellId"</span>,</span>
<span>  spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span>, <span class="st">"z"</span><span class="op">)</span>,</span>
<span>  invert_y <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  extracellular_cellID <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-counts">counts<a class="anchor" aria-label="anchor" href="#arg-counts"></a></dt>
<dd><p>Counts matrix for entire dataset, cells X genes.</p></dd>


<dt id="arg-clust">clust<a class="anchor" aria-label="anchor" href="#arg-clust"></a></dt>
<dd><p>Vector of cluster assignments for each cell in <code>counts</code>, when NULL to automatically assign the cell cluster for each cell based on maximum transcript score of given the provided <code>refProfiles</code></p></dd>


<dt id="arg-refprofiles">refProfiles<a class="anchor" aria-label="anchor" href="#arg-refprofiles"></a></dt>
<dd><p>A matrix of cluster profiles, genes X clusters, default = NULL to use external cluster assignments.
Of note, when <code>refProfiles !=  NULL</code>, genes unique to <code>counts</code> but missing in <code>refProfiles</code> would be omitted from downstream analysis.</p></dd>


<dt id="arg-score-baseline">score_baseline<a class="anchor" aria-label="anchor" href="#arg-score-baseline"></a></dt>
<dd><p>a named vector of score baseline under each cell type listed in <code>refProfiles</code> such that  per cell transcript score higher than the baseline is required to call a cell type of high enough confidence; default = NULL to calculate from <code>counts</code> and <code>refProfiles</code></p></dd>


<dt id="arg-lowercutoff-transnum">lowerCutoff_transNum<a class="anchor" aria-label="anchor" href="#arg-lowercutoff-transnum"></a></dt>
<dd><p>a named vector of transcript number cutoff under each cell type such that higher than the cutoff is required to keep query cell as it is; default = NULL to calculate from <code>counts</code> and <code>refProfiles</code></p></dd>


<dt id="arg-highercutoff-transnum">higherCutoff_transNum<a class="anchor" aria-label="anchor" href="#arg-highercutoff-transnum"></a></dt>
<dd><p>a named vector of transcript number cutoff under each cell type such that lower than the cutoff is required to keep query cell as it is when there is neighbor cell of consistent cell type; default = NULL to calculate from <code>counts</code> and <code>refProfiles</code></p></dd>


<dt id="arg-imputeflag-missingcts">imputeFlag_missingCTs<a class="anchor" aria-label="anchor" href="#arg-imputeflag-missingcts"></a></dt>
<dd><p>flag to impute <code>score_baseline</code>, <code>lowerCutoff_transNum</code>,<code>higherCutoff_transNum</code> for cell types present in <code>refProfiles</code> but missing in the provided transcript data files or the provided baseline and cutoffs; when TRUE, the median values of existing cell types would be used as the values for missing cell types.</p></dd>


<dt id="arg-ctrl-genes">ctrl_genes<a class="anchor" aria-label="anchor" href="#arg-ctrl-genes"></a></dt>
<dd><p>a vector of control genes that are present in input transcript data.frame but not in <code>refProfiles</code> and expect no cell type dependency, e.g. negative control probes; the <code>ctrl_genes</code> would be included in FastReseg analysis. (default = NULL)</p></dd>


<dt id="arg-svmclass-score-cutoff">svmClass_score_cutoff<a class="anchor" aria-label="anchor" href="#arg-svmclass-score-cutoff"></a></dt>
<dd><p>the cutoff of transcript score to separate between high and low score transcripts in SVM, used as the score values for <code>ctrl_genes</code> (default = -2)</p></dd>


<dt id="arg-molecular-distance-cutoff">molecular_distance_cutoff<a class="anchor" aria-label="anchor" href="#arg-molecular-distance-cutoff"></a></dt>
<dd><p>maximum molecule-to-molecule distance within connected transcript group, unit in micron (default = 2.7 micron).
If set to NULL, the pipeline would first randomly choose no more than 2500 cells from up to 10 random picked ROIs with search radius to be 5 times of <code>cellular_distance_cutoff</code>, and then calculate the minimal molecular distance between picked cells. The pipeline would further use the 5 times of 90% quantile of minimal molecular distance as <code>molecular_distance_cutoff</code>. This calculation is slow and is not recommended for large transcript data.frame.</p></dd>


<dt id="arg-cellular-distance-cutoff">cellular_distance_cutoff<a class="anchor" aria-label="anchor" href="#arg-cellular-distance-cutoff"></a></dt>
<dd><p>maximum cell-to-cell distance in x, y between the center of query cells to the center of neighbor cells with direct contact, unit in micron. Default = NULL to use the 2 times of average 2D cell diameter.</p></dd>


<dt id="arg-transcript-df">transcript_df<a class="anchor" aria-label="anchor" href="#arg-transcript-df"></a></dt>
<dd><p>the data.frame of transcript level information with unique CellId, default = NULL to read from the <code>transDF_fileInfo</code></p></dd>


<dt id="arg-transdf-fileinfo">transDF_fileInfo<a class="anchor" aria-label="anchor" href="#arg-transdf-fileinfo"></a></dt>
<dd><p>a data.frame with each row for each individual file of per FOV transcript data.frame within which the coordinates and CellId are unique, columns include the file path of per FOV transcript data.frame file, annotation columns like slide and fov to be used as prefix when creating unique cell_ID across entire dataset; when NULL, use the provided <code>transcript_df</code> directly</p></dd>


<dt id="arg-filepath-coln">filepath_coln<a class="anchor" aria-label="anchor" href="#arg-filepath-coln"></a></dt>
<dd><p>the column name of each individual file of per FOV transcript data.frame in <code>transDF_fileInfo</code></p></dd>


<dt id="arg-prefix-colns">prefix_colns<a class="anchor" aria-label="anchor" href="#arg-prefix-colns"></a></dt>
<dd><p>the column names of annotation in <code>transDF_fileInfo</code>, to be added to the CellId as prefix when creating unique cell_ID for entire dataset; set to NULL if use the original <code>transID_coln</code> or <code>cellID_coln</code></p></dd>


<dt id="arg-fovoffset-colns">fovOffset_colns<a class="anchor" aria-label="anchor" href="#arg-fovoffset-colns"></a></dt>
<dd><p>the column name of coordinate offsets in 1st and 2nd dimension for each per FOV transcript data.frame in <code>transDF_fileInfo</code>, unit in micron
Notice that some assays like SMI has XY axes swapped between stage and each FOV such that <code>fovOffset_colns</code> should be c("stage_Y", "stage_X").</p></dd>


<dt id="arg-pixel-size">pixel_size<a class="anchor" aria-label="anchor" href="#arg-pixel-size"></a></dt>
<dd><p>the micrometer size of image pixel listed in 1st and 2nd dimension of <code>spatLocs_colns</code> of each <code>transcript_df</code></p></dd>


<dt id="arg-zstep-size">zstep_size<a class="anchor" aria-label="anchor" href="#arg-zstep-size"></a></dt>
<dd><p>the micrometer size of z-step for the optional 3rd dimension of <code>spatLocs_colns</code> of each <code>transcript_df</code></p></dd>


<dt id="arg-transid-coln">transID_coln<a class="anchor" aria-label="anchor" href="#arg-transid-coln"></a></dt>
<dd><p>the column name of transcript_ID in <code>transcript_df</code>, default = NULL to use row index of transcript in each <code>transcript_df</code>; when <code>prefix_colns</code> != NULL, unique transcript_id would be generated from <code>prefix_colns</code> and <code>transID_coln</code> in each <code>transcript_df</code></p></dd>


<dt id="arg-transgene-coln">transGene_coln<a class="anchor" aria-label="anchor" href="#arg-transgene-coln"></a></dt>
<dd><p>the column name of target or gene name in <code>transcript_df</code></p></dd>


<dt id="arg-cellid-coln">cellID_coln<a class="anchor" aria-label="anchor" href="#arg-cellid-coln"></a></dt>
<dd><p>the column name of cell_ID in <code>transcript_df</code>; when <code>prefix_colns</code> != NULL, unique cell_ID would be generated from <code>prefix_colns</code> and <code>cellID_coln</code> in each <code>transcript_df</code></p></dd>


<dt id="arg-spatlocs-colns">spatLocs_colns<a class="anchor" aria-label="anchor" href="#arg-spatlocs-colns"></a></dt>
<dd><p>column names for 1st, 2nd and optional 3rd dimension of spatial coordinates in <code>transcript_df</code></p></dd>


<dt id="arg-invert-y">invert_y<a class="anchor" aria-label="anchor" href="#arg-invert-y"></a></dt>
<dd><p>flag to invert y axis of local coordinates during stitching (default = TRUE)</p></dd>


<dt id="arg-extracellular-cellid">extracellular_cellID<a class="anchor" aria-label="anchor" href="#arg-extracellular-cellid"></a></dt>
<dd><p>a vector of cell_ID for extracellular transcripts which would be removed from the resegmention pipeline (default = NULL)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a nested list</p><dl><dt>clust</dt>
<dd><p>vector of cluster assignments for each cell in <code>counts</code>, used in caculating <code>baselineData</code></p></dd>

<dt>refProfiles</dt>
<dd><p>a genes X clusters matrix of cluster-specific reference profiles to use in resegmenation pipeline</p></dd>

<dt>baselineData</dt>
<dd><p>a list of two matrice in cluster X percentile format for the cluster-specific percentile distribution of per cell value; <code>span_score</code> is for the average per molecule transcript tLLR score of each cell, <code>span_transNum</code> is for the transcript number of each cell.</p></dd>

<dt>cutoffs_list</dt>
<dd><p>a list of cutoffs to use in resegmentation pipeline, including, <code>score_baseline</code>, <code>lowerCutoff_transNum</code>, <code>higherCutoff_transNum</code>, <code>cellular_distance_cutoff</code>, <code>molecular_distance_cutoff</code></p></dd>

<dt>ctrl_genes</dt>
<dd><p>a vector of control genes whose transcript scores are set to fixed value for all cell types, return when <code>ctrl_genes</code> is not NULL.</p></dd>

<dt>score_GeneMatrix</dt>
<dd><p>a gene x cell-type score matrix to use in resegmenation pipeline, the scores for <code>ctrl_genes</code> are set to be the same as <code>svmClass_score_cutoff</code></p></dd>

<dt>processed_1st_transDF</dt>
<dd><p>a list of 2 elements for the intracellular and extracellular transcript data.frame of the processed outcomes of 1st transcrip file</p></dd>


</dl><p>The <code>cutoffs_list</code> is a list containing</p><dl><dt>score_baseline</dt>
<dd><p>a named vector of score baseline under each cell type listed in <code>refProfiles</code> such that  per cell transcript score higher than the baseline is required to call a cell type of high enough confidence.</p></dd>

<dt>lowerCutoff_transNum</dt>
<dd><p>a named vector of transcript number cutoff under each cell type such that higher than the cutoff is required to keep query cell as it is.</p></dd>

<dt>higherCutoff_transNum</dt>
<dd><p>a named vector of transcript number cutoff under each cell type such that lower than the cutoff is required to keep query cell as it is when there is neighbor cell of consistent cell type.</p></dd>

<dt>cellular_distance_cutoff</dt>
<dd><p>maximum cell-to-cell distance in x, y between the center of query cells to the center of neighbor cells with direct contact, unit in micron.</p></dd>

<dt>molecular_distance_cutoff</dt>
<dd><p>maximum molecule-to-molecule distance within connected transcript group, unit in micron.</p></dd>


</dl></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span> </span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"mini_transcriptDF"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_CellGeneExpr"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_clust"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_refProfiles"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># cell_ID for extracellualr transcripts</span></span></span>
<span class="r-in"><span><span class="va">extracellular_cellID</span> <span class="op">&lt;-</span> <span class="va">mini_transcriptDF</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">mini_transcriptDF</span><span class="op">$</span><span class="va">CellId</span> <span class="op">==</span><span class="fl">0</span><span class="op">)</span>, <span class="st">'cell_ID'</span><span class="op">]</span> </span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># case 1: use `clust` and `transcript_df` directly, with known distance cutoffs</span></span></span>
<span class="r-in"><span><span class="va">prep_res1</span> <span class="op">&lt;-</span> <span class="fu">runPreprocess</span><span class="op">(</span></span></span>
<span class="r-in"><span>  counts <span class="op">=</span> <span class="va">example_CellGeneExpr</span>,</span></span>
<span class="r-in"><span>  clust <span class="op">=</span> <span class="va">example_clust</span>,</span></span>
<span class="r-in"><span>  refProfiles <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  score_baseline <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  lowerCutoff_transNum <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  higherCutoff_transNum<span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  imputeFlag_missingCTs <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>  ctrl_genes <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  svmClass_score_cutoff <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>,</span></span>
<span class="r-in"><span>  molecular_distance_cutoff <span class="op">=</span> <span class="fl">2.7</span>,</span></span>
<span class="r-in"><span>  cellular_distance_cutoff <span class="op">=</span> <span class="fl">20</span>,</span></span>
<span class="r-in"><span>  transcript_df <span class="op">=</span> <span class="va">mini_transcriptDF</span>, </span></span>
<span class="r-in"><span>  transDF_fileInfo <span class="op">=</span> <span class="cn">NULL</span>, </span></span>
<span class="r-in"><span>  pixel_size <span class="op">=</span> <span class="fl">0.18</span>,</span></span>
<span class="r-in"><span>  zstep_size <span class="op">=</span> <span class="fl">0.8</span>, </span></span>
<span class="r-in"><span>  transID_coln <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  transGene_coln <span class="op">=</span> <span class="st">"target"</span>,</span></span>
<span class="r-in"><span>  cellID_coln <span class="op">=</span> <span class="st">'CellId'</span>,</span></span>
<span class="r-in"><span>  spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'x'</span>,<span class="st">'y'</span>,<span class="st">'z'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  extracellular_cellID <span class="op">=</span> <span class="fl">0</span> </span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 960 common genes among `refProfiles` and `counts`. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use the providied `molecular_distance_cutoff` = 2.7000 for defining direct neighbor cells based on molecule-to-molecule distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use the providied `cellular_distance_cutoff` = 20.0000 for searching of neighbor cells.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># case 2: use `refProfiles` to get `clust`, use `transcript_df` directly, </span></span></span>
<span class="r-in"><span><span class="co"># unknown distance cutoffs</span></span></span>
<span class="r-in"><span><span class="va">prep_res2</span> <span class="op">&lt;-</span> <span class="fu">runPreprocess</span><span class="op">(</span></span></span>
<span class="r-in"><span>  counts <span class="op">=</span> <span class="va">example_CellGeneExpr</span>,</span></span>
<span class="r-in"><span>  clust <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  refProfiles <span class="op">=</span> <span class="va">example_refProfiles</span>,</span></span>
<span class="r-in"><span>  score_baseline <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  lowerCutoff_transNum <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  higherCutoff_transNum<span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  </span></span>
<span class="r-in"><span>  <span class="co"># impute for cell types missing in provided 'transcript_df' </span></span></span>
<span class="r-in"><span>  imputeFlag_missingCTs <span class="op">=</span> <span class="cn">TRUE</span>, </span></span>
<span class="r-in"><span>  </span></span>
<span class="r-in"><span>  ctrl_genes <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  svmClass_score_cutoff <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>,</span></span>
<span class="r-in"><span>  molecular_distance_cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  cellular_distance_cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  transcript_df <span class="op">=</span> <span class="va">mini_transcriptDF</span>, </span></span>
<span class="r-in"><span>  transDF_fileInfo <span class="op">=</span> <span class="cn">NULL</span>, </span></span>
<span class="r-in"><span>  pixel_size <span class="op">=</span> <span class="fl">0.18</span>,</span></span>
<span class="r-in"><span>  zstep_size <span class="op">=</span> <span class="fl">0.8</span>, </span></span>
<span class="r-in"><span>  transID_coln <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  transGene_coln <span class="op">=</span> <span class="st">"target"</span>,</span></span>
<span class="r-in"><span>  cellID_coln <span class="op">=</span> <span class="st">'CellId'</span>,</span></span>
<span class="r-in"><span>  spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'x'</span>,<span class="st">'y'</span>,<span class="st">'z'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  extracellular_cellID <span class="op">=</span> <span class="fl">0</span> </span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 960 common genes among `refProfiles` and `counts`. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No common cell types/clusters found between `clust` and `refProfiles`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Perform cluster assignment based on maximum transcript score given the provided `refProfiles`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extract distance cutoff from first input transcript data.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3 Dimension of spaital coordinates are provided.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 2 times of average 2D cell diameter as cellular_distance_cutoff = 4.3628 for searching of neighbor cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 3D coordinates with variance. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of minimal molecular distance between 1375 cells: 0, 0.01, 0.03, 0.04, 0.05, 0.07, 0.09, 0.11, 0.14, 0.18, 1.24, at quantile = 0%, 10%, 20%, 30%, 40%, 50%, 60%, 70%, 80%, 90%, 100%.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 5 times of 90% quantile of minimal 3D molecular distance between picked cells as `molecular_distance_cutoff` = 0.9003 for defining direct neighbor cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use `molecular_distance_cutoff` = 0.9003 for defining direct neighbor cells based on molecule-to-molecule distance.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use cellular_distance_cutoff = 4.3628 for searching of neighbor cells.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># case 3: provide both `refProfiles` and `clust`, use transDF_fileInfo for </span></span></span>
<span class="r-in"><span><span class="co"># multi-files, no known molecular distance cutoffs</span></span></span>
<span class="r-in"><span><span class="va">dataDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"FastReseg"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">transDF_fileInfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  file_path <span class="op">=</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="va">dataDir</span>,</span></span>
<span class="r-in"><span>                       <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Run4104_FOV001__complete_code_cell_target_call_coord.csv"</span>,</span></span>
<span class="r-in"><span>                         <span class="st">"Run4104_FOV002__complete_code_cell_target_call_coord.csv"</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  slide <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  fov <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  stage_X <span class="op">=</span> <span class="fl">1000</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5.13</span>, <span class="op">-</span><span class="fl">2.701</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  stage_Y <span class="op">=</span> <span class="fl">1000</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.452</span>, <span class="fl">0.081</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">prep_res3</span> <span class="op">&lt;-</span> <span class="fu">runPreprocess</span><span class="op">(</span></span></span>
<span class="r-in"><span>  counts <span class="op">=</span> <span class="va">example_CellGeneExpr</span>,</span></span>
<span class="r-in"><span>  clust <span class="op">=</span> <span class="va">example_clust</span>,</span></span>
<span class="r-in"><span>  refProfiles <span class="op">=</span> <span class="va">example_refProfiles</span>,</span></span>
<span class="r-in"><span>  score_baseline <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  lowerCutoff_transNum <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  higherCutoff_transNum<span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  imputeFlag_missingCTs <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  ctrl_genes <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  svmClass_score_cutoff <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>,</span></span>
<span class="r-in"><span>  molecular_distance_cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  cellular_distance_cutoff <span class="op">=</span> <span class="fl">20</span>,</span></span>
<span class="r-in"><span>  transcript_df <span class="op">=</span> <span class="cn">NULL</span>, </span></span>
<span class="r-in"><span>  transDF_fileInfo <span class="op">=</span> <span class="va">transDF_fileInfo</span>, </span></span>
<span class="r-in"><span>  filepath_coln <span class="op">=</span> <span class="st">'file_path'</span>, </span></span>
<span class="r-in"><span>  prefix_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'slide'</span>,<span class="st">'fov'</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>  fovOffset_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'stage_X'</span>,<span class="st">'stage_Y'</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>  pixel_size <span class="op">=</span> <span class="fl">0.18</span>,</span></span>
<span class="r-in"><span>  zstep_size <span class="op">=</span> <span class="fl">0.8</span>, </span></span>
<span class="r-in"><span>  transID_coln <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  transGene_coln <span class="op">=</span> <span class="st">"target"</span>,</span></span>
<span class="r-in"><span>  cellID_coln <span class="op">=</span> <span class="st">'CellId'</span>,</span></span>
<span class="r-in"><span>  spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'x'</span>,<span class="st">'y'</span>,<span class="st">'z'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  invert_y <span class="op">=</span> <span class="cn">TRUE</span>,  <span class="co"># flip y axis of local image coordinates</span></span></span>
<span class="r-in"><span>  extracellular_cellID <span class="op">=</span> <span class="fl">0</span> </span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 960 common genes among `refProfiles` and `counts`. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Extract distance cutoff from first input transcript data.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 3 Dimension of spaital coordinates are provided.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2 individual per FOV files are provided in `transDF_fileInfo`, use the 1st file to calculate distance cutoffs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> `transID_coln` and `cellID_coln` of each per FOV transcript_df would be re-named based on `prefix_colns` = `slide`,`fov`.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 2 times of average 2D cell diameter as cellular_distance_cutoff = 19.1980 for searching of neighbor cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 3D coordinates with variance. </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of minimal molecular distance between 385 cells: 0, 0.16, 0.22, 0.29, 0.36, 0.44, 0.52, 0.63, 0.76, 0.84, 8.01, at quantile = 0%, 10%, 20%, 30%, 40%, 50%, 60%, 70%, 80%, 90%, 100%.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use 5 times of 90% quantile of minimal 3D molecular distance between picked cells as `molecular_distance_cutoff` = 4.2191 for defining direct neighbor cells.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Use `molecular_distance_cutoff` = 4.2191 for defining direct neighbor cells based on molecule-to-molecule distance.</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Patrick Danaher, Lidan Wu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

