<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Detection and Correction of Cell Segmentation Error Based on Spatial Profile of Transcripts • FastReseg</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Detection and Correction of Cell Segmentation Error Based on Spatial Profile of Transcripts">
<meta name="description" content="An R package for detection and correction of cell segmentation error based on spatial profile of transcripts">
<meta property="og:description" content="An R package for detection and correction of cell segmentation error based on spatial profile of transcripts">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-#E6F3FC" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">FastReseg</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="external-link dropdown-item" href="https://www.biorxiv.org/content/10.1101/2024.12.05.627051v1.abstract">Working principles and Manuscript</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Workflow and core functions</h6></li>
    <li><a class="dropdown-item" href="articles/tutorial.html">FastReseg to detect and correct segmentation error in spatial transcriptome data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Nanostring-Biostats/FastReseg/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="fastreseg">FastReseg<a class="anchor" aria-label="anchor" href="#fastreseg"></a>
</h1></div>
<p>An R package for detection and correction of cell segmentation error based on spatial profile of transcripts</p>
<div class="section level4">
<h4 id="manuscript">
<a href="https://www.biorxiv.org/content/10.1101/2024.12.05.627051v1.abstract" class="external-link">Manuscript</a>:<a class="anchor" aria-label="anchor" href="#manuscript"></a>
</h4>
<p>Wu L, Beechem JM, Danaher P. FastReseg: using transcript locations to refine image-based cell segmentation results in spatial transcriptomics. <em>bioRxiv</em> 2024.12.05.627051; doi: <a href="https://doi.org/10.1101/2024.12.05.627051" class="external-link uri">https://doi.org/10.1101/2024.12.05.627051</a></p>
</div>
<div class="section level3">
<h3 id="dev-notes">Dev notes<a class="anchor" aria-label="anchor" href="#dev-notes"></a>
</h3>
<p><code>FastReseg</code> package processes spatial transcriptome data through 5 different modules:</p>
<ul>
<li>
<p>Preprocess on whole dataset</p>
<ol style="list-style-type: decimal">
<li>
<code><a href="reference/runPreprocess.html">runPreprocess()</a></code> to get the baseline and cutoffs for transcript score and spatial local density from the whole dataset.</li>
</ol>
</li>
<li>
<p>Parallel processing on individual FOV: core wrapper <code><a href="reference/fastReseg_perFOV_full_process.html">fastReseg_perFOV_full_process()</a></code> contains all the 3 modules in this step.</p>
<ol start="2" style="list-style-type: decimal">
<li>
<code><a href="reference/runSegErrorEvaluation.html">runSegErrorEvaluation()</a></code> to detect cells with cell segmentation error based on spatial dependency of transcript score in a given transcript data.frame.</li>
<li>
<code><a href="reference/runTranscriptErrorDetection.html">runTranscriptErrorDetection()</a></code> to identify transcript groups with poor fit to current cell segments.</li>
<li>
<code><a href="reference/runSegRefinement.html">runSegRefinement()</a></code> to re-segment the low-fit transcript groups given their neighborhood.</li>
</ol>
</li>
<li>
<p>Combine outcomes from multiple FOVs into one</p>
<ol start="5" style="list-style-type: decimal">
<li>Pipeline wrappers would combine resegmentation outputs from individual FOVs into one.</li>
</ol>
</li>
</ul>
<p>For convenience, two pipeline wrapper functions are included for streamline processing of multi-FOV dataset to different exit points.</p>
<ul>
<li>
<code><a href="reference/fastReseg_flag_all_errors.html">fastReseg_flag_all_errors()</a></code>: performs preprocess and then evaluates and flags segmentation error in all input files, optional to return the gene expression matrix where all putative contaminating transcripts are trimmed from current cell segments.</li>
<li>
<code><a href="reference/fastReseg_full_pipeline.html">fastReseg_full_pipeline()</a></code>: performs preprocess, detect and correct cell segmentation errors by trimming, splitting and merging given the local neighborhood of poor-fit transcript groups, can process multiple input files in parallel.</li>
</ul>
</div>
<div class="section level3">
<h3 id="system-requirements">System requirements<a class="anchor" aria-label="anchor" href="#system-requirements"></a>
</h3>
<ul>
<li>R (&gt;= 3.5.0)</li>
<li>UNIX, Mac or Windows</li>
<li>see DESCRIPTION for full dependencies</li>
</ul>
</div>
<div class="section level3">
<h3 id="demo">Demo<a class="anchor" aria-label="anchor" href="#demo"></a>
</h3>
<p>See the “vignettes” folder.</p>
<ul>
<li>
<code>tutorial.Rmd</code> and <code>tutorial.html</code> for example usages of streamline pipeline wrappers and modular functions for individual task.</li>
<li>
<code>a__flagErrorOnly_on_SMIobject.R</code> for flagging segmentation errors without correcting, interfacing <code>FastReseg</code> with SMI TAP pipeline (<code>Giotto</code>).</li>
<li>
<code>b__fastReseg_on_SMIobject.R</code> for runing entire resegmentation workfkow on a given dataset, example dataset, interfacing <code>FastReseg</code> with SMI TAP pipeline (<code>Giotto</code>).</li>
</ul>
</div>
<div class="section level3">
<h3 id="workflow">Workflow:<a class="anchor" aria-label="anchor" href="#workflow"></a>
</h3>
<div class="float">
<img src="reference/figures/README-FastReseg_diagram.png" alt="Schematic of FastReseg workflow"><div class="figcaption">Schematic of FastReseg workflow</div>
</div>
</div>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<div class="section level4">
<h4 id="install-the-development-version-from-github">Install the development version from GitHub<a class="anchor" aria-label="anchor" href="#install-the-development-version-from-github"></a>
</h4>
<pre><code><span><span class="co"># to use with R 4.3.x</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"GiottoUtils"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"drieslab/GiottoClass"</span>, upgrade<span class="op">=</span><span class="st">"never"</span>,</span>
<span>  ref <span class="op">=</span> <span class="st">"bec3c79398a9aff299a01cc152ff6f2245a8f965"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"GiottoClass"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"drieslab/GiottoClass"</span>, upgrade<span class="op">=</span><span class="st">"never"</span>,</span>
<span>  ref <span class="op">=</span> <span class="st">"6d9d385beebcc57b78d34ffbe30be1ef0a184681"</span><span class="op">)</span></span></code></pre>
<pre><code><span><span class="co"># install FastReseg from github</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"Nanostring-Biostats/FastReseg"</span>,</span>
<span>                         build_vignettes <span class="op">=</span> <span class="cn">TRUE</span>, ref <span class="op">=</span> <span class="st">"main"</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/Nanostring-Biostats/FastReseg/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/Nanostring-Biostats/FastReseg/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing FastReseg</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Patrick Danaher <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-2844-5883" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Lidan Wu <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-3150-6170" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Patrick Danaher, Lidan Wu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
