<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>NA • FastReseg</title><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="NA"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-#E6F3FC" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">FastReseg</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="external-link dropdown-item" href="https://www.biorxiv.org/content/10.1101/2024.12.05.627051v1.abstract">Working principles and Manuscript</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Workflow and core functions</h6></li>
    <li><a class="dropdown-item" href="articles/tutorial.html">FastReseg to detect and correct segmentation error in spatial transcriptome data</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Nanostring-Biostats/FastReseg/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-title-body">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>NA</h1>
      <small class="dont-index">Source: <a href="https://github.com/Nanostring-Biostats/FastReseg/blob/HEAD/specs.md" class="external-link"><code>specs.md</code></a></small>
    </div>

<div class="section level2">
<h2 id="package-specifications">Package Specifications<a class="anchor" aria-label="anchor" href="#package-specifications"></a></h2>
<div class="section level4">
<h4 id="specs-for-fastreseg_flag_all_errors">Specs for fastReseg_flag_all_errors:<a class="anchor" aria-label="anchor" href="#specs-for-fastreseg_flag_all_errors"></a></h4>
<ul><li>Returns a genes x clusters matrix of cluster-specific reference profiles used in resegmenation pipeline. – test: test_pipeline_wrappers.R#L54</li>
<li>Returns a list of two matrices in cluster x percentile format for the cluster-specific percentile distribution of per cell value; <code>span_score</code> is for the average per molecule transcript tLLR score of each cell, <code>span_transNum</code> is for the transcript number of each cell. – test: test_pipeline_wrappers.R#L55 – test: test_pipeline_wrappers.R#L68</li>
<li>Returns a <code>data.frame</code> for spatial modeling statistics of each cell for all cells in the data set. – test: test_pipeline_wrappers.R#L56</li>
<li>Returns a list with each element to be a vector of <code>UMI_cellID</code> for cells flagged for potential cell segmentation errors within each FOV. – test: test_pipeline_wrappers.R#L57 – test: test_pipeline_wrappers.R#L70</li>
<li>If run with <code>return_trimmed_perCell = TRUE</code>,
<ul><li>Returns a gene x cell count sparse matrix where all putative contaminating transcripts are trimmed. – test: test_pipeline_wrappers.R#L58 – test: test_pipeline_wrappers.R#L65</li>
</ul></li>
<li>If run with <code>transDF_export_option =1</code>,
<ul><li>Saves each of the per FOV outputs as individual files in <code>path_to_output</code> directory; <code>flagged_transDF</code>, <code>modStats_ToFlagCells</code> and <code>classDF_ToFlagTrans</code> would be saved as csv file, respectively. – test: test_pipeline_wrappers.R#L72-79</li>
<li>The saved <code>flagged_transDF</code> csv file should be a transcript <code>data.frame</code> with columns for unique IDs of transcripts <code>UMI_transID</code> and cells <code>UMI_cellID</code>, for global coordinate system <code>x</code>, <code>y</code>, <code>z</code>, and for the goodness-of-fit in original cell segment <code>SMI_class</code>. – test: test_pipeline_wrappers.R#L82</li>
</ul></li>
</ul></div>
<div class="section level4">
<h4 id="specs-for-fastreseg_full_pipeline">Specs for fastReseg_full_pipeline:<a class="anchor" aria-label="anchor" href="#specs-for-fastreseg_full_pipeline"></a></h4>
<ul><li>Returns a genes x clusters matrix of cluster-specific reference profiles used in resegmenation pipeline – test: test_pipeline_wrappers.R#L129</li>
<li>Returns a list of two matrices in cluster x percentile format for the cluster-specific percentile distribution of per cell value; <code>span_score</code> is for the average per molecule transcript tLLR score of each cell, <code>span_transNum</code> is for the transcript number of each cell. – test: test_pipeline_wrappers.R#L130 – test: test_pipeline_wrappers.R#L145</li>
<li>Returns a list of cutoffs used in resegmentation pipeline, including, <code>score_baseline</code>, <code>lowerCutoff_transNum</code>, <code>higherCutoff_transNum</code>, <code>cellular_distance_cutoff</code>, <code>molecular_distance_cutoff</code>. – test: test_pipeline_wrappers.R#L131 – test: test_pipeline_wrappers.R#L146</li>
<li>If run with <code>return_perCellData = TRUE</code>,
<ul><li>Returns a per cell data.table with mean spatial coordinates, new cell type and resegmentation action after resegmentation. – test: test_pipeline_wrappers.R#L133 – test: test_pipeline_wrappers.R#L141</li>
<li>Returns a gene x cell count sparse matrix for updated transcript data.frame after resegmentation. – test: test_pipeline_wrappers.R#L134 – test: test_pipeline_wrappers.R#L142</li>
</ul></li>
<li>If run with <code>save_intermediates = TRUE</code>,
<ul><li>Returns a list of 4 elements describing how the resegmenation would be performed on original transcript <code>data.frame</code>. – test: test_pipeline_wrappers.R#L132 – test: test_pipeline_wrappers.R#L147</li>
<li>Saves all intermediate files and resegmenation outputs of each FOV as single <code>.rds</code> object per FOV. – test: test_pipeline_wrappers.R#L154-157</li>
</ul></li>
<li>If run with <code>save_intermediates = TRUE</code> and <code>transDF_export_option =1</code>,
<ul><li>Saves each of the per FOV outputs as individual files with columns for resegmented outcomes, e.g. <code>updated_cellID</code> and <code>updated_celltype</code>. – test: test_pipeline_wrappers.R#L155 – test: test_pipeline_wrappers.R#L160</li>
</ul></li>
</ul></div>
<div class="section level4">
<h4 id="specs-for-runpreprocess">Specs for runPreprocess:<a class="anchor" aria-label="anchor" href="#specs-for-runpreprocess"></a></h4>
<ul><li>Runs without error when using <code>counts</code>, <code>refProfiles</code> as input but no <code>clust</code>. – test: test_runPreprocess.R#L22-49</li>
<li>Runs without error when using <code>counts</code>, <code>clust</code> as input but no <code>refProfiles</code>. – test: test_runPreprocess.R#L86-113</li>
<li>Runs without error when using transcript <code>data.frame</code> directly as input along with <code>counts</code>, <code>refProfiles</code> but no <code>clust</code>. – test: test_runPreprocess.R#L151-178</li>
<li>For all input combinations,
<ul><li>Returns a vector of cluster assignments for each cell. – test: test_runPreprocess.R#L54 – test: test_runPreprocess.R#L123 – test: test_runPreprocess.R#L190</li>
<li>Returns a genes x clusters matrix of cluster-specific reference profiles used in resegmenation pipeline. – test: test_runPreprocess.R#L55 – test: test_runPreprocess.R#L124 – test: test_runPreprocess.R#L144 – test: test_runPreprocess.R#L191</li>
<li>Returns a list of two matrices in cluster x percentile format for the cluster-specific percentile distribution of per cell value; <code>span_score</code> is for the average per molecule transcript tLLR score of each cell, <code>span_transNum</code> is for the transcript number of each cell. – test: test_runPreprocess.R#L56 – test: test_runPreprocess.R#L71 – test: test_runPreprocess.R#L125 – test: test_runPreprocess.R#L140 – test: test_runPreprocess.R#L192</li>
<li>Returns a list of cutoffs used in resegmentation pipeline, including, <code>score_baseline</code>, <code>lowerCutoff_transNum</code>, <code>higherCutoff_transNum</code>, <code>cellular_distance_cutoff</code>, <code>molecular_distance_cutoff</code>. – test: test_runPreprocess.R#L57 – test: test_runPreprocess.R#L72 – test: test_runPreprocess.R#L126 – test: test_runPreprocess.R#L141 – test: test_runPreprocess.R#L193-194 – test: test_runPreprocess.R#L208-209</li>
<li>Returns a gene x cell-type score matrix to use in resegmenation pipeline. – test: test_runPreprocess.R#L58 – test: test_runPreprocess.R#L127</li>
<li>Returns a list of 2 elements for the intracellular and extracellular transcript <code>data.frame</code> of the processed outcomes of 1st transcrip file. – test: test_runPreprocess.R#L75-83 – test: test_runPreprocess.R#L196-203</li>
</ul></li>
</ul></div>
<div class="section level4">
<h4 id="specs-for-fastreseg_perfov_full_process">Specs for fastReseg_perFOV_full_process:<a class="anchor" aria-label="anchor" href="#specs-for-fastreseg_perfov_full_process"></a></h4>
<ul><li>Returns a updated transcript <code>data.farme</code> with <code>updated_cellID</code> and <code>updated_celltype</code> columns. – test: test_perFOV_processing.R#L57 – test: test_perFOV_processing.R#L71</li>
<li>If run with <code>return_perCellData = TRUE</code>,
<ul><li>Returns a per cell <code>data.table</code> with mean spatial coordinates, new cell type and resegmentation action after resegmentation. – test: test_perFOV_processing.R#L58 – test: test_perFOV_processing.R#L67</li>
<li>Returns a gene x cell count sparse matrix for the updated transcript data.frame after resegmentation. – test: test_perFOV_processing.R#L59 – test: test_perFOV_processing.R#L68</li>
</ul></li>
<li>If run with <code>return_intermediates = TRUE</code>,
<ul><li>Returns a <code>data.frame</code> for spatial modeling statistics of each cell. – test: test_perFOV_processing.R#L53 – test: test_perFOV_processing.R#L72</li>
<li>Returns a <code>data.frame</code> for the group assignment of transcripts within putative wrongly segmented cells. – test: test_perFOV_processing.R#L54 – test: test_perFOV_processing.R#L73</li>
<li>Returns a <code>data.frame</code> for neighborhood environment of low-score transcript groups. – test: test_perFOV_processing.R#L55 – test: test_perFOV_processing.R#L74</li>
<li>Returns a list of 4 elements describing how the resegmenation would be performed on original transcript <code>data.frame</code>. – test: test_perFOV_processing.R#L56 – test: test_perFOV_processing.R#L75</li>
</ul></li>
</ul></div>
<div class="section level4">
<h4 id="specs-for-runsegerrorevaluation">Specs for runSegErrorEvaluation:<a class="anchor" aria-label="anchor" href="#specs-for-runsegerrorevaluation"></a></h4>
<ul><li>Returns a <code>data.frame</code> contains evaluation model statistics in columns for each cell’s potential to have segmentation error. – test: test_perFOV_processing.R#L102-103 – test: test_perFOV_processing.R#L109</li>
<li>Returns a transcript <code>data.frame</code> with 2 additional columns: <code>tLLR_maxCellType</code> for cell types of maxmium transcript score under current segments and <code>score_tLLR_maxCellType</code> for the corresponding transcript score for each transcript. – test: test_perFOV_processing.R#L102-106</li>
</ul></div>
<div class="section level4">
<h4 id="specs-for-runtranscripterrordetection">Specs for runTranscriptErrorDetection:<a class="anchor" aria-label="anchor" href="#specs-for-runtranscripterrordetection"></a></h4>
<ul><li>Runs without error when using <code>dbscan</code> as transcript grouping method. – test: test_perFOV_processing.R#L115-136</li>
<li>Runs without error when using <code>delaunay</code> as transcript grouping method. – test: test_perFOV_processing.R#L140-163</li>
<li>For both transcript grouping methods,</li>
<li>Returns a transcript <code>data.frame</code> containing information for transcript score classifications and spatial group assignments as well as new cell/group ID for downstream resegmentation. – test: test_perFOV_processing.R#L129-136 – test: test_perFOV_processing.R#L156-163</li>
</ul></div>
<div class="section level4">
<h4 id="specs-for-runsegrefinement">Specs for runSegRefinement:<a class="anchor" aria-label="anchor" href="#specs-for-runsegrefinement"></a></h4>
<ul><li>Returns the updated transcript data.frame after resegmentation with <code>updated_cellID</code> and <code>updated_celltype</code> columns. – test: test_perFOV_processing.R#L202 – test: test_perFOV_processing.R#L214 – test: test_perFOV_processing.R#L226</li>
<li>If run with <code>return_intermediates = TRUE</code>,
<ul><li>Returns a <code>data.frame</code> for neighborhood environment of low-score transcript groups. – test: test_perFOV_processing.R#L200 – test: test_perFOV_processing.R#L215 – test: test_perFOV_processing.R#L222</li>
<li>Returns a list of 4 elements describing how the resegmenation would be performed on original transcript <code>data.frame</code>. – test: test_perFOV_processing.R#L201 – test: test_perFOV_processing.R#L216 – test: test_perFOV_processing.R#L223</li>
</ul></li>
<li>If run with <code>return_perCellData = TRUE</code>,
<ul><li>Returns a per cell data.table with mean spatial coordinates, new cell type and resegmentation action after resegmentation. – test: test_perFOV_processing.R#L203 – test: test_perFOV_processing.R#L210 – test: test_perFOV_processing.R#L224</li>
<li>Returns a gene x cell count sparse matrix for updated transcript data.frame after resegmentation. – test: test_perFOV_processing.R#L204 – test: test_perFOV_processing.R#L211 – test: test_perFOV_processing.R#L219 – test: test_perFOV_processing.R#L225</li>
</ul></li>
</ul></div>
</div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Patrick Danaher, Lidan Wu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

