<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>checkTransFileInputsAndLoadFirst — checkTransFileInputsAndLoadFirst • FastReseg</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="checkTransFileInputsAndLoadFirst — checkTransFileInputsAndLoadFirst"><meta name="description" content="check input formats for transcript data.frame file list and load 1st fov"><meta property="og:description" content="check input formats for transcript data.frame file list and load 1st fov"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-#E6F3FC" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FastReseg</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="external-link dropdown-item" href="https://www.biorxiv.org/content/10.1101/2024.12.05.627051v1.abstract">Working principles and Manuscript</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Workflow and core functions</h6></li>
    <li><a class="dropdown-item" href="../articles/tutorial.html">FastReseg to detect and correct segmentation error in spatial transcriptome data</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Nanostring-Biostats/FastReseg/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>checkTransFileInputsAndLoadFirst</h1>
      <small class="dont-index">Source: <a href="https://github.com/Nanostring-Biostats/FastReseg/blob/HEAD/R/file_utilities.R" class="external-link"><code>R/file_utilities.R</code></a></small>
      <div class="d-none name"><code>checkTransFileInputsAndLoadFirst.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>check input formats for transcript data.frame file list and load 1st fov</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">checkTransFileInputsAndLoadFirst</span><span class="op">(</span></span>
<span>  transcript_df <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transDF_fileInfo <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filepath_coln <span class="op">=</span> <span class="st">"file_path"</span>,</span>
<span>  prefix_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"slide"</span>, <span class="st">"fov"</span><span class="op">)</span>,</span>
<span>  fovOffset_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"stage_X"</span>, <span class="st">"stage_Y"</span><span class="op">)</span>,</span>
<span>  pixel_size <span class="op">=</span> <span class="fl">0.18</span>,</span>
<span>  zstep_size <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  transID_coln <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transGene_coln <span class="op">=</span> <span class="st">"target"</span>,</span>
<span>  cellID_coln <span class="op">=</span> <span class="st">"CellId"</span>,</span>
<span>  spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span>, <span class="st">"z"</span><span class="op">)</span>,</span>
<span>  invert_y <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  extracellular_cellID <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-transcript-df">transcript_df<a class="anchor" aria-label="anchor" href="#arg-transcript-df"></a></dt>
<dd><p>the data.frame of transcript level information with unique CellId, set to NULL if read from the <code>transDF_fileInfo</code></p></dd>


<dt id="arg-transdf-fileinfo">transDF_fileInfo<a class="anchor" aria-label="anchor" href="#arg-transdf-fileinfo"></a></dt>
<dd><p>a data.frame with each row for each individual file of per FOV transcript data.frame within which the coordinates and CellId are unique, columns include the file path of per FOV transcript data.frame file, annotation columns like slide and fov to be used as prefix when creating unique cell_ID across entire data set; when NULL, use the provided <code>transcript_df</code> directly</p></dd>


<dt id="arg-filepath-coln">filepath_coln<a class="anchor" aria-label="anchor" href="#arg-filepath-coln"></a></dt>
<dd><p>the column name of each individual file of per FOV transcript data.frame in <code>transDF_fileInfo</code></p></dd>


<dt id="arg-prefix-colns">prefix_colns<a class="anchor" aria-label="anchor" href="#arg-prefix-colns"></a></dt>
<dd><p>the column names of annotation in <code>transDF_fileInfo</code>, to be added to the CellId as prefix when creating unique cell_ID for entire data set; set to NULL if use the original <code>transID_coln</code> or <code>cellID_coln</code></p></dd>


<dt id="arg-fovoffset-colns">fovOffset_colns<a class="anchor" aria-label="anchor" href="#arg-fovoffset-colns"></a></dt>
<dd><p>the column name of coordinate offsets in 1st and 2nd dimension for each per FOV transcript data.frame in <code>transDF_fileInfo</code>, unit in micron
Notice that some assays like SMI has XY axes swapped between stage and each FOV such that <code>fovOffset_colns</code> should be c("stage_Y", "stage_X").</p></dd>


<dt id="arg-pixel-size">pixel_size<a class="anchor" aria-label="anchor" href="#arg-pixel-size"></a></dt>
<dd><p>the micrometer size of image pixel listed in 1st and 2nd dimension of <code>spatLocs_colns</code> of each <code>transcript_df</code></p></dd>


<dt id="arg-zstep-size">zstep_size<a class="anchor" aria-label="anchor" href="#arg-zstep-size"></a></dt>
<dd><p>the micrometer size of z-step for the optional 3rd dimension of <code>spatLocs_colns</code> of each <code>transcript_df</code></p></dd>


<dt id="arg-transid-coln">transID_coln<a class="anchor" aria-label="anchor" href="#arg-transid-coln"></a></dt>
<dd><p>the column name of transcript_ID in <code>transcript_df</code>, default = NULL to use row index of transcript in each <code>transcript_df</code>; when <code>prefix_colns</code> != NULL, unique transcript_id would be generated from <code>prefix_colns</code> and <code>transID_coln</code> in each <code>transcript_df</code></p></dd>


<dt id="arg-transgene-coln">transGene_coln<a class="anchor" aria-label="anchor" href="#arg-transgene-coln"></a></dt>
<dd><p>the column name of target or gene name in <code>transcript_df</code></p></dd>


<dt id="arg-cellid-coln">cellID_coln<a class="anchor" aria-label="anchor" href="#arg-cellid-coln"></a></dt>
<dd><p>the column name of cell_ID in <code>transcript_df</code>; when <code>prefix_colns</code> != NULL, unique cell_ID would be generated from <code>prefix_colns</code> and <code>cellID_coln</code> in each <code>transcript_df</code></p></dd>


<dt id="arg-spatlocs-colns">spatLocs_colns<a class="anchor" aria-label="anchor" href="#arg-spatlocs-colns"></a></dt>
<dd><p>column names for 1st, 2nd and optional 3rd dimension of spatial coordinates in <code>transcript_df</code></p></dd>


<dt id="arg-invert-y">invert_y<a class="anchor" aria-label="anchor" href="#arg-invert-y"></a></dt>
<dd><p>flag to invert y axis of local coordinates during stitching (default = TRUE)</p></dd>


<dt id="arg-extracellular-cellid">extracellular_cellID<a class="anchor" aria-label="anchor" href="#arg-extracellular-cellid"></a></dt>
<dd><p>a vector of cell_ID for extracellular transcripts which would be removed from the resegmention pipeline (default = NULL)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a list contains transcript_df for downstream process and extracellular transcript data.frame
'</p><dl><dt>intraC</dt>
<dd><p>a data.frame for intracellular transcript, <code>UMI_transID</code> and <code>UMI_cellID</code> as column names for unique transcript_id and cell_id, <code>target</code> as column name for target gene name</p></dd>

<dt>extraC</dt>
<dd><p>a data.frame for extracellular transcript, same structure as the <code>intraC</code> data.frame in returned list</p></dd>


</dl></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Patrick Danaher, Lidan Wu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

