<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>per cell gene expression matrix based on raw transcript data files generated by SMI — example_CellGeneExpr • FastReseg</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="per cell gene expression matrix based on raw transcript data files generated by SMI — example_CellGeneExpr"><meta name="description" content="The example dataset is based on the cropped regions from two different FOVs of same SMI RNA readout run for FFPE melanoma tissue sample.
The corresponding raw transcript data files are
extdata/Run4104_FOV001__complete_code_cell_target_call_coord.csv
extdata/Run4104_FOV002__complete_code_cell_target_call_coord.csv
Each raw transcript data files have 1 transcript per row with 25 different meta information in columns. Some important columns are
x
spatial coordinate of the transcript in x-axis of the given FOV, unit in pixel, 0.18um per pixel

y
spatial coordinate of the transcript in y-axis of the given FOV, unit in pixel, 0.18um per pixel

z
spatial coordinate of the transcript in z-axis of the given FOV, unit in z-step, 0.8um per z-step

target
gene identity of given transcript

CellId
the cell label assignment of given transcript within given FOV based on original cell segmentation


The per cell gene expression matrix is calculated from the raw transcript data files to include only the 960 true target genes without including control probes."><meta property="og:description" content="The example dataset is based on the cropped regions from two different FOVs of same SMI RNA readout run for FFPE melanoma tissue sample.
The corresponding raw transcript data files are
extdata/Run4104_FOV001__complete_code_cell_target_call_coord.csv
extdata/Run4104_FOV002__complete_code_cell_target_call_coord.csv
Each raw transcript data files have 1 transcript per row with 25 different meta information in columns. Some important columns are
x
spatial coordinate of the transcript in x-axis of the given FOV, unit in pixel, 0.18um per pixel

y
spatial coordinate of the transcript in y-axis of the given FOV, unit in pixel, 0.18um per pixel

z
spatial coordinate of the transcript in z-axis of the given FOV, unit in z-step, 0.8um per z-step

target
gene identity of given transcript

CellId
the cell label assignment of given transcript within given FOV based on original cell segmentation


The per cell gene expression matrix is calculated from the raw transcript data files to include only the 960 true target genes without including control probes."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-#E6F3FC" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FastReseg</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="external-link dropdown-item" href="https://www.biorxiv.org/content/10.1101/2024.12.05.627051v1.abstract">Working principles and Manuscript</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Workflow and core functions</h6></li>
    <li><a class="dropdown-item" href="../articles/tutorial.html">FastReseg to detect and correct segmentation error in spatial transcriptome data</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Nanostring-Biostats/FastReseg/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>per cell gene expression matrix based on raw transcript data files generated by SMI</h1>
      <small class="dont-index">Source: <a href="https://github.com/Nanostring-Biostats/FastReseg/blob/HEAD/R/data.R" class="external-link"><code>R/data.R</code></a></small>
      <div class="d-none name"><code>example_CellGeneExpr.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The example dataset is based on the cropped regions from two different FOVs of same SMI RNA readout run for FFPE melanoma tissue sample.
The corresponding raw transcript data files are</p><ul><li><p><code>extdata/Run4104_FOV001__complete_code_cell_target_call_coord.csv</code></p></li>
<li><p><code>extdata/Run4104_FOV002__complete_code_cell_target_call_coord.csv</code></p></li>
</ul><p>Each raw transcript data files have 1 transcript per row with 25 different meta information in columns. Some important columns are</p><dl><dt>x</dt>
<dd><p>spatial coordinate of the transcript in x-axis of the given FOV, unit in pixel, 0.18um per pixel</p></dd>

<dt>y</dt>
<dd><p>spatial coordinate of the transcript in y-axis of the given FOV, unit in pixel, 0.18um per pixel</p></dd>

<dt>z</dt>
<dd><p>spatial coordinate of the transcript in z-axis of the given FOV, unit in z-step, 0.8um per z-step</p></dd>

<dt>target</dt>
<dd><p>gene identity of given transcript</p></dd>

<dt>CellId</dt>
<dd><p>the cell label assignment of given transcript within given FOV based on original cell segmentation</p></dd>


</dl><p>The per cell gene expression matrix is calculated from the raw transcript data files to include only the 960 true target genes without including control probes.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">example_CellGeneExpr</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="format">Format<a class="anchor" aria-label="anchor" href="#format"></a></h2>
    <p>An object of class <code>"matrix"</code> with 754 cells in row and 960 genes in columns</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><a href="example_clust.html">example_clust</a> for corresponding cell cluster assignment</p></div>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Patrick Danaher, Lidan Wu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

