<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>get_neighborhood_content — get_neighborhood_content • FastReseg</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="get_neighborhood_content — get_neighborhood_content"><meta name="description" content="find neighbor cells with transcripts that are direct neighbor of chosen_cell, check tLLRv2 score under neighbor cell type, return neighborhood information"><meta property="og:description" content="find neighbor cells with transcripts that are direct neighbor of chosen_cell, check tLLRv2 score under neighbor cell type, return neighborhood information"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-#E6F3FC" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FastReseg</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="external-link dropdown-item" href="https://www.biorxiv.org/content/10.1101/2024.12.05.627051v1.abstract">Working principles and Manuscript</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Workflow and core functions</h6></li>
    <li><a class="dropdown-item" href="../articles/tutorial.html">FastReseg to detect and correct segmentation error in spatial transcriptome data</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Nanostring-Biostats/FastReseg/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>get_neighborhood_content</h1>
      <small class="dont-index">Source: <a href="https://github.com/Nanostring-Biostats/FastReseg/blob/HEAD/R/get_neighborhood_content.R" class="external-link"><code>R/get_neighborhood_content.R</code></a></small>
      <div class="d-none name"><code>get_neighborhood_content.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>find neighbor cells with transcripts that are direct neighbor of chosen_cell, check tLLRv2 score under neighbor cell type, return neighborhood information</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_neighborhood_content</span><span class="op">(</span></span>
<span>  chosen_cells <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">score_GeneMatrix</span>,</span>
<span>  score_baseline <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  neighbor_distance_xy <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  distance_cutoff <span class="op">=</span> <span class="fl">2.7</span>,</span>
<span>  <span class="va">transcript_df</span>,</span>
<span>  cellID_coln <span class="op">=</span> <span class="st">"CellId"</span>,</span>
<span>  celltype_coln <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>  transID_coln <span class="op">=</span> <span class="st">"transcript_id"</span>,</span>
<span>  transGene_coln <span class="op">=</span> <span class="st">"target"</span>,</span>
<span>  transSpatLocs_coln <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span>, <span class="st">"z"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-chosen-cells">chosen_cells<a class="anchor" aria-label="anchor" href="#arg-chosen-cells"></a></dt>
<dd><p>the cell_ID of chosen cells need to be evaluate for re-segmentation</p></dd>


<dt id="arg-score-genematrix">score_GeneMatrix<a class="anchor" aria-label="anchor" href="#arg-score-genematrix"></a></dt>
<dd><p>the gene x cell-type matrix of log-like score of gene in each cell type</p></dd>


<dt id="arg-score-baseline">score_baseline<a class="anchor" aria-label="anchor" href="#arg-score-baseline"></a></dt>
<dd><p>a named vector of score baseline for all cell type listed in score_GeneMatrix</p></dd>


<dt id="arg-neighbor-distance-xy">neighbor_distance_xy<a class="anchor" aria-label="anchor" href="#arg-neighbor-distance-xy"></a></dt>
<dd><p>maximum cell-to-cell distance in x, y between the center of query cells to the center of neighbor cells with direct contact, same unit as input spatial coordinate. Default = NULL to use the 2 times of average 2D cell diameter.</p></dd>


<dt id="arg-distance-cutoff">distance_cutoff<a class="anchor" aria-label="anchor" href="#arg-distance-cutoff"></a></dt>
<dd><p>maximum molecule-to-molecule distance within connected transcript group, same unit as input spatial coordinate (default = 2.7 micron).
If set to NULL, the pipeline would first randomly choose no more than 2500 cells from up to 10 random picked ROIs with search radius to be 5 times of <code>neighbor_distance_xy</code>, and then calculate the minimal molecular distance between picked cells. The pipeline would further use the 5 times of 90% quantile of minimal molecular distance as <code>distance_cutoff</code>. This calculation is slow and is not recommended for large transcript data.frame.</p></dd>


<dt id="arg-transcript-df">transcript_df<a class="anchor" aria-label="anchor" href="#arg-transcript-df"></a></dt>
<dd><p>the data.frame with transcript_id, target/geneName, x, y and cell_id</p></dd>


<dt id="arg-cellid-coln">cellID_coln<a class="anchor" aria-label="anchor" href="#arg-cellid-coln"></a></dt>
<dd><p>the column name of cell_ID in transcript_df</p></dd>


<dt id="arg-celltype-coln">celltype_coln<a class="anchor" aria-label="anchor" href="#arg-celltype-coln"></a></dt>
<dd><p>the column name of cell_type in transcript_df</p></dd>


<dt id="arg-transid-coln">transID_coln<a class="anchor" aria-label="anchor" href="#arg-transid-coln"></a></dt>
<dd><p>the column name of transcript_ID in transcript_df</p></dd>


<dt id="arg-transgene-coln">transGene_coln<a class="anchor" aria-label="anchor" href="#arg-transgene-coln"></a></dt>
<dd><p>the column name of target or gene name in transcript_df</p></dd>


<dt id="arg-transspatlocs-coln">transSpatLocs_coln<a class="anchor" aria-label="anchor" href="#arg-transspatlocs-coln"></a></dt>
<dd><p>the column name of 1st, 2nd, optional 3rd spatial dimension of each transcript in transcript_df</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a data.frame
#'</p><ol><li><p>CellId, original cell id of chosen cells</p></li>
<li><p>cell_type, original cell type of chosen cells</p></li>
<li><p>transcript_num, number of transcripts in chosen cells</p></li>
<li><p>self_celltype, cell type give maximum score for query cell only</p></li>
<li><p>score_under_self, score in query cell under its own maximum celltype</p></li>
<li><p>neighbor_CellId, cell id of neighbor cell whose cell type gives maximum score in query cell among all neighbors, not including query cell itself</p></li>
<li><p>neighbor_celltype, cell type that gives maximum score in query cell among all non-self neighbor cells</p></li>
<li><p>score_under_neighbor, score in query cell under neighbor_celltype</p></li>
</ol></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Locate neighbor cells of each query cell firstly via cell-to-cell distance in 2D plane within neighbor_distance_xy, then via molecule-to-molecule 3D distance within distance_cutoff. If no neighbor cells found for query cell, use the cell id and cell type of query cell to fill in the columns for neighbor cells in returned data.frame</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Patrick Danaher, Lidan Wu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

