<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>FastReseg to detect and correct segmentation error in spatial transcriptome data â€¢ FastReseg</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="FastReseg to detect and correct segmentation error in spatial transcriptome data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-#E6F3FC" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FastReseg</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="external-link dropdown-item" href="https://www.nature.com/articles/s41598-025-08733-5">Working principles and Manuscript</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Workflow and core functions</h6></li>
    <li><a class="dropdown-item" href="../articles/tutorial.html">FastReseg to detect and correct segmentation error in spatial transcriptome data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Nanostring-Biostats/FastReseg/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<script src="tutorial_files/kePrint-0.0.1/kePrint.js"></script><link href="tutorial_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>FastReseg to detect and correct segmentation error in spatial transcriptome data</h1>
                        <h4 data-toc-skip class="author">Lidan Wu</h4>
            
            <h4 data-toc-skip class="date">2025-08-21</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/Nanostring-Biostats/FastReseg/blob/HEAD/vignettes/tutorial.Rmd" class="external-link"><code>vignettes/tutorial.Rmd</code></a></small>
      <div class="d-none name"><code>tutorial.Rmd</code></div>
    </div>

    
    
<p>This tutorial below shows how to apply functions in
<code>FastReseg</code> to perform segmentation error detection and
correction in spatial transcriptome data.</p>
<div class="section level2">
<h2 id="fastreseg-workflow-diagram">FastReseg workflow diagram<a class="anchor" aria-label="anchor" href="#fastreseg-workflow-diagram"></a>
</h2>
<p><code>FastReseg</code> package processes spatial transcriptome data
through 5 different modules and provides 3 wrapper functions for
streamline processing of multi-FOV dataset to different exit points.</p>
<p><img src="README-FastReseg_diagram.png" width="100%"></p>
</div>
<div class="section level2">
<h2 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h2>
<p>The required inputs include:</p>
<ul>
<li><p><code>counts</code>: a cell-by-gene counts matrix for entire
dataset.</p></li>
<li><p><code>clust</code>: a vector of cluster assignments for each cell
in <code>counts</code>; use <code>NULL</code> to automatically assign
the cell cluster for each cell based on maximum transcript score of
given the provided <code>refProfiles</code>.</p></li>
<li><p><code>refProfiles</code>: a gene-by-cluster matrix of
cluster-specific expression profiles; default = <code>NULL</code> to use
external cluster assignments.</p></li>
<li>
<p><code>transDF_fileInfo</code>: a data.frame with each row for
each individual file of per-FOV transcript data.frame, columns include
the file path of per FOV transcript data.frame file, annotation columns
like <code>slide</code> and <code>fov</code> to be used as prefix when
creating unique cell_ID across entire dataset.</p>
<ul>
<li>when <code>NULL</code>, use the transcript data.frame
<code>transcript_df</code> directly.</li>
</ul>
</li>
</ul>
<p>There must be at least one of the <code>clust</code> and
<code>refProfiles</code> provided to run the <code>FastReseg</code>
pipeline. All the spatial coordinates and distance are in the units of
micron for consistency. Please refer to example data coming with the
package to see how it looks like.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load example input data from package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://nanostring-biostats.github.io/FastReseg/">FastReseg</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get cell-by-gene `counts`</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_CellGeneExpr"</span><span class="op">)</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">example_CellGeneExpr</span></span>
<span></span>
<span><span class="co"># get cluster assignment `clust`</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_clust"</span><span class="op">)</span></span>
<span><span class="va">clust</span> <span class="op">&lt;-</span> <span class="va">example_clust</span></span>
<span></span>
<span><span class="co"># get cluster-specific reference profiles `refProfiles`</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_refProfiles"</span><span class="op">)</span></span>
<span><span class="va">refProfiles</span> <span class="op">&lt;-</span> <span class="va">example_refProfiles</span></span>
<span></span>
<span><span class="co"># create `transDF_fileInfo` for multiple per FOV transcript data.frame </span></span>
<span><span class="co"># coordinates for each FOV, `stage_x` and `stage_y`, should have units in micron.</span></span>
<span><span class="va">dataDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"FastReseg"</span><span class="op">)</span></span>
<span><span class="va">transDF_fileInfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>file_path <span class="op">=</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="va">dataDir</span>, </span>
<span>                                                    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Run4104_FOV001__complete_code_cell_target_call_coord.csv"</span>,</span>
<span>                                                      <span class="st">"Run4104_FOV002__complete_code_cell_target_call_coord.csv"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                               slide <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                               fov <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>,</span>
<span>                               stage_X <span class="op">=</span> <span class="fl">1000</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5.13</span>, <span class="op">-</span><span class="fl">2.701</span><span class="op">)</span>,</span>
<span>                               stage_Y <span class="op">=</span> <span class="fl">1000</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.452</span>, <span class="fl">0.081</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Here is an example of per-FOV transcript data.frame with necessary
columns for:</p>
<ul>
<li><p><code>target</code> for gene name;</p></li>
<li><p><code>x</code>, <code>y</code> and optional <code>z</code> for
spatial coordinates of each transcript;</p></li>
<li>
<p><code>UMI_cellID</code> for cell ids of current cell segmentation
and must be unique across all FOVs of the same dataset;</p>
<ul>
<li>When not available, <code><a href="../reference/prepare_perFOV_transDF.html">prepare_perFOV_transDF()</a></code> function
would use per-FOV unique <code>CellId</code> and the provided
<code>prefix_colns = c('slide', 'fov')</code> to generate
<code>UMI_cellID</code> that would be unique across the entire dataset.
Pipeline wrappers, <code><a href="../reference/fastReseg_flag_all_errors.html">fastReseg_flag_all_errors()</a></code> and
<code><a href="../reference/fastReseg_full_pipeline.html">fastReseg_full_pipeline()</a></code>, would also generate unique cell
ids in same manner for the multiple files listed in
<code>transDF_fileInfo</code>.</li>
</ul>
</li>
<li>
<p><code>UMI_transID</code>: transcript ids that are unique across
the dataset;</p>
<ul>
<li>By default, the <code><a href="../reference/prepare_perFOV_transDF.html">prepare_perFOV_transDF()</a></code> function and
the 2 pipeline wrappers would use row index of transcript in each
per-FOV transcript data.frame and <code>UMI_cellID</code> to create
<code>UMI_transID</code> that would be unique across the entire
dataset.</li>
</ul>
</li>
</ul>
<table class="table">
<colgroup>
<col width="17%">
<col width="15%">
<col width="13%">
<col width="13%">
<col width="5%">
<col width="10%">
<col width="8%">
<col width="5%">
<col width="9%">
</colgroup>
<thead><tr>
<th align="left">UMI_transID</th>
<th align="left">UMI_cellID</th>
<th align="right">x</th>
<th align="right">y</th>
<th align="right">z</th>
<th align="left">target</th>
<th align="right">slide</th>
<th align="right">fov</th>
<th align="right">CellId</th>
</tr></thead>
<tbody>
<tr>
<td align="left">t_1_2_10</td>
<td align="left">c_1_2_5</td>
<td align="right">914.8070</td>
<td align="right">-7832.786</td>
<td align="right">2.4</td>
<td align="left">HLA-DRA</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">5</td>
</tr>
<tr>
<td align="left">t_1_2_100005</td>
<td align="left">c_1_2_1270</td>
<td align="right">841.7000</td>
<td align="right">-8045.011</td>
<td align="right">2.4</td>
<td align="left">HLA-E</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">1270</td>
</tr>
<tr>
<td align="left">t_1_2_100006</td>
<td align="left">c_1_2_1270</td>
<td align="right">844.4609</td>
<td align="right">-8045.006</td>
<td align="right">2.4</td>
<td align="left">TWIST1</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">1270</td>
</tr>
<tr>
<td align="left">t_1_2_100007</td>
<td align="left">c_1_2_1286</td>
<td align="right">878.1635</td>
<td align="right">-8045.002</td>
<td align="right">2.4</td>
<td align="left">HLA-B</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">1286</td>
</tr>
<tr>
<td align="left">t_1_2_10001</td>
<td align="left">c_1_2_201</td>
<td align="right">1018.2337</td>
<td align="right">-7859.833</td>
<td align="right">2.4</td>
<td align="left">S100A6</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">201</td>
</tr>
<tr>
<td align="left">t_1_2_100015</td>
<td align="left">c_1_2_1292</td>
<td align="right">885.8180</td>
<td align="right">-8045.029</td>
<td align="right">2.4</td>
<td align="left">HBB</td>
<td align="right">1</td>
<td align="right">2</td>
<td align="right">1292</td>
</tr>
</tbody>
</table>
<p>As mentioned above, <code>FastReseg</code> package provides 2
pipeline wrappers for streamline multi-FOV processing and those wrapper
would go through each per-FOV transcript data.frame listed in
<code>transDF_fileInfo</code> and stitch all FOVs together to get data
with coordinates in the global system as well as transcript and cell IDs
unique across the entire dataset. When preparing the per-FOV data.frame
outside the pipeline wrappers, one can read in each per-FOV file and
process it with <code><a href="../reference/prepare_perFOV_transDF.html">prepare_perFOV_transDF()</a></code> function to get
the unique IDs for cells and transcripts as well as converting the local
coordinates in pixel for each FOV to a global coordinate in micron for
entire dataset.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#  process 1st file in the `transDF_fileInfo` entry </span></span>
<span><span class="va">idx</span> <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="va">rawDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">transDF_fileInfo</span><span class="op">[</span><span class="va">idx</span>, <span class="st">'file_path'</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">rawDF</span>, n <span class="op">=</span> <span class="fl">3L</span><span class="op">)</span></span>
<span><span class="co">#&gt;   fov seed_x seed_y         x        y z     std_x     std_y</span></span>
<span><span class="co">#&gt; 1   1  866.0 1250.5  865.9666 1250.467 1 0.1366284 0.1365927</span></span>
<span><span class="co">#&gt; 2   1 1026.8 1250.6 1026.7778 1250.656 1 0.1202107 0.1424117</span></span>
<span><span class="co">#&gt; 3   1 1008.0 1250.9 1008.0501 1250.933 1 0.1378334 0.1365927</span></span>
<span><span class="co">#&gt;   X95_perc_conf_int_x X95_perc_conf_int_y target target_idx Spot1_count</span></span>
<span><span class="co">#&gt; 1           0.2867654           0.2866904    AXL         70           1</span></span>
<span><span class="co">#&gt; 2           0.1848043           0.2189347   IL16        506           3</span></span>
<span><span class="co">#&gt; 3           0.2892946           0.2866905 S100A6        846           1</span></span>
<span><span class="co">#&gt;   Spot2_count Spot3_count Spot4_count target_call_observations</span></span>
<span><span class="co">#&gt; 1           2           2           1                        6</span></span>
<span><span class="co">#&gt; 2           2           3           1                        9</span></span>
<span><span class="co">#&gt; 3           2           2           1                        6</span></span>
<span><span class="co">#&gt;   target_count_per_feature random_call_probability possible_BC_count</span></span>
<span><span class="co">#&gt; 1                        1             0.002597012                 1</span></span>
<span><span class="co">#&gt; 2                        1             0.002597012                 1</span></span>
<span><span class="co">#&gt; 3                        1             0.010347653                 4</span></span>
<span><span class="co">#&gt;   spots_per_feature multicolor_spots_per_feature call_quality_score CellId</span></span>
<span><span class="co">#&gt; 1                 4                            0       0.0001604543   2932</span></span>
<span><span class="co">#&gt; 2                 4                            0       0.0019040769    864</span></span>
<span><span class="co">#&gt; 3                 5                            0       0.0001396453      0</span></span>
<span><span class="co">#&gt;   CellComp transcript_id</span></span>
<span><span class="co">#&gt; 1  Nuclear         11356</span></span>
<span><span class="co">#&gt; 2  Nuclear         11359</span></span>
<span><span class="co">#&gt; 3        0         11363</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">transcript_df_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepare_perFOV_transDF.html">prepare_perFOV_transDF</a></span><span class="op">(</span>each_transDF <span class="op">=</span> <span class="va">rawDF</span>, </span>
<span>                                            fov_centerLocs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">transDF_fileInfo</span><span class="op">[</span><span class="va">idx</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'stage_X'</span>, <span class="st">'stage_Y'</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                                            prefix_vals <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">transDF_fileInfo</span><span class="op">[</span><span class="va">idx</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'slide'</span>, <span class="st">'fov'</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>, </span>
<span>                                            pixel_size <span class="op">=</span> <span class="fl">0.12</span>, <span class="co"># micron per pixel </span></span>
<span>                                            zstep_size <span class="op">=</span> <span class="fl">0.8</span>, <span class="co"># micron per z step </span></span>
<span>                                            transID_coln <span class="op">=</span> <span class="cn">NULL</span>, <span class="co"># use row index </span></span>
<span>                                            transGene_coln <span class="op">=</span> <span class="st">'target'</span>, <span class="co"># gene name</span></span>
<span>                                            cellID_coln <span class="op">=</span> <span class="st">'CellId'</span>, <span class="co"># cell label unique at FOV level </span></span>
<span>                                            spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'x'</span>, <span class="st">'y'</span>, <span class="st">'z'</span><span class="op">)</span>, <span class="co"># column names for spatial coordinates in pixel for each FOV</span></span>
<span>                                            invert_y <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co"># flip Y axis (default = TRUE) between local image coordinate and global stage coordinate </span></span>
<span>                                            extracellular_cellID <span class="op">=</span> <span class="fl">0</span>, <span class="co"># set this to the cell ID for extracellular transcript, use NULL if your data only contains intracellular transcripts </span></span>
<span>                                            drop_original <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># set to FALSE if want to have columns for original cell ID and spatial coordinates returned in the data.frame</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">transcript_df_all</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 2</span></span>
<span><span class="co">#&gt;  $ intraC:'data.frame':  151712 obs. of  6 variables:</span></span>
<span><span class="co">#&gt;   ..$ UMI_cellID : chr [1:151712] "c_1_1_2932" "c_1_1_864" "c_1_1_2923" "c_1_1_864" ...</span></span>
<span><span class="co">#&gt;   ..$ UMI_transID: chr [1:151712] "t_1_1_1" "t_1_1_2" "t_1_1_5" "t_1_1_6" ...</span></span>
<span><span class="co">#&gt;   ..$ target     : chr [1:151712] "AXL" "IL16" "HSP90AA1" "MZT2A" ...</span></span>
<span><span class="co">#&gt;   ..$ x          : num [1:151712] 5234 5253 5196 5252 5192 ...</span></span>
<span><span class="co">#&gt;   ..$ y          : num [1:151712] -602 -602 -602 -602 -602 ...</span></span>
<span><span class="co">#&gt;   ..$ z          : num [1:151712] 0.8 0.8 0.8 0.8 0.8 0.8 0.8 0.8 0.8 0.8 ...</span></span>
<span><span class="co">#&gt;  $ extraC:'data.frame':  110792 obs. of  6 variables:</span></span>
<span><span class="co">#&gt;   ..$ UMI_cellID : chr [1:110792] "c_1_1_0" "c_1_1_0" "c_1_1_0" "c_1_1_0" ...</span></span>
<span><span class="co">#&gt;   ..$ UMI_transID: chr [1:110792] "t_1_1_3" "t_1_1_4" "t_1_1_8" "t_1_1_10" ...</span></span>
<span><span class="co">#&gt;   ..$ target     : chr [1:110792] "S100A6" "IFI27" "HLA-A" "SOD2" ...</span></span>
<span><span class="co">#&gt;   ..$ x          : num [1:110792] 5251 5192 5306 5251 5269 ...</span></span>
<span><span class="co">#&gt;   ..$ y          : num [1:110792] -602 -602 -602 -602 -603 ...</span></span>
<span><span class="co">#&gt;   ..$ z          : num [1:110792] 0.8 0.8 0.8 0.8 0.8 0.8 0.8 0.8 0.8 0.8 ...</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## we would focus on intracellular transcripts for downstream segmentation error detection</span></span>
<span><span class="va">transcript_df</span> <span class="op">&lt;-</span> <span class="va">transcript_df_all</span><span class="op">[[</span><span class="st">"intraC"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="pipeline-wrapper-functions-for-streamline-processing">Pipeline wrapper functions for streamline processing<a class="anchor" aria-label="anchor" href="#pipeline-wrapper-functions-for-streamline-processing"></a>
</h2>
<p>For streamline processing of big dataset with multiple FOVs, one can
use the pipeline wrapper functions provided in the
<code>FastReseg</code> package without going through the individual
steps that are discussed in later part of this tutorial. Those pipeline
wrappers would preprocess at whole dataset level first to get
appropriate cutoffs, and then perform segmentation evaluation and
optional correction on each FOV, followed by combining per FOV data into
one. Please refer to the manual of each pipeline wrapper for more
details and to section <a href="#processing-each-fov-outside-of-core-wrapper">Processing each FOV
outside of core wrapper</a> for excerpts of example outputs.</p>
<div class="section level3">
<h3 id="pipeline-wraper-for-cell-level-segmentation-detection-across-multi-fov-dataset">Pipeline wraper for cell-level segmentation detection across
multi-FOV dataset<a class="anchor" aria-label="anchor" href="#pipeline-wraper-for-cell-level-segmentation-detection-across-multi-fov-dataset"></a>
</h3>
<p>A common use case of <code>FastReseg</code> is to evaluate the
current cell segmentation of entire multi-FOV dataset and then identify
the transcripts with poor goodness-of-fit to current cell and thus
likely arsing from potential neighborhood contamination. And this could
be done with <code><a href="../reference/fastReseg_flag_all_errors.html">fastReseg_flag_all_errors()</a></code> function.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">flagAll_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fastReseg_flag_all_errors.html">fastReseg_flag_all_errors</a></span><span class="op">(</span></span>
<span>  counts <span class="op">=</span> <span class="va">counts</span>,</span>
<span>  clust <span class="op">=</span> <span class="va">clust</span>,</span>
<span>  refProfiles <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  </span>
<span>  <span class="co"># Similar to `runPreprocess()`, one can use `clust = NULL` if providing `refProfiles`</span></span>
<span>  </span>
<span>  transcript_df <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transDF_fileInfo <span class="op">=</span> <span class="va">transDF_fileInfo</span>,</span>
<span>  filepath_coln <span class="op">=</span> <span class="st">'file_path'</span>,</span>
<span>  prefix_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'slide'</span>,<span class="st">'fov'</span><span class="op">)</span>,</span>
<span>  fovOffset_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'stage_Y'</span>,<span class="st">'stage_X'</span><span class="op">)</span>, <span class="co"># match XY axes between stage and each FOV</span></span>
<span>  pixel_size <span class="op">=</span> <span class="fl">0.18</span>, </span>
<span>  zstep_size <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  transID_coln <span class="op">=</span> <span class="cn">NULL</span>, <span class="co"># row index as transcript_id</span></span>
<span>  transGene_coln <span class="op">=</span> <span class="st">"target"</span>,</span>
<span>  cellID_coln <span class="op">=</span> <span class="st">"CellId"</span>,</span>
<span>  spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>,<span class="st">"y"</span>,<span class="st">"z"</span><span class="op">)</span>,</span>
<span>  extracellular_cellID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>, </span>
<span>  </span>
<span>  flagCell_lrtest_cutoff <span class="op">=</span> <span class="fl">5</span>, <span class="co"># cutoff for flagging wrongly segmented cells</span></span>
<span>  svmClass_score_cutoff <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>, <span class="co"># cutoff for low vs. high transcript score</span></span>
<span>  path_to_output <span class="op">=</span> <span class="st">"res1f_multiFiles"</span>, <span class="co"># path to output folder</span></span>
<span>  return_trimmed_perCell <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co"># flag to return per cell expression matrix after trimming all flagged transcripts </span></span>
<span>  ctrl_genes <span class="op">=</span> <span class="cn">NULL</span> <span class="co"># name for control probes in transcript data.frame, e.g. negative control probes</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">flagAll_res</span><span class="op">)</span></span>
<span><span class="co">#&gt;  num [1:960, 1:14] 1.00e-04 3.28e-04 4.19e-04 1.35e-04 8.07e-05 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:960] "AATK" "ABL1" "ABL2" "ACE" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:14] "c" "e" "f" "b" ...</span></span>
<span><span class="co">#&gt;  $ baselineData                 :List of 2</span></span>
<span><span class="co">#&gt;   ..$ span_score   : num [1:14, 1:5] 0 -1.508 -1.842 -1.806 -0.451 ...</span></span>
<span><span class="co">#&gt;   .. ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. .. ..$ : chr [1:14] "a" "b" "c" "e" ...</span></span>
<span><span class="co">#&gt;   .. .. ..$ : chr [1:5] "0%" "25%" "50%" "75%" ...</span></span>
<span><span class="co">#&gt;   ..$ span_transNum: num [1:14, 1:5] 1 9 16 24 58 13 1 18 1 1 ...</span></span>
<span><span class="co">#&gt;   .. ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. .. ..$ : chr [1:14] "a" "b" "c" "e" ...</span></span>
<span><span class="co">#&gt;   .. .. ..$ : chr [1:5] "0%" "25%" "50%" "75%" ...</span></span>
<span><span class="co">#&gt;  $ combined_modStats_ToFlagCells:'data.frame':   724 obs. of  9 variables:</span></span>
<span><span class="co">#&gt;   ..$ transcript_num  : int [1:724] 438 708 405 505 360 406 408 500 538 387 ...</span></span>
<span><span class="co">#&gt;   ..$ modAlt_rsq      : num [1:724] 0.0497 0.0257 0.0519 0.0122 0.0305 ...</span></span>
<span><span class="co">#&gt;   ..$ lrtest_ChiSq    : num [1:724] 22.32 18.46 21.57 6.18 11.16 ...</span></span>
<span><span class="co">#&gt;   ..$ lrtest_Pr       : num [1:724] 0.00792 0.03016 0.01035 0.72192 0.26483 ...</span></span>
<span><span class="co">#&gt;   ..$ UMI_cellID      : chr [1:724] "c_1_1_1008" "c_1_1_1027" "c_1_1_1042" "c_1_1_1058" ...</span></span>
<span><span class="co">#&gt;   ..$ lrtest_nlog10P  : num [1:724] 2.101 1.52 1.985 0.142 0.577 ...</span></span>
<span><span class="co">#&gt;   ..$ tLLR_maxCellType: chr [1:724] "e" "e" "c" "e" ...</span></span>
<span><span class="co">#&gt;   ..$ flagged         : logi [1:724] FALSE FALSE FALSE FALSE FALSE FALSE ...</span></span>
<span><span class="co">#&gt;   ..$ file_idx        : int [1:724] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#&gt;  $ combined_flaggedCells        :List of 2</span></span>
<span><span class="co">#&gt;   ..$ : chr [1:34] "c_1_1_1185" "c_1_1_1188" "c_1_1_1232" "c_1_1_1261" ...</span></span>
<span><span class="co">#&gt;   ..$ : chr [1:2] "c_1_2_2659" "c_1_2_2733"</span></span>
<span><span class="co">#&gt;  $ trimmed_perCellExprs         :Formal class 'dgCMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">#&gt;   .. ..@ i       : int [1:119728] 29 35 41 48 55 58 60 66 71 75 ...</span></span>
<span><span class="co">#&gt;   .. ..@ p       : int [1:755] 0 177 462 638 860 1028 1215 1372 1613 1844 ...</span></span>
<span><span class="co">#&gt;   .. ..@ Dim     : int [1:2] 960 754</span></span>
<span><span class="co">#&gt;   .. ..@ Dimnames:List of 2</span></span>
<span><span class="co">#&gt;   .. .. ..$ : chr [1:960] "AATK" "ABL1" "ABL2" "ACE" ...</span></span>
<span><span class="co">#&gt;   .. .. ..$ : chr [1:754] "c_1_1_1008" "c_1_1_1027" "c_1_1_1042" "c_1_1_1058" ...</span></span>
<span><span class="co">#&gt;   .. ..@ x       : num [1:119728] 1 1 1 1 1 2 1 1 12 2 ...</span></span>
<span><span class="co">#&gt;   .. ..@ factors : list()</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># outputs in output folder</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"res1f_multiFiles"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "1_classDF_ToFlagTrans.csv"  "1_flagged_transDF.csv"     </span></span>
<span><span class="co">#&gt; [3] "1_modStats_ToFlagCells.csv" "2_classDF_ToFlagTrans.csv" </span></span>
<span><span class="co">#&gt; [5] "2_flagged_transDF.csv"      "2_modStats_ToFlagCells.csv"</span></span></code></pre></div>
<p><code><a href="../reference/fastReseg_flag_all_errors.html">fastReseg_flag_all_errors()</a></code> function takes similar input
arguments as <code><a href="../reference/prepare_perFOV_transDF.html">prepare_perFOV_transDF()</a></code> and
<code><a href="../reference/runPreprocess.html">runPreprocess()</a></code> (discussed in section <a href="#preprocess-on-whole-dataset">Preprocess on whole dataset</a>). It
does dataset preprocessing before segmentation error detection. The
function returns a list of outputs including
<code>combined_modStats_ToFlagCells</code>, which is a data.frame for
spatial modeling statistics of cell-level segmentation evaluation (see
section <a href="#flag-cells-with-putative-segmentation-errors">Flag
cells with putative segmentation errors</a> for an excerpt), and
<code>combined_flaggedCells</code>, which contains cell IDs for all
cells flagged with potential cell segmentation errors in the dataset.
<code><a href="../reference/fastReseg_flag_all_errors.html">fastReseg_flag_all_errors()</a></code> function also exports the
per-FOV outputs as 3 individual files per FOV under
<code>path_to_output</code> directory.</p>
<table class="table table table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
variable
</th>
<th style="text-align:left;">
description
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;width: 25%; white-space: pre-line; text-align: left;">
<code>flagged_transDF</code>
</td>
<td style="text-align:left;width: 70%; white-space: pre-line; text-align: left;">
a transcript data.frame for each FOV, with columns for unique IDs of
transcripts <code>UMI_transID</code> and cells <code>UMI_cellID</code>,
for global coordiante system <code>x</code>, <code>y</code>,
<code>z</code>, and for the goodness-of-fit in original cell segment
<code>SVM_class</code>
</td>
</tr>
<tr>
<td style="text-align:left;width: 25%; white-space: pre-line; text-align: left;">
<code>modStats_ToFlagCells</code>
</td>
<td style="text-align:left;width: 70%; white-space: pre-line; text-align: left;">
a data.frame for spatial modeling statistics of each cell, output of
<code><a href="../reference/runSegErrorEvaluation.html">runSegErrorEvaluation()</a></code> function
</td>
</tr>
<tr>
<td style="text-align:left;width: 25%; white-space: pre-line; text-align: left;">
<code>classDF_ToFlagTrans</code>
</td>
<td style="text-align:left;width: 70%; white-space: pre-line; text-align: left;">
data.frame for the class assignment of transcripts within putative
wrongly segmented cells, output of <code><a href="../reference/flag_bad_transcripts.html">flag_bad_transcripts()</a></code>
function
</td>
</tr>
</tbody>
</table>
<p>One can simply trim off the transcripts with low goodness-of-fit to
current segmentation from the <code>flagged_transDF</code> by removing
the transcripts with <code>SVM_class = 0</code>. Conveniently,
<code><a href="../reference/fastReseg_flag_all_errors.html">fastReseg_flag_all_errors()</a></code> would also return
<code>trimmed_perCellExprs</code>, the gene x cell count matrix where
all putative contaminating transcripts are trimmed, when
<code>return_trimmed_perCell = TRUE</code>.</p>
</div>
<div class="section level3">
<h3 id="full-pipeline-segmentation-detection-and-refinement-across-multi-fov-dataset">Full pipeline segmentation detection and refinement across multi-FOV
dataset<a class="anchor" aria-label="anchor" href="#full-pipeline-segmentation-detection-and-refinement-across-multi-fov-dataset"></a>
</h3>
<p>To perform full pipeline on the entire dataset with more complex
segmentation refinement actions, like splitting, merging and trimming,
one can use <code><a href="../reference/fastReseg_full_pipeline.html">fastReseg_full_pipeline()</a></code> pipeline wrapper.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">refineAll_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fastReseg_full_pipeline.html">fastReseg_full_pipeline</a></span><span class="op">(</span></span>
<span>  counts <span class="op">=</span> <span class="va">counts</span>,</span>
<span>  clust <span class="op">=</span> <span class="va">clust</span>,</span>
<span>  refProfiles <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  </span>
<span>  <span class="co"># Similar to `runPreprocess()`, one can use `clust = NULL` if providing `refProfiles`</span></span>
<span>  </span>
<span>  transcript_df <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transDF_fileInfo <span class="op">=</span> <span class="va">transDF_fileInfo</span>,</span>
<span>  filepath_coln <span class="op">=</span> <span class="st">'file_path'</span>,</span>
<span>  prefix_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'slide'</span>,<span class="st">'fov'</span><span class="op">)</span>,</span>
<span>  fovOffset_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'stage_Y'</span>,<span class="st">'stage_X'</span><span class="op">)</span>,</span>
<span>  pixel_size <span class="op">=</span> <span class="fl">0.18</span>,</span>
<span>  zstep_size <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  transID_coln <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transGene_coln <span class="op">=</span> <span class="st">"target"</span>,</span>
<span>  cellID_coln <span class="op">=</span> <span class="st">"CellId"</span>,</span>
<span>  spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>,<span class="st">"y"</span>,<span class="st">"z"</span><span class="op">)</span>,</span>
<span>  extracellular_cellID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>,</span>
<span>  </span>
<span>  <span class="co"># Similar to `runPreprocess()`, one can set various cutoffs to NULL for automatic calculation from input data</span></span>
<span>  </span>
<span>  <span class="co"># distance cutoff for neighborhood searching at molecular and cellular levels, respectively</span></span>
<span>  molecular_distance_cutoff <span class="op">=</span> <span class="fl">2.7</span>, </span>
<span>  cellular_distance_cutoff <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  </span>
<span>  <span class="co"># cutoffs for transcript scores and number for cells under each cell type</span></span>
<span>  score_baseline <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  lowerCutoff_transNum <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  higherCutoff_transNum<span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  imputeFlag_missingCTs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  </span>
<span>  <span class="co"># Settings for error detection and correction, refer to `runSegRefinement()` for more details</span></span>
<span>  flagCell_lrtest_cutoff <span class="op">=</span> <span class="fl">5</span>, <span class="co"># cutoff to flag for cells with strong spatial dependcy in transcript score profiles</span></span>
<span>  svmClass_score_cutoff <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>,   <span class="co"># cutoff of transcript score to separate between high and low score classes</span></span>
<span>  groupTranscripts_method <span class="op">=</span> <span class="st">"dbscan"</span>,</span>
<span>  spatialMergeCheck_method <span class="op">=</span> <span class="st">"leidenCut"</span>, </span>
<span>  cutoff_spatialMerge <span class="op">=</span> <span class="fl">0.5</span>, <span class="co"># spatial constraint cutoff for a valid merge event</span></span>
<span>  </span>
<span>  path_to_output <span class="op">=</span> <span class="st">"res2_multiFiles"</span>,</span>
<span>  save_intermediates <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co"># flag to return and write intermediate results to disk</span></span>
<span>  return_perCellData <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co"># flag to return per cell level outputs from updated segmentation </span></span>
<span>  combine_extra <span class="op">=</span> <span class="cn">FALSE</span> <span class="co"># flag to include trimmed and extracellular transcripts in the exported `updated_transDF.csv` files </span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">refineAll_res</span><span class="op">)</span></span>
<span><span class="co">#&gt; 60, 1:14] 1.00e-04 3.28e-04 4.19e-04 1.35e-04 8.07e-05 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:960] "AATK" "ABL1" "ABL2" "ACE" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:14] "c" "e" "f" "b" ...</span></span>
<span><span class="co">#&gt;  $ baselineData        :List of 2</span></span>
<span><span class="co">#&gt;   ..$ span_score   : num [1:14, 1:5] 0 -1.508 -1.842 -1.806 -0.451 ...</span></span>
<span><span class="co">#&gt;   .. ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. .. ..$ : chr [1:14] "a" "b" "c" "e" ...</span></span>
<span><span class="co">#&gt;   .. .. ..$ : chr [1:5] "0%" "25%" "50%" "75%" ...</span></span>
<span><span class="co">#&gt;   ..$ span_transNum: num [1:14, 1:5] 1 9 16 24 58 13 1 18 1 1 ...</span></span>
<span><span class="co">#&gt;   .. ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. .. ..$ : chr [1:14] "a" "b" "c" "e" ...</span></span>
<span><span class="co">#&gt;   .. .. ..$ : chr [1:5] "0%" "25%" "50%" "75%" ...</span></span>
<span><span class="co">#&gt;  $ cutoffs_list        :List of 5</span></span>
<span><span class="co">#&gt;   ..$ score_baseline           : Named num [1:14] -1.473 -1.299 -1.171 -1.05 -0.979 ...</span></span>
<span><span class="co">#&gt;   .. ..- attr(*, "names")= chr [1:14] "c" "e" "f" "b" ...</span></span>
<span><span class="co">#&gt;   ..$ lowerCutoff_transNum     : Named num [1:14] 261 239.2 45.5 102.5 3 ...</span></span>
<span><span class="co">#&gt;   .. ..- attr(*, "names")= chr [1:14] "c" "e" "f" "b" ...</span></span>
<span><span class="co">#&gt;   ..$ higherCutoff_transNum    : Named num [1:14] 445 484 95 131 22 ...</span></span>
<span><span class="co">#&gt;   .. ..- attr(*, "names")= chr [1:14] "c" "e" "f" "b" ...</span></span>
<span><span class="co">#&gt;   ..$ cellular_distance_cutoff : num 19.2</span></span>
<span><span class="co">#&gt;   ..$ molecular_distance_cutoff: num 2.7</span></span>
<span><span class="co">#&gt;  $ updated_perCellDT   :Classes 'data.table' and 'data.frame':   754 obs. of  6 variables:</span></span>
<span><span class="co">#&gt;   ..$ updated_cellID  : chr [1:754] "c_1_1_1008" "c_1_1_1027" "c_1_1_1042" "c_1_1_1058" ...</span></span>
<span><span class="co">#&gt;   ..$ updated_celltype: chr [1:754] "e" "e" "c" "e" ...</span></span>
<span><span class="co">#&gt;   ..$ x               : num [1:754] -186 -220 -298 -306 -317 ...</span></span>
<span><span class="co">#&gt;   ..$ y               : num [1:754] 4869 4864 4862 4855 4854 ...</span></span>
<span><span class="co">#&gt;   ..$ z               : num [1:754] 3.28 2.72 2.79 2.95 2.84 ...</span></span>
<span><span class="co">#&gt;   ..$ reSeg_action    : chr [1:754] "none" "none" "none" "none" ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, ".internal.selfref")=&lt;externalptr&gt; </span></span>
<span><span class="co">#&gt;  $ updated_perCellExprs:Formal class 'dgCMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">#&gt;   .. ..@ i       : int [1:119731] 29 35 41 48 55 58 60 66 71 75 ...</span></span>
<span><span class="co">#&gt;   .. ..@ p       : int [1:755] 0 177 462 638 860 1028 1215 1372 1613 1844 ...</span></span>
<span><span class="co">#&gt;   .. ..@ Dim     : int [1:2] 960 754</span></span>
<span><span class="co">#&gt;   .. ..@ Dimnames:List of 2</span></span>
<span><span class="co">#&gt;   .. .. ..$ : chr [1:960] "AATK" "ABL1" "ABL2" "ACE" ...</span></span>
<span><span class="co">#&gt;   .. .. ..$ : chr [1:754] "c_1_1_1008" "c_1_1_1027" "c_1_1_1042" "c_1_1_1058" ...</span></span>
<span><span class="co">#&gt;   .. ..@ x       : num [1:119731] 1 1 1 1 1 2 1 1 12 2 ...</span></span>
<span><span class="co">#&gt;   .. ..@ factors : list()</span></span>
<span><span class="co">#&gt;  $ reseg_actions       :List of 4</span></span>
<span><span class="co">#&gt;   ..$ cells_to_discard    : chr [1:55] "c_1_1_1188_g1" "c_1_1_1232_g1" "c_1_1_1232_g2" "c_1_1_1261_g1" ...</span></span>
<span><span class="co">#&gt;   ..$ cells_to_update     : Named chr [1:2] "c_1_1_1744" "c_1_2_2733"</span></span>
<span><span class="co">#&gt;   .. ..- attr(*, "names")= chr [1:2] "c_1_1_1744_g1" "c_1_2_2733_g5"</span></span>
<span><span class="co">#&gt;   ..$ cells_to_keep       : chr(0) </span></span>
<span><span class="co">#&gt;   ..$ reseg_full_converter: Named chr [1:57] "c_1_1_1744" NA NA NA ...</span></span>
<span><span class="co">#&gt;   .. ..- attr(*, "names")= chr [1:57] "c_1_1_1744_g1" "c_1_1_1188_g1" "c_1_1_1232_g1" "c_1_1_1232_g2" ...</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># outputs in output folder</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"res2_multiFiles"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "1_each_segRes.rds"                          </span></span>
<span><span class="co">#&gt; [2] "1_updated_transDF.csv"                      </span></span>
<span><span class="co">#&gt; [3] "2_each_segRes.rds"                          </span></span>
<span><span class="co">#&gt; [4] "2_updated_transDF.csv"                      </span></span>
<span><span class="co">#&gt; [5] "combined_updated_perCellDT_perCellExprs.rds"</span></span></code></pre></div>
<p><code><a href="../reference/fastReseg_full_pipeline.html">fastReseg_full_pipeline()</a></code> returns the new cell
expression count matrix <code>updated_perCellExprs</code> and spatial
coordinate data.frame <code>updated_perCellDT</code> when
<code>return_perCellData = TRUE</code>. It also writes the transcript
data.frame updated with new cell segmentation outcomes into individual
<code>updated_transDF.csv</code> files at FOV level under
<code>path_to_output</code> directory. Intermediate results, including
<code>modStats_ToFlagCells</code> and <code>groupDF_ToFlagTrans</code>
for cell-level and transcript-level segmentation evaluation, are also
saved into individual <code>each_segRes.rds</code> object at FOV level
when <code>save_intermediates = TRUE</code>. For more details, please
refer to the manuals of <code><a href="../reference/fastReseg_full_pipeline.html">fastReseg_full_pipeline()</a></code> and see
section <a href="#processing-each-fov-outside-of-core-wrapper">Processing each FOV
outside of core wrapper</a> for excerpts of example outputs.</p>
</div>
</div>
<div class="section level2">
<h2 id="modular-functions-for-individual-tasks">Modular functions for individual tasks<a class="anchor" aria-label="anchor" href="#modular-functions-for-individual-tasks"></a>
</h2>
<p>While the above 2 pipeline wrappers provide streamline processing of
multi-FOV dataset, itâ€™s sometimes desired to focus on a representative
subset of the data first and check out the impact of various cutoffs on
re-segmentation performance quickly. To do so, one can rely on
<code><a href="../reference/runPreprocess.html">runPreprocess()</a></code> and
<code><a href="../reference/fastReseg_perFOV_full_process.html">fastReseg_perFOV_full_process()</a></code> as shown in this
section.</p>
<div class="section level3">
<h3 id="preprocess-on-whole-dataset">Preprocess on whole dataset<a class="anchor" aria-label="anchor" href="#preprocess-on-whole-dataset"></a>
</h3>
<p>First, one needs to preprocess at whole dataset scale to get
appropriate baselines and cutoffs for downstream segmentation error
detection and correction at individual FOV level. This could be done
using <code><a href="../reference/runPreprocess.html">runPreprocess()</a></code> function.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prep_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runPreprocess.html">runPreprocess</a></span><span class="op">(</span></span>
<span>  counts <span class="op">=</span> <span class="va">counts</span>, </span>
<span>  </span>
<span>  <span class="co">## when certain cell typing has been done on the dataset with initial cell segmentation,  </span></span>
<span>  <span class="co"># set `refProfiles` to NULL, but use the cell typing assignment in `clust`</span></span>
<span>  clust <span class="op">=</span> <span class="va">clust</span>, </span>
<span>  refProfiles <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  </span>
<span>  <span class="co">## if celll typing has NOT been done on the dataset with initial cell segmentation, </span></span>
<span>  <span class="co"># set `clust` to NULL, but use cluster-specific profiles in `refProfiles` instead</span></span>
<span>  </span>
<span>  <span class="co">## of note, when `refProfiles is not NULL, genes unique to `counts` but missing in `refProfiles` would be omitted from downstream analysis.  </span></span>
<span>  </span>
<span>  <span class="co"># cutoffs for transcript scores and number for cells under each cell type</span></span>
<span>  <span class="co"># if NULL, calculate those cutoffs from `counts`, `clust` and/or `refProfiles` across the entire dataset</span></span>
<span>  score_baseline <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  lowerCutoff_transNum <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  higherCutoff_transNum<span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  imputeFlag_missingCTs <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co"># flag to impute transcript score and number cutoffs for cell types in `refProfiles` but missing in `clust`</span></span>
<span>  </span>
<span>  <span class="co"># genes in `counts` but not in `refProfiles` and expect no cell type dependency, e.g. negative control probes</span></span>
<span>  ctrl_genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="co"># cutoff of transcript score to separate between high and low score transcript classes, used as the score values for `ctrl_genes` </span></span>
<span>  svmClass_score_cutoff <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>,</span>
<span>  </span>
<span>  <span class="co"># distance cutoff for neighborhood searching at molecular and cellular levels, respectively</span></span>
<span>  <span class="co"># if NULL, calculate those distance cutoffs from the first transcript data.frame provided (slow process)</span></span>
<span>  <span class="co"># if values provided in input, no distance calculation would be done </span></span>
<span>  molecular_distance_cutoff <span class="op">=</span> <span class="fl">2.7</span>,</span>
<span>  cellular_distance_cutoff <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  </span>
<span>  transcript_df <span class="op">=</span> <span class="cn">NULL</span>, <span class="co"># take a transcript data.frame as input directly when `transDF_fileInfo = NULL`</span></span>
<span>  transDF_fileInfo <span class="op">=</span> <span class="va">transDF_fileInfo</span>, <span class="co"># data.frame info for multiple perFOV transcript data.frame files</span></span>
<span>  filepath_coln <span class="op">=</span> <span class="st">'file_path'</span>, </span>
<span>  prefix_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'slide'</span>,<span class="st">'fov'</span><span class="op">)</span>, </span>
<span>  fovOffset_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'stage_X'</span>,<span class="st">'stage_Y'</span><span class="op">)</span>, </span>
<span>  </span>
<span>  pixel_size <span class="op">=</span> <span class="fl">0.18</span>, <span class="co"># in micron per pixel</span></span>
<span>  zstep_size <span class="op">=</span> <span class="fl">0.8</span>, <span class="co"># in micron per z step</span></span>
<span>  transID_coln <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transGene_coln <span class="op">=</span> <span class="st">"target"</span>,</span>
<span>  </span>
<span>  <span class="co"># cell ID column in the provided transcript data.frame, which is the 1st file in `transDF_fileInfo` in this example</span></span>
<span>  cellID_coln <span class="op">=</span> <span class="st">'CellId'</span>, </span>
<span>  spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'x'</span>,<span class="st">'y'</span>,<span class="st">'z'</span><span class="op">)</span>, </span>
<span>  extracellular_cellID <span class="op">=</span> <span class="fl">0</span> <span class="co"># cell ID for extracellular transcript </span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## variables passing to the downstream pipeline</span></span>
<span><span class="co"># gene x cell type matrix of transcript score </span></span>
<span><span class="va">score_GeneMatrix</span> <span class="op">&lt;-</span> <span class="va">prep_res</span><span class="op">[[</span><span class="st">'score_GeneMatrix'</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># per cell transcript score baseline for each cell type</span></span>
<span><span class="va">score_baseline</span> <span class="op">&lt;-</span> <span class="va">prep_res</span><span class="op">[[</span><span class="st">'cutoffs_list'</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">'score_baseline'</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># upper and lower limit of per cell transcript number for each cell type</span></span>
<span><span class="va">lowerCutoff_transNum</span> <span class="op">&lt;-</span> <span class="va">prep_res</span><span class="op">[[</span><span class="st">'cutoffs_list'</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">'lowerCutoff_transNum'</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">higherCutoff_transNum</span> <span class="op">&lt;-</span> <span class="va">prep_res</span><span class="op">[[</span><span class="st">'cutoffs_list'</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">'higherCutoff_transNum'</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># distance cutoffs for neighborhood at cellular and molecular levels</span></span>
<span><span class="va">cellular_distance_cutoff</span> <span class="op">&lt;-</span> <span class="va">prep_res</span><span class="op">[[</span><span class="st">'cutoffs_list'</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">'cellular_distance_cutoff'</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">molecular_distance_cutoff</span> <span class="op">&lt;-</span> <span class="va">prep_res</span><span class="op">[[</span><span class="st">'cutoffs_list'</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">'molecular_distance_cutoff'</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><code><a href="../reference/runPreprocess.html">runPreprocess()</a></code> function returns a list of outputs among
which <code>score_GeneMatrix</code> and <code>cutoffs_list</code> would
be passed to downstream resegmentation pipeline.</p>
<table class="table table table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
variable
</th>
<th style="text-align:left;">
description
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;width: 25%; white-space: pre-line; text-align: left;">
<code>clust</code>
</td>
<td style="text-align:left;width: 70%; white-space: pre-line; text-align: left;">
vector of cluster assignments used in caculating
<code>baselineData</code>
</td>
</tr>
<tr>
<td style="text-align:left;width: 25%; white-space: pre-line; text-align: left;">
<code>refProfiles</code>
</td>
<td style="text-align:left;width: 70%; white-space: pre-line; text-align: left;">
gene X cluster matrix of cluster-specific reference profiles to use in
resegmenation pipeline
</td>
</tr>
<tr>
<td style="text-align:left;width: 25%; white-space: pre-line; text-align: left;">
<code>baselineData</code>
</td>
<td style="text-align:left;width: 70%; white-space: pre-line; text-align: left;">
list of two matrice in cluster X percentile format for the
cluster-specific percentile distribution of per cell value in terms of
transcript score and number, respectively
</td>
</tr>
<tr>
<td style="text-align:left;width: 25%; white-space: pre-line; text-align: left;">
<code>cutoffs_list</code>
</td>
<td style="text-align:left;width: 70%; white-space: pre-line; text-align: left;">
<i>list of cutoffs to use in resegmentation pipeline:</i>
</td>
</tr>
<tr>
<td style="text-align:left;width: 25%; white-space: pre-line; text-align: left;">
</td>
<td style="text-align:left;width: 70%; white-space: pre-line; text-align: left;">
<ul>
<li>
<code>score_baseline</code>: named vector of per cell type minimal score
required to call certain cell type of high enough confidence.
</li>
</ul>
<ul>
<li>
<code>lowerCutoff_transNum</code>: named vector of per cell type minimal
transcript number per cell to keep query cell as it is.
</li>
</ul>
<ul>
<li>
<code>higherCutoff_transNum</code>: named vector of per cell type
maxmium transcript number per cell to keep query cell as it is without
merging to neighbor cell of consistent cell type.
</li>
</ul>
<ul>
<li>
<code>cellular_distance_cutoff</code>: maximum cell-to-cell distance in
x, y between the center of query cells to the center of neighbor cells
with direct contact, unit in micron.
</li>
</ul>
<ul>
<li>
<code>molecular_distance_cutoff</code>: maximum molecule-to-molecule
distance within connected transcript group, unit in micron.
</li>
</ul>
</td>
</tr>
<tr>
<td style="text-align:left;width: 25%; white-space: pre-line; text-align: left;">
<code>ctrl_genes</code>
</td>
<td style="text-align:left;width: 70%; white-space: pre-line; text-align: left;">
vector of control genes whose transcript scores are set to fixed value
for all cell types, return when <code>ctrl_genes</code> is not
<code>NULL</code>.
</td>
</tr>
<tr>
<td style="text-align:left;width: 25%; white-space: pre-line; text-align: left;">
<code>score_GeneMatrix</code>
</td>
<td style="text-align:left;width: 70%; white-space: pre-line; text-align: left;">
gene x cell-type score matrix to use in resegmenation pipeline, the
scores for <code>ctrl_genes</code> are set to be the same as
<code>svmClass_score_cutoff</code>
</td>
</tr>
<tr>
<td style="text-align:left;width: 25%; white-space: pre-line; text-align: left;">
<code>processed_1st_transDF</code>
</td>
<td style="text-align:left;width: 70%; white-space: pre-line; text-align: left;">
list of 2 elements for the intracellular and extracellular transcript
data.frame of the processed outcomes of 1st transcrip file
</td>
</tr>
</tbody>
</table>
<div class="section level5">
<h5 id="initial-cell-typing-and-control-genes">initial cell typing and control genes<a class="anchor" aria-label="anchor" href="#initial-cell-typing-and-control-genes"></a>
</h5>
<p>The above example of <code><a href="../reference/runPreprocess.html">runPreprocess()</a></code> is using
<code>clust</code> as input, assuming the single-cell dataset that has
gone through certain type of cell typing algorithm using the initial
cell segmentation. In case of no cell typing has been done on the input
dataset, one could set <code>clust</code> to <code>NULL</code>, but
provide cluster-specific profiles in <code>refProfiles</code> as input.
The <code><a href="../reference/runPreprocess.html">runPreprocess()</a></code> function would do quick supervised cell
typing on the input dataset given the initial cell segmentation. Of
note, when <code>refProfiles</code> is provided to
<code><a href="../reference/runPreprocess.html">runPreprocess()</a></code> , genes unique to <code>counts</code> but
missing in <code>refProfiles</code> would be omitted from downstream
analysis. To include all genes in <code>counts</code>, one could set
those unique genes as <code>ctrl_genes</code> whose expression profiles
are expected to either show no strong cell type dependency or be a very
small fraction of the total per cell expression. Alternatively, one can
do a quick cell typing using the <code><a href="../reference/get_baselineCT.html">get_baselineCT()</a></code> function
as the following before feeding the <code>clust</code> to
<code><a href="../reference/runPreprocess.html">runPreprocess()</a></code> with <code>refProfiles</code> set to
<code>NULL</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">baselineData</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_baselineCT.html">get_baselineCT</a></span><span class="op">(</span>refProfiles <span class="op">=</span> <span class="va">refProfiles</span>, counts <span class="op">=</span> <span class="va">counts</span>, clust <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="va">clust</span> <span class="op">&lt;-</span> <span class="va">baselineData</span><span class="op">[[</span><span class="st">'clust_used'</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="distance-cutoffs-defining-neighborhood">distance cutoffs defining neighborhood<a class="anchor" aria-label="anchor" href="#distance-cutoffs-defining-neighborhood"></a>
</h5>
<p>During the pre-processing step, one would also need to define the
distance cutoffs for downstream neighborhood search during segmentation
refinement. These cutoffs could be defined based on prior knowledge
directly or calculated from the provided <code>transcript_df</code>
using either <code><a href="../reference/runPreprocess.html">runPreprocess()</a></code> or
<code><a href="../reference/choose_distance_cutoff.html">choose_distance_cutoff()</a></code> function.</p>
<ul>
<li>
<p><code>cellular_distance_cutoff</code> is defined as maximum
cell-to-cell distance in x, y between the center of query cells to the
center of neighbor cells with direct contact.</p>
<ul>
<li>When set to <code>NULL</code> in the input of
<code><a href="../reference/runPreprocess.html">runPreprocess()</a></code> function, the function calculates average
2D cell diameter from the input transcript data.frame and use 2 times of
the mean cell diameter as <code>cellular_distance_cutoff</code>.</li>
</ul>
</li>
<li>
<p><code>molecular_distance_cutoff</code> is defined as maximum
molecule-to-molecule distance within connected transcript groups
belonging to same source cells. One can decide this value based on the
expected spot density within each cell.</p>
<ul>
<li>When set to <code>NULL</code> in the input of
<code><a href="../reference/runPreprocess.html">runPreprocess()</a></code> function, the function would first randomly
choose <code>sampleSize_cellNum = 2500</code> number of cells from
<code>sampleSize_nROI = 10</code> number of randomly picked ROIs with
search radius to be 5 times of <code>cellular_distance_cutoff</code>,
and then calculate the minimal molecular distance between picked
cells.</li>
</ul>
</li>
</ul>
<p>Below is an example to calculate distance cutoff from the input
transcript data.frame outside the <code><a href="../reference/runPreprocess.html">runPreprocess()</a></code> function
using <code><a href="../reference/choose_distance_cutoff.html">choose_distance_cutoff()</a></code>, which offers more control
on the distance cutoff calculation setup and may allow faster
calculation than doing it within the <code><a href="../reference/runPreprocess.html">runPreprocess()</a></code>
function.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## for demonstration purpose, use the example `mini_transcriptDF`</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">mini_transcriptDF</span><span class="op">)</span></span>
<span><span class="va">transcript_df</span> <span class="op">&lt;-</span> <span class="va">mini_transcriptDF</span></span>
<span></span>
<span></span>
<span><span class="co">## get distance cutoffs</span></span>
<span><span class="va">distCutoffs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/choose_distance_cutoff.html">choose_distance_cutoff</a></span><span class="op">(</span></span>
<span>  <span class="co"># allow to choose any transcript data.frame that is representative to entire dataset</span></span>
<span>  <span class="co"># while `runPreprocess()` uses the first provided transcript data.frame in the file list</span></span>
<span>  <span class="va">transcript_df</span>, </span>
<span>  </span>
<span>  <span class="co"># allow to use 2D spatial coordinates here since transcript is more dense in 2D, </span></span>
<span>  <span class="co"># 2D calculation of distance cutoff would be faster than 3D calculation used in `runPreprocess()` </span></span>
<span>  spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'x'</span>,<span class="st">'y'</span><span class="op">)</span>, </span>
<span>  </span>
<span>  transID_coln <span class="op">=</span> <span class="st">'UMI_transID'</span>,</span>
<span>  cellID_coln <span class="op">=</span> <span class="st">'UMI_cellID'</span>, </span>
<span>  extracellular_cellID <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  </span>
<span>  <span class="co"># flag to calculate `molecular_distance_cutoff` from input data, slower process</span></span>
<span>  run_molecularDist <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="co"># configs on random sampling of cells</span></span>
<span>  sampleSize_nROI <span class="op">=</span> <span class="fl">10</span>, </span>
<span>  sampleSize_cellNum <span class="op">=</span> <span class="fl">2500</span>, </span>
<span>  seed <span class="op">=</span> <span class="fl">123</span> <span class="op">)</span></span>
<span><span class="co">#&gt; Use 2 times of average 2D cell diameter as cellular_distance_cutoff = 24.2375 for searching of neighbor cells.</span></span>
<span><span class="co">#&gt; Identified 2D coordinates with variance.</span></span>
<span><span class="co">#&gt; Warning: data contain duplicated points</span></span>
<span><span class="co">#&gt; Distribution of minimal molecular distance between 1375 cells: 0, 0.04, 0.07, 0.09, 0.12, 0.15, 0.19, 0.23, 0.28, 0.35, 3.49, at quantile = 0%, 10%, 20%, 30%, 40%, 50%, 60%, 70%, 80%, 90%, 100%.</span></span>
<span><span class="co">#&gt; Use 5 times of 90% quantile of minimal 2D molecular distance between picked cells as `molecular_distance_cutoff` = 1.7655 for defining direct neighbor cells.</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">molecular_distance_cutoff</span> <span class="op">&lt;-</span> <span class="va">distCutoffs</span><span class="op">[[</span><span class="st">'molecular_distance_cutoff'</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">cellular_distance_cutoff</span> <span class="op">&lt;-</span> <span class="va">distCutoffs</span><span class="op">[[</span><span class="st">'cellular_distance_cutoff'</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p>Empirically, letâ€™s set 20um and 2um values for the two cutoffs,
respectively, for dataset on human tissue with 100+ plex target gene in
panel.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellular_distance_cutoff</span> <span class="op">=</span> <span class="fl">20</span></span>
<span><span class="va">molecular_distance_cutoff</span> <span class="op">=</span> <span class="fl">2</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## for demonstration purpose, use the saved baseline values paired with example `mini_transcriptDF`</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_baselineCT"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">score_baseline</span> <span class="op">=</span> <span class="va">example_baselineCT</span><span class="op">[[</span><span class="st">"span_score"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="st">"25%"</span><span class="op">]</span></span>
<span><span class="va">lowerCutoff_transNum</span> <span class="op">=</span> <span class="va">example_baselineCT</span><span class="op">[[</span><span class="st">"span_transNum"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="st">"25%"</span><span class="op">]</span></span>
<span><span class="va">higherCutoff_transNum</span> <span class="op">=</span> <span class="va">example_baselineCT</span><span class="op">[[</span><span class="st">"span_transNum"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="st">"50%"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># calculate log-likelihood of each gene under each cell type and center the score matrix on per gene basis</span></span>
<span><span class="va">score_GeneMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scoreGenesInRef.html">scoreGenesInRef</a></span><span class="op">(</span>genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">refProfiles</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                                    ref_profiles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmax</a></span><span class="op">(</span><span class="va">refProfiles</span>, <span class="fl">1e-5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p>The following sections would operate on one per-FOV transcript
data.frame at a time. For processing on multiple FOVs, please either
refer to the provided pipeline wrapper functions or create your own
wrapper around the code chucks listed below.</p>
</blockquote>
</div>
</div>
<div class="section level3">
<h3 id="core-wrapper-for-perfov-processing">Core wrapper for perFOV processing<a class="anchor" aria-label="anchor" href="#core-wrapper-for-perfov-processing"></a>
</h3>
<p>With the appropriate cutoffs identified in preprocess step, one could
push a single FOV <code>transcript_df</code> through the full pipeline
using a core wrapper function,
<code><a href="../reference/fastReseg_perFOV_full_process.html">fastReseg_perFOV_full_process()</a></code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">mini_transcriptDF</span><span class="op">)</span></span>
<span><span class="va">extracellular_cellID</span> <span class="op">&lt;-</span> <span class="va">mini_transcriptDF</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">mini_transcriptDF</span><span class="op">$</span><span class="va">CellId</span> <span class="op">==</span><span class="fl">0</span><span class="op">)</span>, <span class="st">'cell_ID'</span><span class="op">]</span></span>
<span></span>
<span><span class="va">finalRes_perFOV</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fastReseg_perFOV_full_process.html">fastReseg_perFOV_full_process</a></span><span class="op">(</span></span>
<span>  score_GeneMatrix <span class="op">=</span> <span class="va">score_GeneMatrix</span>, </span>
<span>  transcript_df <span class="op">=</span> <span class="va">mini_transcriptDF</span>, </span>
<span>  transID_coln <span class="op">=</span> <span class="st">'UMI_transID'</span>,</span>
<span>  transGene_coln <span class="op">=</span> <span class="st">"target"</span>,</span>
<span>  cellID_coln <span class="op">=</span> <span class="st">'UMI_cellID'</span>, </span>
<span>  spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'x'</span>,<span class="st">'y'</span>,<span class="st">'z'</span><span class="op">)</span>, </span>
<span>  extracellular_cellID <span class="op">=</span> <span class="va">extracellular_cellID</span>, </span>
<span>  flagModel_TransNum_cutoff <span class="op">=</span> <span class="fl">50</span>, </span>
<span>  flagCell_lrtest_cutoff <span class="op">=</span> <span class="va">flagCell_lrtest_cutoff</span>,</span>
<span>  svmClass_score_cutoff <span class="op">=</span> <span class="va">svmClass_score_cutoff</span>, </span>
<span>  molecular_distance_cutoff <span class="op">=</span> <span class="va">molecular_distance_cutoff</span>,</span>
<span>  cellular_distance_cutoff <span class="op">=</span> <span class="va">cellular_distance_cutoff</span>,</span>
<span>  score_baseline <span class="op">=</span> <span class="va">score_baseline</span>, </span>
<span>  lowerCutoff_transNum <span class="op">=</span> <span class="va">lowerCutoff_transNum</span>, </span>
<span>  higherCutoff_transNum <span class="op">=</span> <span class="va">higherCutoff_transNum</span>,</span>
<span>  </span>
<span>  <span class="co"># default to "dbscan" for spatial grouping of transcripts, alternative to use "delaunay"</span></span>
<span>  groupTranscripts_method <span class="op">=</span> <span class="st">"dbscan"</span>,</span>
<span>  </span>
<span>  <span class="co"># default to "leidenCut" for decision based on Leiden clustering of transcript coordinates, alternative to use "geometryDiff" for geometric analysis</span></span>
<span>  spatialMergeCheck_method <span class="op">=</span> <span class="st">"leidenCut"</span>, </span>
<span>  </span>
<span>  cutoff_spatialMerge <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  return_intermediates <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  return_perCellData <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>  includeAllRefGenes <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="processing-each-fov-outside-of-core-wrapper">Processing each FOV outside of core wrapper<a class="anchor" aria-label="anchor" href="#processing-each-fov-outside-of-core-wrapper"></a>
</h3>
<p>As illustrated in the <a href="#fastreseg-workflow-diagram">FastReseg
workflow diagram</a>, <code><a href="../reference/fastReseg_perFOV_full_process.html">fastReseg_perFOV_full_process()</a></code> would
process a single-FOV transcript data through a series of modules. This
section provides a breakdown on what each module does and explains on
their results with example dataset and figures. One can play with the
input cutoffs and see how it affects each step.</p>
<div class="section level4">
<h4 id="flag-cells-with-putative-segmentation-errors">Flag cells with putative segmentation errors<a class="anchor" aria-label="anchor" href="#flag-cells-with-putative-segmentation-errors"></a>
</h4>
<p>After pre-processing, we are ready to evaluate each cell on their
potential to have cell segmentation errors.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSegErrorEvaluation.html">runSegErrorEvaluation</a></span><span class="op">(</span></span>
<span>  score_GeneMatrix<span class="op">=</span> <span class="va">score_GeneMatrix</span>, </span>
<span>  transcript_df <span class="op">=</span> <span class="va">transcript_df</span>, </span>
<span>  cellID_coln <span class="op">=</span> <span class="st">'UMI_cellID'</span>, </span>
<span>  transID_coln <span class="op">=</span> <span class="st">'UMI_transID'</span>,</span>
<span>  transGene_coln <span class="op">=</span> <span class="st">'target'</span>,</span>
<span>  spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'x'</span>,<span class="st">'y'</span>,<span class="st">'z'</span><span class="op">)</span>,</span>
<span>  <span class="co"># cutoff of transcript number to do spatial modeling</span></span>
<span>  flagModel_TransNum_cutoff <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> </span>
<span><span class="co">#&gt; Found 960 common genes among transcript_df and score_GeneMatrix.</span></span>
<span><span class="co">#&gt; Found 1375 cells and assigned cell type based on the provided `refProfiles` cluster profiles.</span></span>
<span><span class="co">#&gt; Run linear regreassion in 3 Dimension.</span></span>
<span><span class="co">#&gt; Warning in score_cell_segmentation_error(chosen_cells =</span></span>
<span><span class="co">#&gt; names(celltype_cellVector), : Below model_cutoff = 50, skip 37 cells with fewer</span></span>
<span><span class="co">#&gt; transcripts. Move forward with remaining 1338 cells.</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">modStats_ToFlagCells</span> <span class="op">&lt;-</span> <span class="va">outs</span> <span class="op">[[</span><span class="st">'modStats_ToFlagCells'</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">transcript_df</span> <span class="op">&lt;-</span> <span class="va">outs</span><span class="op">[[</span><span class="st">'transcript_df'</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">outs</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># transcript data.frame with additional columns for cell types and transcript scores under current cell segmentation </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">transcript_df</span>, n <span class="op">=</span> <span class="fl">3L</span><span class="op">)</span></span>
<span><span class="co">#&gt;    UMI_transID UMI_cellID        x         y   z  target slide fov CellId</span></span>
<span><span class="co">#&gt; 1     t_1_2_10    c_1_2_5 914.8070 -7832.786 2.4 HLA-DRA     1   2      5</span></span>
<span><span class="co">#&gt; 2 t_1_2_100005 c_1_2_1270 841.7000 -8045.011 2.4   HLA-E     1   2   1270</span></span>
<span><span class="co">#&gt; 3 t_1_2_100006 c_1_2_1270 844.4609 -8045.006 2.4  TWIST1     1   2   1270</span></span>
<span><span class="co">#&gt;   tLLR_maxCellType score_tLLR_maxCellType</span></span>
<span><span class="co">#&gt; 1                f             -0.4227132</span></span>
<span><span class="co">#&gt; 2                d             -0.5070020</span></span>
<span><span class="co">#&gt; 3                d             -2.2769526</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># model statistics</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">modStats_ToFlagCells</span><span class="op">)</span></span>
<span><span class="co">#&gt;            transcript_num modAlt_rsq lrtest_ChiSq    lrtest_Pr UMI_cellID</span></span>
<span><span class="co">#&gt; c_1_2_1000            502 0.05528780     28.55123 7.711518e-04 c_1_2_1000</span></span>
<span><span class="co">#&gt; c_1_2_1001            141 0.13150271     19.87971 1.080087e-02 c_1_2_1001</span></span>
<span><span class="co">#&gt; c_1_2_1009            734 0.03751650     28.06695 9.296016e-04 c_1_2_1009</span></span>
<span><span class="co">#&gt; c_1_2_1010            480 0.03255556     15.88670 6.928571e-02 c_1_2_1010</span></span>
<span><span class="co">#&gt; c_1_2_1011            227 0.09747293     23.28035 5.596473e-03 c_1_2_1011</span></span>
<span><span class="co">#&gt; c_1_2_1012            799 0.05433535     44.63795 1.076376e-06 c_1_2_1012</span></span>
<span><span class="co">#&gt;            lrtest_nlog10P tLLR_maxCellType</span></span>
<span><span class="co">#&gt; c_1_2_1000       3.112860                d</span></span>
<span><span class="co">#&gt; c_1_2_1001       1.966541                f</span></span>
<span><span class="co">#&gt; c_1_2_1009       3.031703                d</span></span>
<span><span class="co">#&gt; c_1_2_1010       1.159356                d</span></span>
<span><span class="co">#&gt; c_1_2_1011       2.252086                d</span></span>
<span><span class="co">#&gt; c_1_2_1012       5.968036                d</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># histogram for spatial dependency in all cells</span></span>
<span><span class="va">tmp_flag</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">modStats_ToFlagCells</span><span class="op">$</span><span class="va">lrtest_nlog10P</span><span class="op">)</span><span class="op">)</span> <span class="co"># exclude cells with too few transcript number</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">modStats_ToFlagCells</span><span class="op">$</span><span class="va">lrtest_nlog10P</span><span class="op">[</span><span class="va">tmp_flag</span><span class="op">]</span>, breaks <span class="op">=</span> <span class="st">"FD"</span>, </span>
<span>     main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Histogram of spatial dependency, mean = "</span>, </span>
<span>                   <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">modStats_ToFlagCells</span><span class="op">$</span><span class="va">lrtest_nlog10P</span><span class="op">[</span><span class="va">tmp_flag</span><span class="op">]</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">modStats_ToFlagCells</span><span class="op">$</span><span class="va">lrtest_nlog10P</span><span class="op">[</span><span class="va">tmp_flag</span><span class="op">]</span><span class="op">)</span>, col<span class="op">=</span><span class="st">"red"</span>, lwd<span class="op">=</span><span class="fl">3</span>, lty<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="figures%2Feval_cellSeg-1.png" width="100%"></p>
<p>The function above returns the statistics for evaluating each cell
for spatial dependent model against null model. Based on the P value,
<code>lrtest_Pr</code> or the negative log10 value
<code>lrtest_nlog10P</code>, one can select for cells with strong
spatial dependency in transcript score profile. Those cells are likely
to contain contaminating transcripts for neighbor cells.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># cutoff to flag for cells with strong spatial dependcy in transcript score profiles</span></span>
<span><span class="va">flagCell_lrtest_cutoff</span>  <span class="op">=</span> <span class="fl">5</span></span>
<span></span>
<span><span class="va">modStats_ToFlagCells</span><span class="op">[[</span><span class="st">'flagged'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">modStats_ToFlagCells</span><span class="op">[[</span><span class="st">'lrtest_nlog10P'</span><span class="op">]</span><span class="op">]</span> <span class="op">&gt;</span> <span class="va">flagCell_lrtest_cutoff</span> <span class="op">)</span></span>
<span><span class="va">flagged_cells</span> <span class="op">&lt;-</span> <span class="va">modStats_ToFlagCells</span><span class="op">[[</span><span class="st">'UMI_cellID'</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">modStats_ToFlagCells</span><span class="op">[[</span><span class="st">'flagged'</span><span class="op">]</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%d cells, %.4f of all evaluated cells, are flagged for resegmentation with lrtest_nlog10P &gt; %.1f."</span>, </span>
<span>                 <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">flagged_cells</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">flagged_cells</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">modStats_ToFlagCells</span><span class="op">)</span>, <span class="va">flagCell_lrtest_cutoff</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; 373 cells, 0.2788 of all evaluated cells, are flagged for resegmentation with lrtest_nlog10P &gt; 5.0.</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># spatial plot some flagged cells with various degrees of spatial dependency in transcript profiles</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">modStats_ToFlagCells</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">modStats_ToFlagCells</span><span class="op">$</span><span class="va">UMI_cellID</span></span>
<span><span class="va">cells_to_plot</span> <span class="op">&lt;-</span> <span class="va">modStats_ToFlagCells</span><span class="op">[</span><span class="va">flagged_cells</span>, <span class="st">'lrtest_nlog10P'</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cells_to_plot</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">flagged_cells</span></span>
<span><span class="va">cells_to_plot</span> <span class="op">&lt;-</span> <span class="va">cells_to_plot</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">cells_to_plot</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">cells_to_plot</span> <span class="op">&lt;-</span> <span class="va">cells_to_plot</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">cells_to_plot</span><span class="op">)</span>, by <span class="op">=</span> <span class="fl">25</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSpatialScoreMultiCells.html">plotSpatialScoreMultiCells</a></span><span class="op">(</span>chosen_cells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cells_to_plot</span><span class="op">)</span>, </span>
<span>                           cell_labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">cells_to_plot</span>, <span class="fl">2</span><span class="op">)</span>, </span>
<span>                           transcript_df <span class="op">=</span> <span class="va">transcript_df</span>, </span>
<span>                           cellID_coln <span class="op">=</span> <span class="st">"UMI_cellID"</span>, </span>
<span>                           transID_coln <span class="op">=</span> <span class="st">"UMI_transID"</span>,</span>
<span>                           score_coln <span class="op">=</span> <span class="st">"score_tLLR_maxCellType"</span>, </span>
<span>                           spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>,<span class="st">"y"</span><span class="op">)</span>,</span>
<span>                           point_size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="figures%2Ffig_flagCells-1.png" width="100%"></p>
</div>
<div class="section level4">
<h4 id="identify-wrongly-segmented-transcript-groups">Identify wrongly segmented transcript groups<a class="anchor" aria-label="anchor" href="#identify-wrongly-segmented-transcript-groups"></a>
</h4>
<p>Under the assumption that the contamination from neighbor cells would
result in patches of low-score transcript groups in space, we first
separate the transcripts within each flagged cells into high and low
score groups and then divide the transcripts of low score into different
spatially connected groups assuming they might arise from different
source cells in neighborhood.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># cutoff of transcript score to separate between high and low score transcripts </span></span>
<span><span class="va">svmClass_score_cutoff</span> <span class="op">=</span> <span class="op">-</span><span class="fl">2</span> </span>
<span></span>
<span><span class="co"># a list of arguments to pass to `e1071::svm` function to define the strength of spatial connectivity </span></span>
<span><span class="va">svm_args</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>kernel <span class="op">=</span> <span class="st">"radial"</span>, </span>
<span>                scale <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                gamma <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span></span>
<span></span>
<span><span class="va">groupDF_ToFlagTrans</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runTranscriptErrorDetection.html">runTranscriptErrorDetection</a></span><span class="op">(</span>chosen_cells <span class="op">=</span> <span class="va">flagged_cells</span>,</span>
<span>                                                   score_GeneMatrix <span class="op">=</span> <span class="va">score_GeneMatrix</span>, </span>
<span>                                                   transcript_df <span class="op">=</span> <span class="va">transcript_df</span>, <span class="co"># include column for transcript score under current cell segmentation</span></span>
<span>                                                   cellID_coln <span class="op">=</span> <span class="st">"UMI_cellID"</span>, </span>
<span>                                                   transID_coln <span class="op">=</span> <span class="st">"UMI_transID"</span>, </span>
<span>                                                   <span class="co"># column for transcript score in current cell segment</span></span>
<span>                                                   score_coln <span class="op">=</span> <span class="st">'score_tLLR_maxCellType'</span>,</span>
<span>                                                   spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>,<span class="st">"y"</span>,<span class="st">"z"</span><span class="op">)</span>,</span>
<span>                                                   model_cutoff <span class="op">=</span> <span class="fl">50</span>, </span>
<span>                                                   score_cutoff <span class="op">=</span> <span class="va">svmClass_score_cutoff</span>, </span>
<span>                                                   svm_args <span class="op">=</span> <span class="va">svm_args</span>,</span>
<span>                                                   <span class="co"># maximum molecule-to-molecule distance within same transcript group </span></span>
<span>                                                   distance_cutoff <span class="op">=</span> <span class="va">molecular_distance_cutoff</span>, </span>
<span>                                                   <span class="co"># use "dbscan" method for spatial grouping of transcripts, alternative to use "delaunay"</span></span>
<span>                                                   groupTranscripts_method <span class="op">=</span> <span class="st">"dbscan"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Run SVM in 3 Dimension.</span></span>
<span><span class="co">#&gt; Found 373 common cells and 960 common genes among chosen_cells, transcript_df, and score_GeneMatrix.</span></span>
<span><span class="co">#&gt; Warning in flag_bad_transcripts(chosen_cells = chosen_cells, score_GeneMatrix =</span></span>
<span><span class="co">#&gt; score_GeneMatrix, : Below model_cutoff = 50, skip 0 cells with fewer</span></span>
<span><span class="co">#&gt; transcripts. Move forward with remaining 373 cells.</span></span>
<span><span class="co">#&gt; Warning in flag_bad_transcripts(chosen_cells = chosen_cells, score_GeneMatrix =</span></span>
<span><span class="co">#&gt; score_GeneMatrix, : Skip 0 cells with all transcripts in same class given</span></span>
<span><span class="co">#&gt; `score_cutoff = -2`. Move forward with remaining 373 cells.</span></span>
<span><span class="co">#&gt; Remove 0 cells with raw transcript score all in same class based on cutoff -2.00 when running spatial SVM model.</span></span>
<span><span class="co">#&gt; Do spatial network analysis in 3 Dimension.</span></span>
<span><span class="co">#&gt; 10652 chosen_transcripts are found in common cells.</span></span>
<span><span class="co">#&gt; SVM spatial model further identified 17 cells with transcript score all in same class, exclude from transcript group analysis.</span></span>
<span><span class="co">#&gt; Found 960 common genes among transcript_df and score_GeneMatrix.</span></span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">groupDF_ToFlagTrans</span><span class="op">)</span></span>
<span><span class="co">#&gt;    UMI_transID UMI_cellID        x         y   z  target slide fov CellId</span></span>
<span><span class="co">#&gt; 1 t_1_2_100019 c_1_2_1280 1075.583 -8045.011 2.4    PSAP     1   2   1280</span></span>
<span><span class="co">#&gt; 2 t_1_2_100021 c_1_2_1311 1180.613 -8045.024 2.4    RXRB     1   2   1311</span></span>
<span><span class="co">#&gt; 3 t_1_2_100042 c_1_2_1311 1180.446 -8045.052 2.4  S100A6     1   2   1311</span></span>
<span><span class="co">#&gt; 4 t_1_2_100054 c_1_2_1280 1070.457 -8045.092 2.4    ABL2     1   2   1280</span></span>
<span><span class="co">#&gt; 5 t_1_2_100055 c_1_2_1297 1109.888 -8045.084 2.4 IL22RA1     1   2   1297</span></span>
<span><span class="co">#&gt; 6 t_1_2_100062 c_1_2_1280 1078.530 -8045.144 2.4    CD63     1   2   1280</span></span>
<span><span class="co">#&gt;   tLLR_maxCellType score_tLLR_maxCellType     DecVal SVM_class SVM_cell_type</span></span>
<span><span class="co">#&gt; 1                d            -0.84881774  1.0703177         1             d</span></span>
<span><span class="co">#&gt; 2                d            -0.83865972  0.9651801         1             d</span></span>
<span><span class="co">#&gt; 3                d             0.00000000  1.0000416         1             d</span></span>
<span><span class="co">#&gt; 4                d            -1.57303427  1.0000334         1             d</span></span>
<span><span class="co">#&gt; 5                f            -4.08493706 -0.8650251         1             f</span></span>
<span><span class="co">#&gt; 6                d            -0.01510279  1.0003084         1             d</span></span>
<span><span class="co">#&gt;   connect_group tmp_cellID group_maxCellType</span></span>
<span><span class="co">#&gt; 1             0 c_1_2_1280                 d</span></span>
<span><span class="co">#&gt; 2             0 c_1_2_1311                 d</span></span>
<span><span class="co">#&gt; 3             0 c_1_2_1311                 d</span></span>
<span><span class="co">#&gt; 4             0 c_1_2_1280                 d</span></span>
<span><span class="co">#&gt; 5             0 c_1_2_1297                 f</span></span>
<span><span class="co">#&gt; 6             0 c_1_2_1280                 d</span></span></code></pre></div>
<p>The function above returns a transcript data.frame for the flagged
cells with results in spatial-dependent score classification and spatial
group ID assignment.</p>
<ul>
<li>
<p><code>SVM_class</code> shows the transcript score classification,
<code>0</code> for low score below cutoff, <code>1</code> for high score
above cutoff; the corresponding decision values of svm model output are
listed in <code>DecVal</code>.</p>
<ul>
<li>One can use <code>SVM_class</code> to select all low-score
transcript groups and then remove them from the original transcript
data.frame all together. This approach effectively trims off the
putative contaminating transcripts from current cell segmentation
without more complex refinement.</li>
</ul>
</li>
<li><p><code>connect_group</code> shows the spatial group ID assigned to
each transcripts, while the corresponding cell types with maximum
transcript scores under the given transcript groups are listed in
<code>group_maxCellType</code>. <code>0</code> for transcript group with
high score under the putative cell type of current cell
segmentation.</p></li>
<li><p><code>tmp_cellID</code> is the column for new cell IDs with which
each identified low-score transcript group is assigned with a unique new
name to separate from its original cell. Transcript groups with high
score would keep the same cell ID as the corresponding original
cells.</p></li>
</ul>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># spatial plot for `SVM_class`, the high vs. low score classification of transcript groups in flagged cells</span></span>
<span><span class="fu"><a href="../reference/plotSpatialScoreMultiCells.html">plotSpatialScoreMultiCells</a></span><span class="op">(</span>chosen_cells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cells_to_plot</span><span class="op">)</span>,</span>
<span>                           cell_labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">cells_to_plot</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>                           transcript_df <span class="op">=</span> <span class="va">groupDF_ToFlagTrans</span>,</span>
<span>                           cellID_coln <span class="op">=</span> <span class="st">"UMI_cellID"</span>,</span>
<span>                           transID_coln <span class="op">=</span> <span class="st">"UMI_transID"</span>,</span>
<span>                           score_coln <span class="op">=</span> <span class="st">"SVM_class"</span>,</span>
<span>                           spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>,<span class="st">"y"</span><span class="op">)</span>,</span>
<span>                           point_size <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>                           plot_discrete <span class="op">=</span> <span class="cn">T</span>, </span>
<span>                           title <span class="op">=</span> <span class="st">"transcript score classification"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># spatial plot for `connect_group`, the spatial group ID for transcripts within each cell</span></span>
<span><span class="fu"><a href="../reference/plotSpatialScoreMultiCells.html">plotSpatialScoreMultiCells</a></span><span class="op">(</span>chosen_cells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cells_to_plot</span><span class="op">)</span>,</span>
<span>                           cell_labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">cells_to_plot</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>                           transcript_df <span class="op">=</span> <span class="va">groupDF_ToFlagTrans</span>,</span>
<span>                           cellID_coln <span class="op">=</span> <span class="st">"UMI_cellID"</span>,</span>
<span>                           transID_coln <span class="op">=</span> <span class="st">"UMI_transID"</span>,</span>
<span>                           score_coln <span class="op">=</span> <span class="st">"connect_group"</span>,</span>
<span>                           spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>,<span class="st">"y"</span><span class="op">)</span>,</span>
<span>                           point_size <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>                           plot_discrete <span class="op">=</span> <span class="cn">T</span>, </span>
<span>                           title <span class="op">=</span> <span class="st">"spatial connected transcript groups"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># spatial plot for `group_maxCellType`, the cell type with maximum score for each transcript group</span></span>
<span><span class="fu"><a href="../reference/plotSpatialScoreMultiCells.html">plotSpatialScoreMultiCells</a></span><span class="op">(</span>chosen_cells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cells_to_plot</span><span class="op">)</span>,</span>
<span>                           cell_labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">cells_to_plot</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>                           transcript_df <span class="op">=</span> <span class="va">groupDF_ToFlagTrans</span>,</span>
<span>                           cellID_coln <span class="op">=</span> <span class="st">"UMI_cellID"</span>,</span>
<span>                           transID_coln <span class="op">=</span> <span class="st">"UMI_transID"</span>,</span>
<span>                           score_coln <span class="op">=</span> <span class="st">"group_maxCellType"</span>,</span>
<span>                           spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>,<span class="st">"y"</span><span class="op">)</span>,</span>
<span>                           point_size <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>                           plot_discrete <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                           title <span class="op">=</span> <span class="st">"putative cell type for each transcript group"</span><span class="op">)</span></span></code></pre></div>
<p><img src="figures%2Ffig_flagTranscripts-1.png" width="100%"><img src="figures%2Ffig_flagTranscripts-2.png" width="100%"><img src="figures%2Ffig_flagTranscripts-3.png" width="100%"></p>
</div>
<div class="section level4">
<h4 id="get-ready-for-segmentation-refinement">Get ready for segmentation refinement<a class="anchor" aria-label="anchor" href="#get-ready-for-segmentation-refinement"></a>
</h4>
<p>If one would like to perform more sophisticated cell segmentation
refinement than simple trimming for all low-score transcript groups, one
should update the original transcript data.frame with the new cell IDs
and maximum cell types from the output of
<code><a href="../reference/runTranscriptErrorDetection.html">runTranscriptErrorDetection()</a></code> function to keep those
identified low-score transcript groups separated from their original
cell segment assignment before further segmentation refinement.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># update the transcript_df with flagged transcript_group</span></span>
<span><span class="va">reSeg_ready_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepResegDF.html">prepResegDF</a></span><span class="op">(</span>transcript_df <span class="op">=</span> <span class="va">transcript_df</span>, </span>
<span>                               groupDF_ToFlagTrans <span class="op">=</span> <span class="va">groupDF_ToFlagTrans</span>,</span>
<span>                               cellID_coln <span class="op">=</span> <span class="st">"UMI_cellID"</span>,</span>
<span>                               transID_coln <span class="op">=</span> <span class="st">"UMI_transID"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># `tmp_cellID` and `group_maxCellType` now contains info for all cells including the identified transcript groups </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">reSeg_ready_res</span><span class="op">[[</span><span class="st">"reseg_transcript_df"</span><span class="op">]</span><span class="op">]</span>, n <span class="op">=</span> <span class="fl">3L</span><span class="op">)</span></span>
<span><span class="co">#&gt;    UMI_transID UMI_cellID        x         y   z  target slide fov CellId</span></span>
<span><span class="co">#&gt; 1     t_1_2_10    c_1_2_5 914.8070 -7832.786 2.4 HLA-DRA     1   2      5</span></span>
<span><span class="co">#&gt; 2 t_1_2_100005 c_1_2_1270 841.7000 -8045.011 2.4   HLA-E     1   2   1270</span></span>
<span><span class="co">#&gt; 3 t_1_2_100006 c_1_2_1270 844.4609 -8045.006 2.4  TWIST1     1   2   1270</span></span>
<span><span class="co">#&gt;   tLLR_maxCellType score_tLLR_maxCellType connect_group tmp_cellID</span></span>
<span><span class="co">#&gt; 1                f             -0.4227132             0    c_1_2_5</span></span>
<span><span class="co">#&gt; 2                d             -0.5070020             0 c_1_2_1270</span></span>
<span><span class="co">#&gt; 3                d             -2.2769526             0 c_1_2_1270</span></span>
<span><span class="co">#&gt;   group_maxCellType</span></span>
<span><span class="co">#&gt; 1                 f</span></span>
<span><span class="co">#&gt; 2                 d</span></span>
<span><span class="co">#&gt; 3                 d</span></span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># cells or group IDs for neighborhood evaluation </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">reSeg_ready_res</span><span class="op">[[</span><span class="st">"groups_to_reseg"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "c_1_2_1300_g1" "c_1_2_1340_g1" "c_1_2_1311_g1" "c_1_2_1376_g1"</span></span>
<span><span class="co">#&gt; [5] "c_1_2_1380_g1" "c_1_2_1356_g1"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="perform-segmentation-refinement">Perform segmentation refinement<a class="anchor" aria-label="anchor" href="#perform-segmentation-refinement"></a>
</h4>
<p>For more sophisticated cell segmentation refinement, we first
evaluate the neighborhood environment of each low-score transcript
groups in terms of their goodness of fit to nearby potential source
cells, and then determine the corresponding resegmentation operation by
comparing the neighborhood analysis results with the baseline data for
each cell type derived from the original dataset. Lastly, we apply those
resegmentation operations to the original transcript data.frame to get
the refined segmentation outcomes. Each identified low-score transcript
group may experience the following refinement operation: trimming
(removing to extracellular space), splitting (keeping as new cell ID by
itself), or merging (assigning same cell ID as that of neighboring
cells/transcript groups of same cell type).</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">finalRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runSegRefinement.html">runSegRefinement</a></span><span class="op">(</span></span>
<span>  score_GeneMatrix <span class="op">=</span> <span class="va">score_GeneMatrix</span>,  </span>
<span>  chosen_cells <span class="op">=</span> <span class="va">reSeg_ready_res</span><span class="op">[[</span><span class="st">"groups_to_reseg"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>  reseg_transcript_df <span class="op">=</span> <span class="va">reSeg_ready_res</span><span class="op">[[</span><span class="st">"reseg_transcript_df"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>  reseg_cellID_coln <span class="op">=</span> <span class="st">"tmp_cellID"</span>, </span>
<span>  reseg_celltype_coln <span class="op">=</span> <span class="st">"group_maxCellType"</span>, </span>
<span>  transID_coln <span class="op">=</span> <span class="st">"UMI_transID"</span>,</span>
<span>  transGene_coln <span class="op">=</span> <span class="st">"target"</span>, </span>
<span>  transSpatLocs_coln <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'x'</span>,<span class="st">'y'</span>,<span class="st">'z'</span><span class="op">)</span>,</span>
<span>  score_baseline <span class="op">=</span> <span class="va">score_baseline</span>, </span>
<span>  lowerCutoff_transNum <span class="op">=</span> <span class="va">lowerCutoff_transNum</span>, </span>
<span>  higherCutoff_transNum<span class="op">=</span> <span class="va">higherCutoff_transNum</span>, </span>
<span>  neighbor_distance_xy <span class="op">=</span> <span class="va">cellular_distance_cutoff</span>,</span>
<span>  distance_cutoff <span class="op">=</span> <span class="va">molecular_distance_cutoff</span>,</span>
<span>  </span>
<span>  <span class="co"># apply spatial constraint on cell merging vs. splitting </span></span>
<span>  <span class="co"># default to "leidenCut" for decision based on Leiden clustering of transcript coordinates, alternative use "geometryDiff" for geometric analysis </span></span>
<span>  spatialMergeCheck_method <span class="op">=</span> <span class="st">"leidenCut"</span>, </span>
<span>  <span class="co"># list of leiden clustering configuration to pass to `igraph::cluster_leiden()` function</span></span>
<span>  leiden_config <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>objective_function <span class="op">=</span> <span class="st">"CPM"</span>,</span>
<span>                       resolution_parameter <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                       beta <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>                       n_iterations <span class="op">=</span> <span class="fl">200</span><span class="op">)</span>, </span>
<span>  <span class="co"># minimal percentage of transcripts shared membership between query cell and neighbor cells in leiden clustering results for a valid merging event</span></span>
<span>  cutoff_spatialMerge <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  </span>
<span>  return_intermediates <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co"># flag to return intermediate outputs for neighborhood evaluation and resegmentation actions</span></span>
<span>  return_perCellData <span class="op">=</span> <span class="cn">TRUE</span>,  <span class="co"># flag to return additional per cell data.frame and expression matrix</span></span>
<span>  includeAllRefGenes <span class="op">=</span> <span class="cn">FALSE</span> <span class="co"># flag to include all genes in `score_GeneMatrix` in the returned `updated_perCellExprs` with missing genes of value </span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Perform leiden clustering at resolution_parameter = 1.000.</span></span>
<span><span class="co">#&gt; Create Delanay network when config$method is NULL.</span></span>
<span><span class="co">#&gt; Name the spatial network based on method as `Delaunay_network` when config$name is NULL.</span></span>
<span><span class="co">#&gt; Use neighbor_distance_xy = 20.0000 for searching of neighbor cells.</span></span>
<span><span class="co">#&gt; Use distance_cutoff = 2.0000 for defining direct neighbor cells based on molecule-to-molecule distance.</span></span>
<span><span class="co">#&gt; Use first 2D for searching cell neighborhood, but all 3 Dimension to identify direct neighbors based on molecular distance.</span></span>
<span><span class="co">#&gt; Found 2458 common cells and 960 common genes among transcript_df, cell_networkDT, and score_GeneMatrix.</span></span>
<span><span class="co">#&gt; 1083 chosen_cells are found in common cells.</span></span>
<span><span class="co">#&gt; Use `leidenCut` method to evaluate putative merging event in space.</span></span>
<span><span class="co">#&gt; Perform leiden clustering at resolution_parameter = 1.000.</span></span>
<span><span class="co">#&gt; A valid merging event must have query cell with 0.500 transcript shared same membership as neighbor cell of consistent cell type.</span></span>
<span><span class="co">#&gt; Run delanuay network in 3 Dimension.</span></span>
<span><span class="co">#&gt; Perform ledien clustering on 66 potential merging events.</span></span>
<span><span class="co">#&gt; (`c_1_2_1339_g4`, `c_1_2_1403_g6`) cell pair with all 7 transcripts in same z plane, run 2D network analysis.</span></span>
<span><span class="co">#&gt; (`c_1_2_1403_g6`, `c_1_2_1339_g4`) cell pair with all 7 transcripts in same z plane, run 2D network analysis.</span></span>
<span><span class="co">#&gt; (`c_1_2_921_g8`, `c_1_2_941_g1`) cell pair with all 15 transcripts in same z plane, run 2D network analysis.</span></span>
<span><span class="co">#&gt; (`c_1_2_941_g1`, `c_1_2_921_g8`) cell pair with all 15 transcripts in same z plane, run 2D network analysis.</span></span>
<span><span class="co">#&gt; Found 1084 common cells and 960 common genes among `names(reseg_full_converter)`, `transcript_df`, and `score_GeneMatrix`.</span></span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span><span class="co"># updated transcript data.frame with `updated_cellID` column reflecting the new transcript assignment to cells after refinement</span></span>
<span><span class="co"># only includes transcripts that are intracellular in original segmentation and are genes in `score_GeneMatrix` </span></span>
<span><span class="va">updated_transDF</span> <span class="op">&lt;-</span> <span class="va">finalRes</span><span class="op">[[</span><span class="st">"updated_transDF"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">updated_transDF</span>, n <span class="op">=</span> <span class="fl">3L</span><span class="op">)</span></span>
<span><span class="co">#&gt;    UMI_transID UMI_cellID        x         y   z  target slide fov CellId</span></span>
<span><span class="co">#&gt; 1     t_1_2_10    c_1_2_5 914.8070 -7832.786 2.4 HLA-DRA     1   2      5</span></span>
<span><span class="co">#&gt; 2 t_1_2_100005 c_1_2_1270 841.7000 -8045.011 2.4   HLA-E     1   2   1270</span></span>
<span><span class="co">#&gt; 3 t_1_2_100006 c_1_2_1270 844.4609 -8045.006 2.4  TWIST1     1   2   1270</span></span>
<span><span class="co">#&gt;   tLLR_maxCellType score_tLLR_maxCellType connect_group tmp_cellID</span></span>
<span><span class="co">#&gt; 1                f             -0.4227132             0    c_1_2_5</span></span>
<span><span class="co">#&gt; 2                d             -0.5070020             0 c_1_2_1270</span></span>
<span><span class="co">#&gt; 3                d             -2.2769526             0 c_1_2_1270</span></span>
<span><span class="co">#&gt;   group_maxCellType updated_cellID updated_celltype score_updated_celltype</span></span>
<span><span class="co">#&gt; 1                 f        c_1_2_5                f             -0.4227132</span></span>
<span><span class="co">#&gt; 2                 d     c_1_2_1270                d             -0.5070020</span></span>
<span><span class="co">#&gt; 3                 d     c_1_2_1270                d             -2.2769526</span></span></code></pre></div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># updated per cell data.table after refinement, with resegmentation operation recorded in `reSeg_action`</span></span>
<span><span class="va">updated_perCellDT</span> <span class="op">&lt;-</span> <span class="va">finalRes</span><span class="op">[[</span><span class="st">'updated_perCellDT'</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">updated_perCellDT</span><span class="op">)</span></span>
<span><span class="co">#&gt; Key: &lt;updated_cellID&gt;</span></span>
<span><span class="co">#&gt;    updated_cellID updated_celltype         x         y        z reSeg_action</span></span>
<span><span class="co">#&gt;            &lt;char&gt;           &lt;char&gt;     &lt;num&gt;     &lt;num&gt;    &lt;num&gt;       &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:     c_1_2_1000                d 1061.5538 -7993.376 2.691633         none</span></span>
<span><span class="co">#&gt; 2:     c_1_2_1001                f 1195.9138 -7993.106 4.192908         none</span></span>
<span><span class="co">#&gt; 3:     c_1_2_1009                d  807.5324 -7996.263 3.311172         none</span></span>
<span><span class="co">#&gt; 4:     c_1_2_1010                d  928.8692 -7996.915 3.523333         none</span></span>
<span><span class="co">#&gt; 5:     c_1_2_1011                d 1135.7938 -7996.114 2.759471         none</span></span>
<span><span class="co">#&gt; 6:     c_1_2_1012                d 1146.3545 -7996.048 2.354226         trim</span></span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">updated_perCellDT</span><span class="op">[[</span><span class="st">'reSeg_action'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; merge_or_flagged              new             none             trim </span></span>
<span><span class="co">#&gt;                4                3             1017              354</span></span></code></pre></div>
<p><code><a href="../reference/runSegRefinement.html">runSegRefinement()</a></code> function returns a list of outputs
with updated cell segmentation that could be used in further single cell
analysis.</p>
<table class="table table table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
variable
</th>
<th style="text-align:left;">
description
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;width: 25%; white-space: pre-line; text-align: left;">
<code>updated_transDF</code>
</td>
<td style="text-align:left;width: 70%; white-space: pre-line; text-align: left;">
the updated transcript_df with <code>updated_cellID</code> and
<code>updated_celltype</code> columns reflecting the new segmentation
outcomes
</td>
</tr>
<tr>
<td style="text-align:left;width: 25%; white-space: pre-line; text-align: left;">
<code>neighborhoodDF_ToReseg</code>
</td>
<td style="text-align:left;width: 70%; white-space: pre-line; text-align: left;">
data.frame for neighborhood environment of low-score transcript groups,
output of <code>neighborhood_for_resegment_spatstat()</code> function,
return when <code>return_intermediates = TRUE</code>
</td>
</tr>
<tr>
<td style="text-align:left;width: 25%; white-space: pre-line; text-align: left;">
<code>reseg_actions</code>
</td>
<td style="text-align:left;width: 70%; white-space: pre-line; text-align: left;">
<i>list of 4 elements describing how the resegmenation would be
performed on <code>reseg_transcript_df</code> by the group assignment of
transcripts listed in <code>reseg_cellID_coln</code>, return when
<code>return_intermediates = TRUE</code></i>
</td>
</tr>
<tr>
<td style="text-align:left;width: 25%; white-space: pre-line; text-align: left;">
</td>
<td style="text-align:left;width: 70%; white-space: pre-line; text-align: left;">
<ul>
<li>
<code>cells_to_discard</code>: a vector of cell ID that should be
discarded during resegmentation.
</li>
</ul>
<ul>
<li>
<code>cells_to_update</code>: a named vector of cell ID whether the
cell_ID in name would be replaced with new cell_ID in value.
</li>
</ul>
<ul>
<li>
<code>cells_to_keep</code>: a vector of cell ID that should be kept as
it is.
</li>
</ul>
<ul>
<li>
<code>reseg_full_converter</code>: a single named vector of cell ID to
update the original cell ID, assign <code>NA</code> for
<code>cells_to_discard</code>.
</li>
</ul>
</td>
</tr>
<tr>
<td style="text-align:left;width: 25%; white-space: pre-line; text-align: left;">
<code>updated_perCellDT</code>
</td>
<td style="text-align:left;width: 70%; white-space: pre-line; text-align: left;">
per cell data.table with mean spatial coordinates, new cell type and
resegmentation action after resegmentation, return when
<code>return_perCellData = TRUE</code>
</td>
</tr>
<tr>
<td style="text-align:left;width: 25%; white-space: pre-line; text-align: left;">
<code>updated_perCellExprs</code>
</td>
<td style="text-align:left;width: 70%; white-space: pre-line; text-align: left;">
gene x cell count sparse matrix for updated transcript data.frame after
resegmentation, return when <code>return_perCellData = TRUE</code>
</td>
</tr>
</tbody>
</table>
<div class="section level5">
<h5 id="visualize-example-resegmentation-outcomes">Visualize example resegmentation outcomes<a class="anchor" aria-label="anchor" href="#visualize-example-resegmentation-outcomes"></a>
</h5>
<p>Letâ€™s check on a few cells which got changed by the segmentation
refinement process.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># choose example cells got segmentation refinement operations of various kinds</span></span>
<span><span class="va">trimmed_cells_to_plot</span> <span class="op">&lt;-</span> <span class="va">updated_perCellDT</span><span class="op">[</span><span class="va">reSeg_action</span> <span class="op">==</span> <span class="st">"trim"</span> <span class="op">&amp;</span> <span class="va">updated_cellID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cells_to_plot</span><span class="op">)</span>, <span class="va">updated_cellID</span><span class="op">]</span></span>
<span><span class="va">changes_to_plot</span> <span class="op">&lt;-</span> <span class="va">updated_perCellDT</span><span class="op">[</span><span class="op">!</span><span class="va">reSeg_action</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"none"</span>, <span class="st">"trim"</span><span class="op">)</span> <span class="op">|</span> <span class="va">updated_cellID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">trimmed_cells_to_plot</span> , <span class="va">.SD</span>, .SDcols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'updated_cellID'</span>, <span class="st">'reSeg_action'</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">changes_to_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">changes_to_plot</span>, </span>
<span>                         <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">updated_transDF</span><span class="op">[</span><span class="va">updated_transDF</span><span class="op">[[</span><span class="st">'updated_cellID'</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">changes_to_plot</span><span class="op">[[</span><span class="st">'updated_cellID'</span><span class="op">]</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'UMI_cellID'</span>, <span class="st">'updated_cellID'</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">changes_to_plot</span><span class="op">[</span>, <span class="va">oriCellNum_in_newCell</span> <span class="op">:=</span> <span class="va">.N</span>, by <span class="op">=</span> <span class="va">updated_cellID</span> <span class="op">]</span></span>
<span><span class="va">changes_to_plot</span><span class="op">[</span>, <span class="va">newCellNum_from_oriCell</span> <span class="op">:=</span> <span class="va">.N</span>, by <span class="op">=</span> <span class="va">UMI_cellID</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## segmentation changes for the selected cells with original cell_ID in `UMI_cellID` column</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">changes_to_plot</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">reSeg_action</span>, <span class="op">-</span> <span class="va">newCellNum_from_oriCell</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;     updated_cellID     reSeg_action UMI_cellID oriCellNum_in_newCell</span></span>
<span><span class="co">#&gt;             &lt;char&gt;           &lt;char&gt;     &lt;char&gt;                 &lt;int&gt;</span></span>
<span><span class="co">#&gt;  1:  c_1_2_1481_g5 merge_or_flagged c_1_2_1493                     2</span></span>
<span><span class="co">#&gt;  2:  c_1_2_1481_g5 merge_or_flagged c_1_2_1481                     2</span></span>
<span><span class="co">#&gt;  3:     c_1_2_1655 merge_or_flagged c_1_2_1655                     2</span></span>
<span><span class="co">#&gt;  4:     c_1_2_1655 merge_or_flagged c_1_2_1668                     2</span></span>
<span><span class="co">#&gt;  5:     c_1_2_1683 merge_or_flagged c_1_2_1683                     2</span></span>
<span><span class="co">#&gt;  6:     c_1_2_1683 merge_or_flagged c_1_2_1732                     2</span></span>
<span><span class="co">#&gt;  7:     c_1_2_2243 merge_or_flagged c_1_2_2243                     1</span></span>
<span><span class="co">#&gt;  8:  c_1_2_2549_g2              new c_1_2_2549                     1</span></span>
<span><span class="co">#&gt;  9:  c_1_2_1502_g1              new c_1_2_1502                     1</span></span>
<span><span class="co">#&gt; 10:  c_1_2_1758_g1              new c_1_2_1758                     1</span></span>
<span><span class="co">#&gt; 11:     c_1_2_2549             trim c_1_2_2549                     1</span></span>
<span><span class="co">#&gt; 12:     c_1_2_1341             trim c_1_2_1341                     1</span></span>
<span><span class="co">#&gt; 13:     c_1_2_1485             trim c_1_2_1485                     1</span></span>
<span><span class="co">#&gt; 14:     c_1_2_1646             trim c_1_2_1646                     1</span></span>
<span><span class="co">#&gt; 15:     c_1_2_1751             trim c_1_2_1751                     1</span></span>
<span><span class="co">#&gt; 16:     c_1_2_1763             trim c_1_2_1763                     1</span></span>
<span><span class="co">#&gt; 17:     c_1_2_2062             trim c_1_2_2062                     1</span></span>
<span><span class="co">#&gt; 18:     c_1_2_2332             trim c_1_2_2332                     1</span></span>
<span><span class="co">#&gt; 19:     c_1_2_2623             trim c_1_2_2623                     1</span></span>
<span><span class="co">#&gt; 20:     c_1_2_2688             trim c_1_2_2688                     1</span></span>
<span><span class="co">#&gt; 21:      c_1_2_352             trim  c_1_2_352                     1</span></span>
<span><span class="co">#&gt; 22:      c_1_2_392             trim  c_1_2_392                     1</span></span>
<span><span class="co">#&gt; 23:     c_1_2_4147             trim c_1_2_4147                     1</span></span>
<span><span class="co">#&gt; 24:      c_1_2_521             trim  c_1_2_521                     1</span></span>
<span><span class="co">#&gt; 25:      c_1_2_714             trim  c_1_2_714                     1</span></span>
<span><span class="co">#&gt;     updated_cellID     reSeg_action UMI_cellID oriCellNum_in_newCell</span></span>
<span><span class="co">#&gt;     newCellNum_from_oriCell</span></span>
<span><span class="co">#&gt;                       &lt;int&gt;</span></span>
<span><span class="co">#&gt;  1:                       1</span></span>
<span><span class="co">#&gt;  2:                       1</span></span>
<span><span class="co">#&gt;  3:                       1</span></span>
<span><span class="co">#&gt;  4:                       1</span></span>
<span><span class="co">#&gt;  5:                       1</span></span>
<span><span class="co">#&gt;  6:                       1</span></span>
<span><span class="co">#&gt;  7:                       1</span></span>
<span><span class="co">#&gt;  8:                       2</span></span>
<span><span class="co">#&gt;  9:                       1</span></span>
<span><span class="co">#&gt; 10:                       1</span></span>
<span><span class="co">#&gt; 11:                       2</span></span>
<span><span class="co">#&gt; 12:                       1</span></span>
<span><span class="co">#&gt; 13:                       1</span></span>
<span><span class="co">#&gt; 14:                       1</span></span>
<span><span class="co">#&gt; 15:                       1</span></span>
<span><span class="co">#&gt; 16:                       1</span></span>
<span><span class="co">#&gt; 17:                       1</span></span>
<span><span class="co">#&gt; 18:                       1</span></span>
<span><span class="co">#&gt; 19:                       1</span></span>
<span><span class="co">#&gt; 20:                       1</span></span>
<span><span class="co">#&gt; 21:                       1</span></span>
<span><span class="co">#&gt; 22:                       1</span></span>
<span><span class="co">#&gt; 23:                       1</span></span>
<span><span class="co">#&gt; 24:                       1</span></span>
<span><span class="co">#&gt; 25:                       1</span></span>
<span><span class="co">#&gt;     newCellNum_from_oriCell</span></span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span><span class="co"># get all transcript in the source cells for plotting</span></span>
<span><span class="va">transDF_to_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span></span>
<span>  <span class="va">reSeg_ready_res</span><span class="op">[[</span><span class="st">"reseg_transcript_df"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">reSeg_ready_res</span><span class="op">[[</span><span class="st">"reseg_transcript_df"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"UMI_cellID"</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">changes_to_plot</span><span class="op">[[</span><span class="st">'UMI_cellID'</span><span class="op">]</span><span class="op">]</span>, <span class="op">]</span>, </span>
<span>  <span class="va">updated_transDF</span><span class="op">[</span><span class="va">updated_transDF</span><span class="op">[[</span><span class="st">"UMI_cellID"</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">changes_to_plot</span><span class="op">[[</span><span class="st">'UMI_cellID'</span><span class="op">]</span><span class="op">]</span>, <span class="op">]</span>, </span>
<span>  all.x <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># label changed with respect to original cell segmentation</span></span>
<span><span class="va">trim_idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">transDF_to_plot</span><span class="op">[[</span><span class="st">'updated_cellID'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">alter_idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">transDF_to_plot</span><span class="op">[[</span><span class="st">'updated_cellID'</span><span class="op">]</span><span class="op">]</span> <span class="op">!=</span> <span class="va">transDF_to_plot</span><span class="op">[[</span><span class="st">'UMI_cellID'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">transDF_to_plot</span><span class="op">[[</span><span class="st">'changed'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'none'</span></span>
<span><span class="va">transDF_to_plot</span><span class="op">[[</span><span class="st">'changed'</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">alter_idx</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'move_to_others'</span></span>
<span><span class="va">transDF_to_plot</span><span class="op">[[</span><span class="st">'changed'</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">trim_idx</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'trim'</span></span>
<span></span>
<span><span class="co"># spatial plot for modification on original source cells</span></span>
<span><span class="va">cells_to_plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">changes_to_plot</span><span class="op">[[</span><span class="st">'UMI_cellID'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotSpatialScoreMultiCells.html">plotSpatialScoreMultiCells</a></span><span class="op">(</span>chosen_cells <span class="op">=</span> <span class="va">cells_to_plot2</span>,</span>
<span>                           cell_labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">modStats_ToFlagCells</span><span class="op">[</span><span class="va">cells_to_plot2</span>, <span class="st">'lrtest_nlog10P'</span><span class="op">]</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>                           transcript_df <span class="op">=</span> <span class="va">transDF_to_plot</span>,</span>
<span>                           cellID_coln <span class="op">=</span> <span class="st">"UMI_cellID"</span>,</span>
<span>                           transID_coln <span class="op">=</span> <span class="st">"UMI_transID"</span>,</span>
<span>                           score_coln <span class="op">=</span> <span class="st">"changed"</span>,</span>
<span>                           spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>,<span class="st">"y"</span><span class="op">)</span>,</span>
<span>                           point_size <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>                           plot_discrete <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                           title <span class="op">=</span> <span class="st">"examples of orignal cells got trimmed or split"</span><span class="op">)</span></span></code></pre></div>
<p><img src="figures%2Ffig_newSeg-1.png" width="100%"></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># spatial plot for new cells merged from multiple original cells</span></span>
<span><span class="va">cells_to_plot2</span> <span class="op">&lt;-</span> <span class="va">changes_to_plot</span><span class="op">[</span><span class="va">oriCellNum_in_newCell</span> <span class="op">&gt;</span><span class="fl">1</span>,<span class="op">]</span></span>
<span><span class="va">fig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotSpatialScoreMultiCells.html">plotSpatialScoreMultiCells</a></span><span class="op">(</span>chosen_cells <span class="op">=</span> <span class="va">cells_to_plot2</span><span class="op">[[</span><span class="st">'updated_cellID'</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                                  cell_labels <span class="op">=</span> <span class="va">cells_to_plot2</span><span class="op">[[</span><span class="st">'reSeg_action'</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                                  transcript_df <span class="op">=</span> <span class="va">transDF_to_plot</span>,</span>
<span>                                  cellID_coln <span class="op">=</span> <span class="st">"updated_cellID"</span>,</span>
<span>                                  transID_coln <span class="op">=</span> <span class="st">"UMI_transID"</span>,</span>
<span>                                  score_coln <span class="op">=</span> <span class="st">"UMI_cellID"</span>,</span>
<span>                                  spatLocs_colns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>,<span class="st">"y"</span><span class="op">)</span>,</span>
<span>                                  point_size <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                                  plot_discrete <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                                  title <span class="op">=</span> <span class="st">"examples of new cells merged from multiple orignal cells"</span><span class="op">)</span>   </span>
<span><span class="va">fig</span> <span class="op">&lt;-</span> <span class="va">fig</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">fig</span><span class="op">)</span></span></code></pre></div>
<p><img src="figures%2Ffig_merge-1.png" width="100%"></p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Etc/UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] FastReseg_1.0.3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] gtable_0.3.6         xfun_0.48            bslib_0.8.0         </span></span>
<span><span class="co">#&gt;  [4] ggplot2_3.5.1        htmlwidgets_1.6.2    lattice_0.22-6      </span></span>
<span><span class="co">#&gt;  [7] vctrs_0.6.3          tools_4.4.1          spatstat.utils_3.1-0</span></span>
<span><span class="co">#&gt; [10] generics_0.1.3       parallel_4.4.1       proxy_0.4-27        </span></span>
<span><span class="co">#&gt; [13] tibble_3.2.1         fansi_1.0.6          highr_0.11          </span></span>
<span><span class="co">#&gt; [16] pkgconfig_2.0.3      Matrix_1.6-5         data.table_1.16.2   </span></span>
<span><span class="co">#&gt; [19] checkmate_2.3.2      RColorBrewer_1.1-3   desc_1.4.3          </span></span>
<span><span class="co">#&gt; [22] lifecycle_1.0.4      compiler_4.4.1       farver_2.1.2        </span></span>
<span><span class="co">#&gt; [25] stringr_1.4.0        deldir_2.0-4         GiottoUtils_0.2.4   </span></span>
<span><span class="co">#&gt; [28] textshaping_0.4.0    munsell_0.5.1        terra_1.7-39        </span></span>
<span><span class="co">#&gt; [31] codetools_0.2-20     class_7.3-22         htmltools_0.5.8.1   </span></span>
<span><span class="co">#&gt; [34] GiottoClass_0.2.3    sass_0.4.9           yaml_2.3.10         </span></span>
<span><span class="co">#&gt; [37] pillar_1.9.0         pkgdown_2.1.1        jquerylib_0.1.4     </span></span>
<span><span class="co">#&gt; [40] cachem_1.1.0         dbscan_1.1-10        abind_1.4-5         </span></span>
<span><span class="co">#&gt; [43] spatstat.geom_2.4-0  gtools_3.9.5         tidyselect_1.2.1    </span></span>
<span><span class="co">#&gt; [46] digest_0.6.29        stringi_1.8.4        reshape2_1.4.4      </span></span>
<span><span class="co">#&gt; [49] dplyr_1.0.10         magic_1.6-1          labeling_0.4.3      </span></span>
<span><span class="co">#&gt; [52] polyclip_1.10-7      fastmap_1.2.0        grid_4.4.1          </span></span>
<span><span class="co">#&gt; [55] colorspace_2.1-1     cli_3.6.3            concaveman_1.1.0    </span></span>
<span><span class="co">#&gt; [58] magrittr_2.0.3       utf8_1.2.4           e1071_1.7-16        </span></span>
<span><span class="co">#&gt; [61] spatstat.data_3.1-2  withr_3.0.2          scales_1.3.0        </span></span>
<span><span class="co">#&gt; [64] backports_1.5.0      rmarkdown_2.26       igraph_2.1.1        </span></span>
<span><span class="co">#&gt; [67] ragg_1.3.3           zoo_1.8-12           kableExtra_1.4.0    </span></span>
<span><span class="co">#&gt; [70] evaluate_1.0.1       knitr_1.46           lmtest_0.9-40       </span></span>
<span><span class="co">#&gt; [73] geometry_0.4.7       viridisLite_0.4.2    rlang_1.1.1         </span></span>
<span><span class="co">#&gt; [76] Rcpp_1.0.12          glue_1.8.0           xml2_1.3.6          </span></span>
<span><span class="co">#&gt; [79] svglite_2.1.3        rstudioapi_0.17.1    jsonlite_1.8.9      </span></span>
<span><span class="co">#&gt; [82] plyr_1.8.7           R6_2.5.1             systemfonts_1.1.0   </span></span>
<span><span class="co">#&gt; [85] fs_1.5.2</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Patrick Danaher, Lidan Wu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
